<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>9.6 ChIP Quality Control | Computational Genomics with R</title>
  <meta name="description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="9.6 ChIP Quality Control | Computational Genomics with R" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://compmgenomr.github.io/book/" />
  <meta property="og:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />
  <meta property="og:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="github-repo" content="compgenomr/bookdown" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9.6 ChIP Quality Control | Computational Genomics with R" />
  
  <meta name="twitter:description" content="A guide to computationa genomics using R. The book covers fundemental topics with practical examples for an interdisciplinery audience" />
  <meta name="twitter:image" content="https://compmgenomr.github.io/book/images/cover.jpg" />

<meta name="author" content="Altuna Akalin">


<meta name="date" content="2019-09-24">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="computational-analysis-of-chip-seq-data.html">
<link rel="next" href="peak-calling.html">
<script src="libs/jquery/jquery.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />







<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-83786243-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-83786243-1');
</script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Computational Genomics with R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html"><i class="fa fa-check"></i>Who is this book for?</a><ul>
<li class="chapter" data-level="" data-path="who-is-this-book-for.html"><a href="who-is-this-book-for.html#what-will-you-get-out-of-this"><i class="fa fa-check"></i>What will you get out of this?</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="structure-of-the-book.html"><a href="structure-of-the-book.html"><i class="fa fa-check"></i>Structure of the book</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html"><i class="fa fa-check"></i>Software information and conventions</a><ul>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#assignment-operator-convention"><i class="fa fa-check"></i>Assignment operator convention</a></li>
<li class="chapter" data-level="" data-path="software-information-and-conventions.html"><a href="software-information-and-conventions.html#packages-needed-to-run-the-book-code"><i class="fa fa-check"></i>Packages needed to run the book code</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="data-for-the-book.html"><a href="data-for-the-book.html"><i class="fa fa-check"></i>Data for the book</a></li>
<li class="chapter" data-level="" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i>Acknowledgements</a></li>
<li class="chapter" data-level="" data-path="how-to-contribute.html"><a href="how-to-contribute.html"><i class="fa fa-check"></i>How to contribute</a></li>
<li class="chapter" data-level="" data-path="trainings-and-consultation.html"><a href="trainings-and-consultation.html"><i class="fa fa-check"></i>Trainings and consultation</a><ul>
<li class="chapter" data-level="" data-path="trainings-and-consultation.html"><a href="trainings-and-consultation.html#custom-training-and-workshops"><i class="fa fa-check"></i>Custom Training and Workshops</a></li>
<li class="chapter" data-level="" data-path="trainings-and-consultation.html"><a href="trainings-and-consultation.html#consultation"><i class="fa fa-check"></i>Consultation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="about-the-authors.html"><a href="about-the-authors.html"><i class="fa fa-check"></i>About the Authors</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction to Genomics</a><ul>
<li class="chapter" data-level="1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html"><i class="fa fa-check"></i><b>1.1</b> Genes, DNA and central dogma</a><ul>
<li class="chapter" data-level="1.1.1" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-genome"><i class="fa fa-check"></i><b>1.1.1</b> What is a genome?</a></li>
<li class="chapter" data-level="1.1.2" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-is-a-gene"><i class="fa fa-check"></i><b>1.1.2</b> What is a gene?</a></li>
<li class="chapter" data-level="1.1.3" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#how-genes-are-controlled-the-transcriptional-and-the-post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.1.3</b> How genes are controlled ? The transcriptional and the post-transcriptional regulation</a></li>
<li class="chapter" data-level="1.1.4" data-path="genes-dna-and-central-dogma.html"><a href="genes-dna-and-central-dogma.html#what-does-a-gene-look-like"><i class="fa fa-check"></i><b>1.1.4</b> What does a gene look like?</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html"><i class="fa fa-check"></i><b>1.2</b> Elements of gene regulation</a><ul>
<li class="chapter" data-level="1.2.1" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.1</b> Transcriptional regulation</a></li>
<li class="chapter" data-level="1.2.2" data-path="elements-of-gene-regulation.html"><a href="elements-of-gene-regulation.html#post-transcriptional-regulation"><i class="fa fa-check"></i><b>1.2.2</b> Post-transcriptional regulation</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="shaping-the-genome-dna-mutation.html"><a href="shaping-the-genome-dna-mutation.html"><i class="fa fa-check"></i><b>1.3</b> Shaping the genome: DNA mutation</a></li>
<li class="chapter" data-level="1.4" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html"><i class="fa fa-check"></i><b>1.4</b> High-throughput experimental methods in genomics</a><ul>
<li class="chapter" data-level="1.4.1" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#the-general-idea-behind-high-throughput-techniques"><i class="fa fa-check"></i><b>1.4.1</b> The general idea behind high-throughput techniques</a></li>
<li class="chapter" data-level="1.4.2" data-path="high-throughput-experimental-methods-in-genomics.html"><a href="high-throughput-experimental-methods-in-genomics.html#high-throughput-sequencing"><i class="fa fa-check"></i><b>1.4.2</b> High-throughput sequencing</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="visualization-and-data-repositories-for-genomics.html"><a href="visualization-and-data-repositories-for-genomics.html"><i class="fa fa-check"></i><b>1.5</b> Visualization and data repositories for genomics</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="Rintro.html"><a href="Rintro.html"><i class="fa fa-check"></i><b>2</b> Introduction to R for Genomic Data Analysis</a><ul>
<li class="chapter" data-level="2.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html"><i class="fa fa-check"></i><b>2.1</b> Steps of (genomic) data analysis</a><ul>
<li class="chapter" data-level="2.1.1" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-collection"><i class="fa fa-check"></i><b>2.1.1</b> Data collection</a></li>
<li class="chapter" data-level="2.1.2" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-quality-check-and-cleaning"><i class="fa fa-check"></i><b>2.1.2</b> Data quality check and cleaning</a></li>
<li class="chapter" data-level="2.1.3" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#data-processing"><i class="fa fa-check"></i><b>2.1.3</b> Data processing</a></li>
<li class="chapter" data-level="2.1.4" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#exploratory-data-analysis-and-modeling"><i class="fa fa-check"></i><b>2.1.4</b> Exploratory data analysis and modeling</a></li>
<li class="chapter" data-level="2.1.5" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#visualization-and-reporting"><i class="fa fa-check"></i><b>2.1.5</b> Visualization and reporting</a></li>
<li class="chapter" data-level="2.1.6" data-path="steps-of-genomic-data-analysis.html"><a href="steps-of-genomic-data-analysis.html#why-use-r-for-genomics"><i class="fa fa-check"></i><b>2.1.6</b> Why use R for genomics ?</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2.2</b> Getting started with R</a><ul>
<li class="chapter" data-level="2.2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages"><i class="fa fa-check"></i><b>2.2.1</b> Installing packages</a></li>
<li class="chapter" data-level="2.2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-in-custom-locations"><i class="fa fa-check"></i><b>2.2.2</b> Installing packages in custom locations</a></li>
<li class="chapter" data-level="2.2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-on-functions-and-packages"><i class="fa fa-check"></i><b>2.2.3</b> Getting help on functions and packages</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="computations-in-r.html"><a href="computations-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Computations in R</a></li>
<li class="chapter" data-level="2.4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>2.4</b> Data structures</a><ul>
<li class="chapter" data-level="2.4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>2.4.1</b> Vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>2.4.2</b> Matrices</a></li>
<li class="chapter" data-level="2.4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>2.4.3</b> Data Frames</a></li>
<li class="chapter" data-level="2.4.4" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>2.4.4</b> Lists</a></li>
<li class="chapter" data-level="2.4.5" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>2.4.5</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="data-types.html"><a href="data-types.html"><i class="fa fa-check"></i><b>2.5</b> Data types</a></li>
<li class="chapter" data-level="2.6" data-path="reading-and-writing-data.html"><a href="reading-and-writing-data.html"><i class="fa fa-check"></i><b>2.6</b> Reading and writing data</a></li>
<li class="chapter" data-level="2.7" data-path="plotting-in-r.html"><a href="plotting-in-r.html"><i class="fa fa-check"></i><b>2.7</b> Plotting in R</a></li>
<li class="chapter" data-level="2.8" data-path="saving-plots.html"><a href="saving-plots.html"><i class="fa fa-check"></i><b>2.8</b> Saving plots</a></li>
<li class="chapter" data-level="2.9" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html"><i class="fa fa-check"></i><b>2.9</b> Functions and control structures (for, if/else etc.)</a><ul>
<li class="chapter" data-level="2.9.1" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#user-defined-functions"><i class="fa fa-check"></i><b>2.9.1</b> User defined functions</a></li>
<li class="chapter" data-level="2.9.2" data-path="functions-and-control-structures-for-ifelse-etc-.html"><a href="functions-and-control-structures-for-ifelse-etc-.html#loops-and-looping-structures-in-r"><i class="fa fa-check"></i><b>2.9.2</b> Loops and looping structures in R</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>2.10</b> Exercises</a><ul>
<li class="chapter" data-level="2.10.1" data-path="exercises.html"><a href="exercises.html#computations-in-r-1"><i class="fa fa-check"></i><b>2.10.1</b> Computations in R</a></li>
<li class="chapter" data-level="2.10.2" data-path="exercises.html"><a href="exercises.html#data-structures-in-r"><i class="fa fa-check"></i><b>2.10.2</b> Data structures in R</a></li>
<li class="chapter" data-level="2.10.3" data-path="exercises.html"><a href="exercises.html#reading-in-and-writing-data-out-in-r"><i class="fa fa-check"></i><b>2.10.3</b> Reading in and writing data out in R</a></li>
<li class="chapter" data-level="2.10.4" data-path="exercises.html"><a href="exercises.html#plotting-in-r-1"><i class="fa fa-check"></i><b>2.10.4</b> Plotting in R</a></li>
<li class="chapter" data-level="2.10.5" data-path="exercises.html"><a href="exercises.html#functions-and-control-structures-for-ifelse-etc.-1"><i class="fa fa-check"></i><b>2.10.5</b> Functions and control structures (for, if/else etc.)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="stats.html"><a href="stats.html"><i class="fa fa-check"></i><b>3</b> Statistics for Genomics</a><ul>
<li class="chapter" data-level="3.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><i class="fa fa-check"></i><b>3.1</b> How to summarize collection of data points: The idea behind statistical distributions</a><ul>
<li class="chapter" data-level="3.1.1" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-central-tendency-mean-and-median"><i class="fa fa-check"></i><b>3.1.1</b> Describing the central tendency: mean and median</a></li>
<li class="chapter" data-level="3.1.2" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#describing-the-spread-measurements-of-variation"><i class="fa fa-check"></i><b>3.1.2</b> Describing the spread: measurements of variation</a></li>
<li class="chapter" data-level="3.1.3" data-path="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html"><a href="how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions.html#precision-of-estimates-confidence-intervals"><i class="fa fa-check"></i><b>3.1.3</b> Precision of estimates: Confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html"><i class="fa fa-check"></i><b>3.2</b> How to test for differences between samples</a><ul>
<li class="chapter" data-level="3.2.1" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#randomization-based-testing-for-difference-of-the-means"><i class="fa fa-check"></i><b>3.2.1</b> randomization based testing for difference of the means</a></li>
<li class="chapter" data-level="3.2.2" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#using-t-test-for-difference-of-the-means-between-two-samples"><i class="fa fa-check"></i><b>3.2.2</b> Using t-test for difference of the means between two samples</a></li>
<li class="chapter" data-level="3.2.3" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#multiple-testing-correction"><i class="fa fa-check"></i><b>3.2.3</b> multiple testing correction</a></li>
<li class="chapter" data-level="3.2.4" data-path="how-to-test-for-differences-between-samples.html"><a href="how-to-test-for-differences-between-samples.html#moderated-t-tests-using-information-from-multiple-comparisons"><i class="fa fa-check"></i><b>3.2.4</b> moderated t-tests: using information from multiple comparisons</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html"><i class="fa fa-check"></i><b>3.3</b> Relationship between variables: linear models and correlation</a><ul>
<li class="chapter" data-level="3.3.1" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-fit-a-line"><i class="fa fa-check"></i><b>3.3.1</b> How to fit a line</a></li>
<li class="chapter" data-level="3.3.2" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#how-to-estimate-the-error-of-the-coefficients"><i class="fa fa-check"></i><b>3.3.2</b> How to estimate the error of the coefficients</a></li>
<li class="chapter" data-level="3.3.3" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#accuracy-of-the-model"><i class="fa fa-check"></i><b>3.3.3</b> Accuracy of the model</a></li>
<li class="chapter" data-level="3.3.4" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-with-categorical-variables"><i class="fa fa-check"></i><b>3.3.4</b> Regression with categorical variables</a></li>
<li class="chapter" data-level="3.3.5" data-path="relationship-between-variables-linear-models-and-correlation.html"><a href="relationship-between-variables-linear-models-and-correlation.html#regression-pitfalls"><i class="fa fa-check"></i><b>3.3.5</b> Regression pitfalls</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="exercises-1.html"><a href="exercises-1.html"><i class="fa fa-check"></i><b>3.4</b> Exercises</a><ul>
<li class="chapter" data-level="3.4.1" data-path="exercises-1.html"><a href="exercises-1.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1"><i class="fa fa-check"></i><b>3.4.1</b> How to summarize collection of data points: The idea behind statistical distributions</a></li>
<li class="chapter" data-level="3.4.2" data-path="exercises-1.html"><a href="exercises-1.html#how-to-test-for-differences-in-samples"><i class="fa fa-check"></i><b>3.4.2</b> How to test for differences in samples</a></li>
<li class="chapter" data-level="3.4.3" data-path="exercises-1.html"><a href="exercises-1.html#relationship-between-variables-linear-models-and-correlation-1"><i class="fa fa-check"></i><b>3.4.3</b> Relationship between variables: linear models and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="unsupervisedLearning.html"><a href="unsupervisedLearning.html"><i class="fa fa-check"></i><b>4</b> Exploratory Data Analysis with Unsupervised Machine Learning</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html"><i class="fa fa-check"></i><b>4.1</b> Clustering: grouping samples based on their similarity</a><ul>
<li class="chapter" data-level="4.1.1" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#distance-metrics"><i class="fa fa-check"></i><b>4.1.1</b> Distance metrics</a></li>
<li class="chapter" data-level="4.1.2" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#hiearchical-clustering"><i class="fa fa-check"></i><b>4.1.2</b> Hiearchical clustering</a></li>
<li class="chapter" data-level="4.1.3" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#k-means-clustering"><i class="fa fa-check"></i><b>4.1.3</b> K-means clustering</a></li>
<li class="chapter" data-level="4.1.4" data-path="clustering-grouping-samples-based-on-their-similarity.html"><a href="clustering-grouping-samples-based-on-their-similarity.html#how-to-choose-k-the-number-of-clusters"><i class="fa fa-check"></i><b>4.1.4</b> how to choose “k”, the number of clusters</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><i class="fa fa-check"></i><b>4.2</b> Dimensionality reduction techniques: visualizing complex data sets in 2D</a><ul>
<li class="chapter" data-level="4.2.1" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#principal-component-analysis"><i class="fa fa-check"></i><b>4.2.1</b> Principal component analysis</a></li>
<li class="chapter" data-level="4.2.2" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#other-dimension-reduction-techniques-using-other-matrix-factorization-methods"><i class="fa fa-check"></i><b>4.2.2</b> Other dimension reduction techniques using other matrix factorization methods</a></li>
<li class="chapter" data-level="4.2.3" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#multi-dimensional-scaling"><i class="fa fa-check"></i><b>4.2.3</b> Multi-dimensional scaling</a></li>
<li class="chapter" data-level="4.2.4" data-path="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html"><a href="dimensionality-reduction-techniques-visualizing-complex-data-sets-in-2d.html#t-distributed-stochastic-neighbor-embedding-t-sne"><i class="fa fa-check"></i><b>4.2.4</b> t-Distributed Stochastic Neighbor Embedding (t-SNE)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="exercises-2.html"><a href="exercises-2.html"><i class="fa fa-check"></i><b>4.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="supervisedLearning.html"><a href="supervisedLearning.html"><i class="fa fa-check"></i><b>5</b> Predictive Modeling with Supervised Machine Learning</a><ul>
<li class="chapter" data-level="5.1" data-path="how-machine-learning-models-are-fit.html"><a href="how-machine-learning-models-are-fit.html"><i class="fa fa-check"></i><b>5.1</b> How machine learning models are fit?</a><ul>
<li class="chapter" data-level="5.1.1" data-path="how-machine-learning-models-are-fit.html"><a href="how-machine-learning-models-are-fit.html#machine-learning-vs-statistics"><i class="fa fa-check"></i><b>5.1.1</b> Machine learning vs Statistics</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="steps-in-supervised-machine-learning.html"><a href="steps-in-supervised-machine-learning.html"><i class="fa fa-check"></i><b>5.2</b> Steps in supervised machine learning</a></li>
<li class="chapter" data-level="5.3" data-path="use-case-disease-subtype-from-genomics-data.html"><a href="use-case-disease-subtype-from-genomics-data.html"><i class="fa fa-check"></i><b>5.3</b> Use case: Disease subtype from genomics data</a></li>
<li class="chapter" data-level="5.4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>5.4</b> Data preprocessing</a><ul>
<li class="chapter" data-level="5.4.1" data-path="data-preprocessing.html"><a href="data-preprocessing.html#data-transformation"><i class="fa fa-check"></i><b>5.4.1</b> data transformation</a></li>
<li class="chapter" data-level="5.4.2" data-path="data-preprocessing.html"><a href="data-preprocessing.html#filtering-data-and-scaling"><i class="fa fa-check"></i><b>5.4.2</b> Filtering data and scaling</a></li>
<li class="chapter" data-level="5.4.3" data-path="data-preprocessing.html"><a href="data-preprocessing.html#dealing-with-missing-values"><i class="fa fa-check"></i><b>5.4.3</b> Dealing with missing values</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="splitting-the-data.html"><a href="splitting-the-data.html"><i class="fa fa-check"></i><b>5.5</b> Splitting the data</a><ul>
<li class="chapter" data-level="5.5.1" data-path="splitting-the-data.html"><a href="splitting-the-data.html#holdout-test-dataset"><i class="fa fa-check"></i><b>5.5.1</b> Holdout test dataset</a></li>
<li class="chapter" data-level="5.5.2" data-path="splitting-the-data.html"><a href="splitting-the-data.html#cross-validation"><i class="fa fa-check"></i><b>5.5.2</b> Cross-validation</a></li>
<li class="chapter" data-level="5.5.3" data-path="splitting-the-data.html"><a href="splitting-the-data.html#bootstrap-resampling"><i class="fa fa-check"></i><b>5.5.3</b> Bootstrap resampling</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="predicting-the-subtype-with-k-nearest-neighbors.html"><a href="predicting-the-subtype-with-k-nearest-neighbors.html"><i class="fa fa-check"></i><b>5.6</b> Predicting the subtype with k-nearest neighbors</a></li>
<li class="chapter" data-level="5.7" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html"><i class="fa fa-check"></i><b>5.7</b> Assessing the performance of our model</a><ul>
<li class="chapter" data-level="5.7.1" data-path="assessing-the-performance-of-our-model.html"><a href="assessing-the-performance-of-our-model.html#receiver-operating-characteristic-roc-curves"><i class="fa fa-check"></i><b>5.7.1</b> Receiver Operating Characteristic (ROC) Curves</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html"><i class="fa fa-check"></i><b>5.8</b> Model tuning and avoiding overfitting</a><ul>
<li class="chapter" data-level="5.8.1" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#model-complexity-and-bias-variance-trade-off"><i class="fa fa-check"></i><b>5.8.1</b> Model complexity and bias variance trade-off</a></li>
<li class="chapter" data-level="5.8.2" data-path="model-tuning-and-avoiding-overfitting.html"><a href="model-tuning-and-avoiding-overfitting.html#data-split-strategies-for-model-tuning-and-testing"><i class="fa fa-check"></i><b>5.8.2</b> Data split strategies for model tuning and testing</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="variable-importance.html"><a href="variable-importance.html"><i class="fa fa-check"></i><b>5.9</b> Variable importance</a></li>
<li class="chapter" data-level="5.10" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html"><i class="fa fa-check"></i><b>5.10</b> How to deal with class imbalance</a><ul>
<li class="chapter" data-level="5.10.1" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#sampling-for-class-balance"><i class="fa fa-check"></i><b>5.10.1</b> Sampling for class balance</a></li>
<li class="chapter" data-level="5.10.2" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#altering-case-weights"><i class="fa fa-check"></i><b>5.10.2</b> Altering case weights</a></li>
<li class="chapter" data-level="5.10.3" data-path="how-to-deal-with-class-imbalance.html"><a href="how-to-deal-with-class-imbalance.html#selecting-different-classification-score-cutoffs"><i class="fa fa-check"></i><b>5.10.3</b> selecting different classification score cutoffs</a></li>
</ul></li>
<li class="chapter" data-level="5.11" data-path="dealing-with-correlated-predictors.html"><a href="dealing-with-correlated-predictors.html"><i class="fa fa-check"></i><b>5.11</b> Dealing with correlated predictors</a></li>
<li class="chapter" data-level="5.12" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html"><i class="fa fa-check"></i><b>5.12</b> Trees and forests: Random forests in action</a><ul>
<li class="chapter" data-level="5.12.1" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#decision-trees"><i class="fa fa-check"></i><b>5.12.1</b> decision trees</a></li>
<li class="chapter" data-level="5.12.2" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#trees-to-forests"><i class="fa fa-check"></i><b>5.12.2</b> Trees to forests</a></li>
<li class="chapter" data-level="5.12.3" data-path="trees-and-forests-random-forests-in-action.html"><a href="trees-and-forests-random-forests-in-action.html#variable-importance-1"><i class="fa fa-check"></i><b>5.12.3</b> Variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html"><i class="fa fa-check"></i><b>5.13</b> Logistic regression and regularization</a><ul>
<li class="chapter" data-level="5.13.1" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#regularization-in-order-to-avoid-overfitting"><i class="fa fa-check"></i><b>5.13.1</b> regularization in order to avoid overfitting</a></li>
<li class="chapter" data-level="5.13.2" data-path="logistic-regression-and-regularization.html"><a href="logistic-regression-and-regularization.html#variable-importance-2"><i class="fa fa-check"></i><b>5.13.2</b> variable importance</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html"><i class="fa fa-check"></i><b>5.14</b> Other supervised algorithms</a><ul>
<li class="chapter" data-level="5.14.1" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#gradient-boosting"><i class="fa fa-check"></i><b>5.14.1</b> Gradient boosting</a></li>
<li class="chapter" data-level="5.14.2" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#support-vector-machines-svm"><i class="fa fa-check"></i><b>5.14.2</b> Support Vector Machines (SVM)</a></li>
<li class="chapter" data-level="5.14.3" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#neural-networks-and-deep-versions-of-it"><i class="fa fa-check"></i><b>5.14.3</b> Neural networks and deep versions of it</a></li>
<li class="chapter" data-level="5.14.4" data-path="other-supervised-algorithms.html"><a href="other-supervised-algorithms.html#ensemble-learning"><i class="fa fa-check"></i><b>5.14.4</b> Ensemble learning</a></li>
</ul></li>
<li class="chapter" data-level="5.15" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html"><i class="fa fa-check"></i><b>5.15</b> Predicting continuous variables: regression with machine learning</a><ul>
<li class="chapter" data-level="5.15.1" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#use-case-predicting-age-from-dna-methylation"><i class="fa fa-check"></i><b>5.15.1</b> Use case: Predicting age from DNA methylation</a></li>
<li class="chapter" data-level="5.15.2" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#reading-and-processing-the-data"><i class="fa fa-check"></i><b>5.15.2</b> reading and processing the data</a></li>
<li class="chapter" data-level="5.15.3" data-path="predicting-continuous-variables-regression-with-machine-learning.html"><a href="predicting-continuous-variables-regression-with-machine-learning.html#running-random-forest-regression"><i class="fa fa-check"></i><b>5.15.3</b> Running random forest regression</a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="exercises-3.html"><a href="exercises-3.html"><i class="fa fa-check"></i><b>5.16</b> Exercises</a><ul>
<li class="chapter" data-level="5.16.1" data-path="exercises-3.html"><a href="exercises-3.html#classification"><i class="fa fa-check"></i><b>5.16.1</b> classification</a></li>
<li class="chapter" data-level="5.16.2" data-path="exercises-3.html"><a href="exercises-3.html#regression"><i class="fa fa-check"></i><b>5.16.2</b> Regression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="genomicIntervals.html"><a href="genomicIntervals.html"><i class="fa fa-check"></i><b>6</b> Operations on Genomic Intervals and Genome Arithmetic</a><ul>
<li class="chapter" data-level="6.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html"><i class="fa fa-check"></i><b>6.1</b> Operations on Genomic Intervals with GenomicRanges package</a><ul>
<li class="chapter" data-level="6.1.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>6.1.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="6.1.2" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>6.1.2</b> Getting genomic regions into R as GRanges objects</a></li>
<li class="chapter" data-level="6.1.3" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>6.1.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html"><i class="fa fa-check"></i><b>6.2</b> Dealing with mapped high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="6.2.1" data-path="dealing-with-mapped-high-throughput-sequencing-reads.html"><a href="dealing-with-mapped-high-throughput-sequencing-reads.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>6.2.1</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html"><i class="fa fa-check"></i><b>6.3</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="6.3.1" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>6.3.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html"><i class="fa fa-check"></i><b>6.4</b> Genomic intervals with more information: SummarizedExperiment class</a><ul>
<li class="chapter" data-level="6.4.1" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#create-a-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.1</b> Create a SummarizedExperiment object</a></li>
<li class="chapter" data-level="6.4.2" data-path="genomic-intervals-with-more-information-summarizedexperiment-class.html"><a href="genomic-intervals-with-more-information-summarizedexperiment-class.html#subset-and-manipulate-the-summarizedexperiment-object"><i class="fa fa-check"></i><b>6.4.2</b> Subset and manipulate the SummarizedExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html"><i class="fa fa-check"></i><b>6.5</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="6.5.1" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#visualizing-intervals-on-a-locus-of-interest"><i class="fa fa-check"></i><b>6.5.1</b> Visualizing intervals on a locus of interest</a></li>
<li class="chapter" data-level="6.5.2" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#summaries-of-genomic-intervals-on-multiple-loci"><i class="fa fa-check"></i><b>6.5.2</b> Summaries of genomic intervals on multiple loci</a></li>
<li class="chapter" data-level="6.5.3" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>6.5.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="exercises-4.html"><a href="exercises-4.html"><i class="fa fa-check"></i><b>6.6</b> Exercises</a><ul>
<li class="chapter" data-level="6.6.1" data-path="exercises-4.html"><a href="exercises-4.html#operations-on-genomic-intervals-with-genomicranges-package-1"><i class="fa fa-check"></i><b>6.6.1</b> Operations on Genomic Intervals with GenomicRanges package</a></li>
<li class="chapter" data-level="6.6.2" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-mapped-high-throughput-sequencing-reads-1"><i class="fa fa-check"></i><b>6.6.2</b> Dealing with mapped high-throughput sequencing reads</a></li>
<li class="chapter" data-level="6.6.3" data-path="exercises-4.html"><a href="exercises-4.html#dealing-with-contiguous-scores-over-the-genome"><i class="fa fa-check"></i><b>6.6.3</b> Dealing with contiguous scores over the genome</a></li>
<li class="chapter" data-level="6.6.4" data-path="exercises-4.html"><a href="exercises-4.html#visualizing-and-summarizing-genomic-intervals-1"><i class="fa fa-check"></i><b>6.6.4</b> Visualizing and summarizing genomic intervals</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="processingReads.html"><a href="processingReads.html"><i class="fa fa-check"></i><b>7</b> Quality Check, Processing and Alignment of High-throughput Sequencing Reads</a><ul>
<li class="chapter" data-level="7.1" data-path="fasta-and-fastq-formats.html"><a href="fasta-and-fastq-formats.html"><i class="fa fa-check"></i><b>7.1</b> FASTA and FASTQ formats</a></li>
<li class="chapter" data-level="7.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html"><i class="fa fa-check"></i><b>7.2</b> Quality check on sequencing reads</a><ul>
<li class="chapter" data-level="7.2.1" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-quality-per-basecycle"><i class="fa fa-check"></i><b>7.2.1</b> Sequence quality per base/cycle</a></li>
<li class="chapter" data-level="7.2.2" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#sequence-content-per-basecycle"><i class="fa fa-check"></i><b>7.2.2</b> Sequence content per base/cycle</a></li>
<li class="chapter" data-level="7.2.3" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#read-frequency-plot"><i class="fa fa-check"></i><b>7.2.3</b> Read frequency plot</a></li>
<li class="chapter" data-level="7.2.4" data-path="quality-check-on-sequencing-reads.html"><a href="quality-check-on-sequencing-reads.html#other-quality-metrics-and-qc-tools"><i class="fa fa-check"></i><b>7.2.4</b> Other quality metrics and QC tools</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="filtering-and-trimming-reads.html"><a href="filtering-and-trimming-reads.html"><i class="fa fa-check"></i><b>7.3</b> Filtering and trimming reads</a></li>
<li class="chapter" data-level="7.4" data-path="mappingaligning-reads-to-the-genome.html"><a href="mappingaligning-reads-to-the-genome.html"><i class="fa fa-check"></i><b>7.4</b> Mapping/aligning reads to the genome</a></li>
<li class="chapter" data-level="7.5" data-path="further-processing-of-aligned-reads.html"><a href="further-processing-of-aligned-reads.html"><i class="fa fa-check"></i><b>7.5</b> Further processing of aligned reads</a></li>
<li class="chapter" data-level="7.6" data-path="exercises-5.html"><a href="exercises-5.html"><i class="fa fa-check"></i><b>7.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="rnaseqanalysis.html"><a href="rnaseqanalysis.html"><i class="fa fa-check"></i><b>8</b> RNA-seq Analysis</a><ul>
<li class="chapter" data-level="8.1" data-path="what-is-gene-expression.html"><a href="what-is-gene-expression.html"><i class="fa fa-check"></i><b>8.1</b> What is gene expression?</a></li>
<li class="chapter" data-level="8.2" data-path="methods-to-detect-gene-expression.html"><a href="methods-to-detect-gene-expression.html"><i class="fa fa-check"></i><b>8.2</b> Methods to detect gene expression</a></li>
<li class="chapter" data-level="8.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><i class="fa fa-check"></i><b>8.3</b> Gene Expression Analysis Using High-throughput Sequencing Technologies</a><ul>
<li class="chapter" data-level="8.3.1" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#processing-raw-data"><i class="fa fa-check"></i><b>8.3.1</b> Processing raw data</a></li>
<li class="chapter" data-level="8.3.2" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#alignment"><i class="fa fa-check"></i><b>8.3.2</b> Alignment</a></li>
<li class="chapter" data-level="8.3.3" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#quantification"><i class="fa fa-check"></i><b>8.3.3</b> Quantification</a></li>
<li class="chapter" data-level="8.3.4" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#within-sample-normalization-of-the-read-counts"><i class="fa fa-check"></i><b>8.3.4</b> Within sample normalization of the read counts</a></li>
<li class="chapter" data-level="8.3.5" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#computing-different-normalization-schemes-in-r"><i class="fa fa-check"></i><b>8.3.5</b> Computing different normalization schemes in R</a></li>
<li class="chapter" data-level="8.3.6" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#exploratory-analysis-of-the-read-count-table"><i class="fa fa-check"></i><b>8.3.6</b> Exploratory analysis of the read count table</a></li>
<li class="chapter" data-level="8.3.7" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#differential-expression-analysis"><i class="fa fa-check"></i><b>8.3.7</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.3.8" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#functional-enrichment-analysis"><i class="fa fa-check"></i><b>8.3.8</b> Functional Enrichment Analysis</a></li>
<li class="chapter" data-level="8.3.9" data-path="gene-expression-analysis-using-high-throughput-sequencing-technologies.html"><a href="gene-expression-analysis-using-high-throughput-sequencing-technologies.html#accounting-for-additional-sources-of-variation"><i class="fa fa-check"></i><b>8.3.9</b> Accounting for additional sources of variation</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="other-applications-of-rna-seq.html"><a href="other-applications-of-rna-seq.html"><i class="fa fa-check"></i><b>8.4</b> Other applications of RNA-seq</a></li>
<li class="chapter" data-level="8.5" data-path="exercises-6.html"><a href="exercises-6.html"><i class="fa fa-check"></i><b>8.5</b> Exercises</a><ul>
<li class="chapter" data-level="8.5.1" data-path="exercises-6.html"><a href="exercises-6.html#exploring-the-count-tables"><i class="fa fa-check"></i><b>8.5.1</b> Exploring the count tables</a></li>
<li class="chapter" data-level="8.5.2" data-path="exercises-6.html"><a href="exercises-6.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>8.5.2</b> Differential expression analysis</a></li>
<li class="chapter" data-level="8.5.3" data-path="exercises-6.html"><a href="exercises-6.html#functional-enrichment-analysis-1"><i class="fa fa-check"></i><b>8.5.3</b> Functional enrichment analysis</a></li>
<li class="chapter" data-level="8.5.4" data-path="exercises-6.html"><a href="exercises-6.html#removing-unwanted-variation-from-the-expression-data"><i class="fa fa-check"></i><b>8.5.4</b> Removing unwanted variation from the expression data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="chipseq.html"><a href="chipseq.html"><i class="fa fa-check"></i><b>9</b> ChIP-seq analysis</a><ul>
<li class="chapter" data-level="9.1" data-path="regulatory-protein-dna-interactions.html"><a href="regulatory-protein-dna-interactions.html"><i class="fa fa-check"></i><b>9.1</b> Regulatory Protein-DNA Interactions</a></li>
<li class="chapter" data-level="9.2" data-path="measuring-protein-dna-interactions-with-chip-seq.html"><a href="measuring-protein-dna-interactions-with-chip-seq.html"><i class="fa fa-check"></i><b>9.2</b> Measuring Protein-DNA interactions with ChIP-seq</a></li>
<li class="chapter" data-level="9.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><i class="fa fa-check"></i><b>9.3</b> Factors that affect ChIP-seq experiment and analysis quality</a><ul>
<li class="chapter" data-level="9.3.1" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#antibody-specificity"><i class="fa fa-check"></i><b>9.3.1</b> Antibody specificity</a></li>
<li class="chapter" data-level="9.3.2" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#sequencing-depth"><i class="fa fa-check"></i><b>9.3.2</b> Sequencing depth</a></li>
<li class="chapter" data-level="9.3.3" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#pcr-duplication"><i class="fa fa-check"></i><b>9.3.3</b> PCR duplication</a></li>
<li class="chapter" data-level="9.3.4" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#biological-replicates"><i class="fa fa-check"></i><b>9.3.4</b> Biological replicates</a></li>
<li class="chapter" data-level="9.3.5" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#control-experiments"><i class="fa fa-check"></i><b>9.3.5</b> Control experiments</a></li>
<li class="chapter" data-level="9.3.6" data-path="factors-that-affect-chip-seq-experiment-and-analysis-quality.html"><a href="factors-that-affect-chip-seq-experiment-and-analysis-quality.html#using-tagged-proteins"><i class="fa fa-check"></i><b>9.3.6</b> Using tagged proteins</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html"><i class="fa fa-check"></i><b>9.4</b> Pre-processing ChIP data</a><ul>
<li class="chapter" data-level="9.4.1" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html#read-quality-control"><i class="fa fa-check"></i><b>9.4.1</b> Read quality control</a></li>
<li class="chapter" data-level="9.4.2" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html#adapter-and-quality-trimming"><i class="fa fa-check"></i><b>9.4.2</b> Adapter and Quality trimming</a></li>
<li class="chapter" data-level="9.4.3" data-path="pre-processing-chip-data.html"><a href="pre-processing-chip-data.html#mapping"><i class="fa fa-check"></i><b>9.4.3</b> Mapping</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="computational-analysis-of-chip-seq-data.html"><a href="computational-analysis-of-chip-seq-data.html"><i class="fa fa-check"></i><b>9.5</b> Computational Analysis of ChIP-seq data</a><ul>
<li class="chapter" data-level="9.5.1" data-path="computational-analysis-of-chip-seq-data.html"><a href="computational-analysis-of-chip-seq-data.html#prerequisites"><i class="fa fa-check"></i><b>9.5.1</b> Prerequisites</a></li>
<li class="chapter" data-level="9.5.2" data-path="computational-analysis-of-chip-seq-data.html"><a href="computational-analysis-of-chip-seq-data.html#the-data"><i class="fa fa-check"></i><b>9.5.2</b> The Data</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="chip-quality-control.html"><a href="chip-quality-control.html"><i class="fa fa-check"></i><b>9.6</b> ChIP Quality Control</a><ul>
<li class="chapter" data-level="9.6.1" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sample-clustering"><i class="fa fa-check"></i><b>9.6.1</b> Sample Clustering</a></li>
<li class="chapter" data-level="9.6.2" data-path="chip-quality-control.html"><a href="chip-quality-control.html#visualization-the-genome-browser"><i class="fa fa-check"></i><b>9.6.2</b> Visualization the Genome Browser</a></li>
<li class="chapter" data-level="9.6.3" data-path="chip-quality-control.html"><a href="chip-quality-control.html#plus-and-minus-strand-cross-correlation"><i class="fa fa-check"></i><b>9.6.3</b> Plus and minus strand cross-correlation</a></li>
<li class="chapter" data-level="9.6.4" data-path="chip-quality-control.html"><a href="chip-quality-control.html#gc-bias-quantification"><i class="fa fa-check"></i><b>9.6.4</b> GC bias quantification</a></li>
<li class="chapter" data-level="9.6.5" data-path="chip-quality-control.html"><a href="chip-quality-control.html#sequence-read-genomic-distribution"><i class="fa fa-check"></i><b>9.6.5</b> Sequence read genomic distribution</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="peak-calling.html"><a href="peak-calling.html"><i class="fa fa-check"></i><b>9.7</b> Peak Calling</a><ul>
<li class="chapter" data-level="9.7.1" data-path="peak-calling.html"><a href="peak-calling.html#types-of-chip-experiments"><i class="fa fa-check"></i><b>9.7.1</b> Types of ChIP experiments</a></li>
<li class="chapter" data-level="9.7.2" data-path="peak-calling.html"><a href="peak-calling.html#peak-calling---sharp-experiments"><i class="fa fa-check"></i><b>9.7.2</b> Peak Calling - Sharp Experiments</a></li>
<li class="chapter" data-level="9.7.3" data-path="peak-calling.html"><a href="peak-calling.html#broad-regions"><i class="fa fa-check"></i><b>9.7.3</b> Broad regions</a></li>
<li class="chapter" data-level="9.7.4" data-path="peak-calling.html"><a href="peak-calling.html#peak-quality-control"><i class="fa fa-check"></i><b>9.7.4</b> Peak quality control</a></li>
<li class="chapter" data-level="9.7.5" data-path="peak-calling.html"><a href="peak-calling.html#peak-annotation"><i class="fa fa-check"></i><b>9.7.5</b> Peak Annotation</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="motif-discovery.html"><a href="motif-discovery.html"><i class="fa fa-check"></i><b>9.8</b> Motif Discovery</a><ul>
<li class="chapter" data-level="9.8.1" data-path="motif-discovery.html"><a href="motif-discovery.html#motif-annotation"><i class="fa fa-check"></i><b>9.8.1</b> Motif annotation</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="what-to-do-next.html"><a href="what-to-do-next.html"><i class="fa fa-check"></i><b>9.9</b> What to do next?</a></li>
<li class="chapter" data-level="9.10" data-path="exercises-7.html"><a href="exercises-7.html"><i class="fa fa-check"></i><b>9.10</b> Exercises:</a><ul>
<li class="chapter" data-level="9.10.1" data-path="exercises-7.html"><a href="exercises-7.html#basic"><i class="fa fa-check"></i><b>9.10.1</b> Basic</a></li>
<li class="chapter" data-level="9.10.2" data-path="exercises-7.html"><a href="exercises-7.html#intermediate"><i class="fa fa-check"></i><b>9.10.2</b> Intermediate</a></li>
<li class="chapter" data-level="9.10.3" data-path="exercises-7.html"><a href="exercises-7.html#advanced"><i class="fa fa-check"></i><b>9.10.3</b> Advanced</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="bsseq.html"><a href="bsseq.html"><i class="fa fa-check"></i><b>10</b> DNA methylation analysis using bisulfite sequencing data</a><ul>
<li class="chapter" data-level="10.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html"><i class="fa fa-check"></i><b>10.1</b> What is DNA methylation ?</a><ul>
<li class="chapter" data-level="10.1.1" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-dna-methylation-is-set"><i class="fa fa-check"></i><b>10.1.1</b> How DNA methylation is set ?</a></li>
<li class="chapter" data-level="10.1.2" data-path="what-is-dna-methylation.html"><a href="what-is-dna-methylation.html#how-to-measure-dna-methylation-with-bisulfite-sequencing"><i class="fa fa-check"></i><b>10.1.2</b> How to measure DNA methylation with bisulfite-sequencing</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="analyzing-dna-methylation-data.html"><a href="analyzing-dna-methylation-data.html"><i class="fa fa-check"></i><b>10.2</b> Analyzing DNA methylation data</a></li>
<li class="chapter" data-level="10.3" data-path="processing-raw-data-and-getting-data-into-r.html"><a href="processing-raw-data-and-getting-data-into-r.html"><i class="fa fa-check"></i><b>10.3</b> Processing raw data and getting data into R</a></li>
<li class="chapter" data-level="10.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html"><i class="fa fa-check"></i><b>10.4</b> Data filtering and exploratory analysis</a><ul>
<li class="chapter" data-level="10.4.1" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#reading-methylation-call-files"><i class="fa fa-check"></i><b>10.4.1</b> Reading methylation call files</a></li>
<li class="chapter" data-level="10.4.2" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#further-quality-check"><i class="fa fa-check"></i><b>10.4.2</b> Further quality check</a></li>
<li class="chapter" data-level="10.4.3" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#merging-samples-into-a-single-table"><i class="fa fa-check"></i><b>10.4.3</b> Merging samples into a single table</a></li>
<li class="chapter" data-level="10.4.4" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#filtering-cpgs"><i class="fa fa-check"></i><b>10.4.4</b> Filtering CpGs</a></li>
<li class="chapter" data-level="10.4.5" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#clustering-samples"><i class="fa fa-check"></i><b>10.4.5</b> Clustering samples</a></li>
<li class="chapter" data-level="10.4.6" data-path="data-filtering-and-exploratory-analysis.html"><a href="data-filtering-and-exploratory-analysis.html#principal-component-analysis-1"><i class="fa fa-check"></i><b>10.4.6</b> Principal component analysis</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="extracting-interesting-regions-segmentation-and-differential-methylation.html"><a href="extracting-interesting-regions-segmentation-and-differential-methylation.html"><i class="fa fa-check"></i><b>10.5</b> Extracting interesting regions: segmentation and differential methylation</a><ul>
<li class="chapter" data-level="10.5.1" data-path="extracting-interesting-regions-segmentation-and-differential-methylation.html"><a href="extracting-interesting-regions-segmentation-and-differential-methylation.html#differential-methylation"><i class="fa fa-check"></i><b>10.5.1</b> Differential methylation</a></li>
<li class="chapter" data-level="10.5.2" data-path="extracting-interesting-regions-segmentation-and-differential-methylation.html"><a href="extracting-interesting-regions-segmentation-and-differential-methylation.html#methylation-segmentation"><i class="fa fa-check"></i><b>10.5.2</b> Methylation segmentation</a></li>
<li class="chapter" data-level="10.5.3" data-path="extracting-interesting-regions-segmentation-and-differential-methylation.html"><a href="extracting-interesting-regions-segmentation-and-differential-methylation.html#working-with-large-files"><i class="fa fa-check"></i><b>10.5.3</b> Working with large files</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="annotation-of-dmrsdmcs-and-segments.html"><a href="annotation-of-dmrsdmcs-and-segments.html"><i class="fa fa-check"></i><b>10.6</b> Annotation of DMRs/DMCs and segments</a></li>
<li class="chapter" data-level="10.7" data-path="other-r-packages-that-can-be-used-for-methylation-analysis.html"><a href="other-r-packages-that-can-be-used-for-methylation-analysis.html"><i class="fa fa-check"></i><b>10.7</b> Other R packages that can be used for methylation analysis</a></li>
<li class="chapter" data-level="10.8" data-path="exercises-8.html"><a href="exercises-8.html"><i class="fa fa-check"></i><b>10.8</b> Exercises</a><ul>
<li class="chapter" data-level="10.8.1" data-path="exercises-8.html"><a href="exercises-8.html#exercise-1"><i class="fa fa-check"></i><b>10.8.1</b> Exercise 1</a></li>
<li class="chapter" data-level="10.8.2" data-path="exercises-8.html"><a href="exercises-8.html#exercise-2"><i class="fa fa-check"></i><b>10.8.2</b> Exercise 2</a></li>
</ul></li>
<li class="chapter" data-level="10.9" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>10.9</b> References</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiomics.html"><a href="multiomics.html"><i class="fa fa-check"></i><b>11</b> Multi-omics Analysis</a><ul>
<li class="chapter" data-level="11.0.1" data-path="multiomics.html"><a href="multiomics.html#use-case-multi-omics-data-from-colorectal-cancer"><i class="fa fa-check"></i><b>11.0.1</b> Use case: Multi-omics data from colorectal cancer</a></li>
<li class="chapter" data-level="11.1" data-path="latent-variable-models-for-multi-omics-integration.html"><a href="latent-variable-models-for-multi-omics-integration.html"><i class="fa fa-check"></i><b>11.1</b> Latent variable models for multi-omics integration</a></li>
<li class="chapter" data-level="11.2" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><i class="fa fa-check"></i><b>11.2</b> Matrix factorization methods for unsupervised multi-omics data integration</a><ul>
<li class="chapter" data-level="11.2.1" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#multiple-factor-analysis"><i class="fa fa-check"></i><b>11.2.1</b> Multiple Factor Analysis</a></li>
<li class="chapter" data-level="11.2.2" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#joint-non-negative-matrix-factorization"><i class="fa fa-check"></i><b>11.2.2</b> Joint Non-negative Matrix Factorization</a></li>
<li class="chapter" data-level="11.2.3" data-path="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html"><a href="matrix-factorization-methods-for-unsupervised-multi-omics-data-integration.html#icluster"><i class="fa fa-check"></i><b>11.2.3</b> iCluster</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html"><i class="fa fa-check"></i><b>11.3</b> Clustering using latent factors</a><ul>
<li class="chapter" data-level="11.3.1" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#one-hot-clustering"><i class="fa fa-check"></i><b>11.3.1</b> One-hot clustering</a></li>
<li class="chapter" data-level="11.3.2" data-path="clustering-using-latent-factors.html"><a href="clustering-using-latent-factors.html#k-means-clustering-1"><i class="fa fa-check"></i><b>11.3.2</b> K-means clustering</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html"><i class="fa fa-check"></i><b>11.4</b> Biological interpretation of latent factors</a><ul>
<li class="chapter" data-level="11.4.1" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#inspection-of-feature-weights-in-loading-vectors"><i class="fa fa-check"></i><b>11.4.1</b> Inspection of feature weights in loading vectors</a></li>
<li class="chapter" data-level="11.4.2" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#making-sense-of-factors-using-enrichment-analysis"><i class="fa fa-check"></i><b>11.4.2</b> Making sense of factors using enrichment analysis</a></li>
<li class="chapter" data-level="11.4.3" data-path="biological-interpretation-of-latent-factors.html"><a href="biological-interpretation-of-latent-factors.html#interpretation-using-additional-covariates"><i class="fa fa-check"></i><b>11.4.3</b> Interpretation using additional covariates</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="exercises-9.html"><a href="exercises-9.html"><i class="fa fa-check"></i><b>11.5</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="_blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Genomics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="chip-quality-control" class="section level2">
<h2><span class="header-section-number">9.6</span> ChIP Quality Control</h2>
<p>While the goal of the read quality assessment is to check whether the sequencing
produced high enough number of high quality reads;
the goal of the ChIP quality control to ascertaining whether the chromatin immunoprecipitation
enrichment was successful.
This is a crucial step in the ChIP-seq analysis because it can help us
identify low quality ChIP samples, and give information about which experimental
steps went wrong.</p>
<p>There are four steps in the ChIP quality control:</p>
<ol style="list-style-type: decimal">
<li><p>Sample correlation clustering - clustering of the pair-wise correlations between
genome-wide signal profiles.</p></li>
<li><p>Data visualization in a genomic browser.</p></li>
<li><p>Average fragment length determination - determining whether the ChIP enriched
for fragments of a certain length.</p></li>
<li><p>Visualization of GC bias. Here we will plot the ChIP enrichment Vs the
average GC content in the corresponding genomic bin.</p></li>
</ol>
<div id="sample-clustering" class="section level3">
<h3><span class="header-section-number">9.6.1</span> Sample Clustering</h3>
<p>Clustering is an ordering procedure which groups samples by similarity -
the more similar samples are grouped closer to one another.
Clustering of ChIP signal profiles is used for two purposes:
The first one is to ascertain whether there is concordance between
biological replicates - biological replicates should show greater similarity
than ChIP of different proteins.
The second function is to see whether our experiments conform to known prior knowledge.
For example, we would expect to see proteins greater similarity between proteins
which belong to the same quaternary complex.</p>
<p>To quantify the ChIP signal we will firstly construct 1 kilobase wide tilling
windows over the genome, and subsequently count the number of reads
in each window, for each experiment. We will then normalize the counts, to
account for different total number of reads in each experiment, and finally
calculate the correlation between all pairs of samples.
Although this procedure represents a crude way of data quantification, it provides sufficient
information to ascertain the data quality.</p>
<p>Using the <strong>GenomeInfoDb</strong> we will first fetch the chromosome lengths corresponding
to the hg38 version of the human genome, and filter the length for human
chromosome 21.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the chromosome info package</span>
<span class="kw">library</span>(GenomeInfoDb)

<span class="co"># fetch the chromosome lengths for the human genome</span>
hg_chrs =<span class="st"> </span><span class="kw">fetchExtendedChromInfoFromUCSC</span>(<span class="st">&#39;hg38&#39;</span>)

<span class="co"># find the length of chromosome 21</span>
hg_chrs =<span class="st"> </span><span class="kw">subset</span>(hg_chrs, <span class="kw">grepl</span>(<span class="st">&#39;chr21$&#39;</span>,UCSC_seqlevel))</code></pre>
<p><strong>tileGenome</strong> function from the <strong>GenomicRanges</strong> package constructs equally sized
windows over the genome of interest.
The function takes two arguments:</p>
<ol style="list-style-type: decimal">
<li><p>A vector of chromosome lengths</p></li>
<li><p>Window size</p></li>
</ol>
<p>Firstly, we convert the chromosome lengths <em>data.frame</em> into a <em>named vector</em></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># downloaded hg_chrs is a data.frame object, we need to convert the data.frame</span>
<span class="co"># into a named vector</span>
seqlengths =<span class="st"> </span><span class="kw">with</span>(hg_chrs, <span class="kw">setNames</span>(UCSC_seqlength, UCSC_seqlevel))</code></pre>
<p>Then we construct the windows.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the genomic ranges package</span>
<span class="kw">library</span>(GenomicRanges)

<span class="co"># tileGenome function returns a list of GRanges of a given width, spanning</span>
<span class="co"># the whole chromosome</span>
tilling_window =<span class="st"> </span><span class="kw">tileGenome</span>(seqlengths, <span class="dt">tilewidth=</span><span class="dv">1000</span>)

<span class="co"># unlist converts the list to one GRanges object</span>
tilling_window =<span class="st"> </span><span class="kw">unlist</span>(tilling_window)</code></pre>
<pre><code>## GRanges object with 46710 ranges and 0 metadata columns:
##           seqnames            ranges strand
##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##       [1]    chr21            1-1000      *
##       [2]    chr21         1001-2000      *
##       [3]    chr21         2001-3000      *
##       [4]    chr21         3001-4000      *
##       [5]    chr21         4001-5000      *
##       ...      ...               ...    ...
##   [46706]    chr21 46704985-46705984      *
##   [46707]    chr21 46705985-46706984      *
##   [46708]    chr21 46706985-46707984      *
##   [46709]    chr21 46707985-46708984      *
##   [46710]    chr21 46708985-46709983      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome</code></pre>
<p>We will use the <strong>summarizeOverlaps</strong> function from the <strong>GenomicAlignments</strong> package
to count the number of reads in each genomic window.
The function will do the counting automatically for all our experiments.
<strong>summarizeOverlaps</strong> function returns a <strong>SummarizedExperiment</strong> object.
The object contains the counts, genomic ranges which were used for the quantification,
and the sample descriptions.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load GenomicAlignments</span>
<span class="kw">library</span>(GenomicAlignments)

<span class="co"># fetch bam files from the data folder</span>
bam_files =<span class="st"> </span><span class="kw">list.files</span>(data_path, <span class="dt">full.names=</span><span class="ot">TRUE</span>, <span class="dt">pattern=</span><span class="st">&#39;bam$&#39;</span>)

<span class="co"># use summarizeOverlaps to count the reads</span>
so =<span class="st"> </span><span class="kw">summarizeOverlaps</span>(tilling_window, bam_files)

<span class="co"># extract the counts from the SummarizedExperiment</span>
counts =<span class="st"> </span><span class="kw">assays</span>(so)[[<span class="dv">1</span>]]</code></pre>
<p>Different ChIP experiments were sequenced to different depth - each experiment
contains different number of reads. To remove the effect of the experimental
depth on the quantification, the samples need to be normalized.
Standard normalization procedure, for ChIP data, is to divide the
counts in each tilling window by the total number of sequenced reads, and
multiply it with a constant factor (to avoid extremely small numbers).
This normalization procedure is called the <strong>cpm</strong> - counts per million.</p>
<p><span class="math display">\[
CPM = counts * (10^{6} / total\&gt;number\&gt;of\&gt;reads)
\]</span></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate the cpm from the counts matrix</span>
<span class="co"># the following command works because R calculates everything by columns</span>
cpm =<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(counts)<span class="op">*</span>(<span class="dv">1000000</span><span class="op">/</span><span class="kw">colSums</span>(counts)))</code></pre>
<p>We remove all tiles which do not have overlapping reads.
Tiles with 0 counts do not provide any additional discriminatory power, rather,
they introduce artificial similarity between the samples (i.e. samples with
a only a handful of bound regions will have a lot of tiles with 0 counts, while
they do not have to have any overlapping enriched tiles).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># remove all tiles which do not contain reads</span>
cpm =<span class="st"> </span>cpm[<span class="kw">rowSums</span>(cpm) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>,]</code></pre>
<p>We use the sub function to shorten the column names of the cpm matrix.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># change the formatting of the column names</span>
<span class="co"># remove the .chr21.bam suffix</span>
<span class="kw">colnames</span>(cpm) =<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;.chr21.bam&#39;</span>,<span class="st">&#39;&#39;</span>,   <span class="kw">colnames</span>(cpm))

<span class="co"># remove the GM12878_hg38 prefix</span>
<span class="kw">colnames</span>(cpm) =<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;GM12878_hg38_&#39;</span>,<span class="st">&#39;&#39;</span>,<span class="kw">colnames</span>(cpm))</code></pre>
<p>Finally we calculate the pairwise pearson correlation coefficient using the
<strong>cor</strong> function.
The function takes as input an region X sample count matrix, and returns
a sample X sample matrix, where each field contains the correlation coefficient
between two samples.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculates the pearson correlation coefficient between the samples</span>
correlation_matrix =<span class="st"> </span><span class="kw">cor</span>(cpm, <span class="dt">method=</span><span class="st">&#39;pearson&#39;</span>)</code></pre>
<p>The <em>Heatmap</em> function from the <strong>ComplexHeatmap</strong> <span class="citation">(Z. Gu, Eils, and Schlesner <a href="#ref-Gu_2016">2016</a><a href="#ref-Gu_2016">b</a>)</span> package is used to visualize
the correlation coefficient.
The function automatically performs hierarchical clustering - it groups the
samples which have the highest pairwise correlation.
The diagonal represents the correlation of each sample with itself.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load ComplexHeatmap</span>
<span class="kw">library</span>(ComplexHeatmap)

<span class="co"># load the circlize package, and define the color palette which will be</span>
<span class="co"># used in the heatmap</span>
<span class="kw">library</span>(circlize)
heatmap_col =<span class="st"> </span>circlize<span class="op">::</span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">1</span>),<span class="kw">c</span>(<span class="st">&#39;blue&#39;</span>,<span class="st">&#39;white&#39;</span>,<span class="st">&#39;red&#39;</span>))

<span class="co"># plot the heatmap using the Heatmap function</span>
<span class="kw">Heatmap</span>(correlation_matrix, 
        <span class="dt">col =</span> heatmap_col)</code></pre>
<div class="figure" style="text-align: center"><span id="fig:sample-clustering-complex-heatmap"></span>
<img src="compgenomrReloaded_files/figure-html/sample-clustering-complex-heatmap-1.png" alt="Heatmap showing ChIP-seq sample similarity using Pearson correlation coefficient" width="672" />
<p class="caption">
FIGURE 9.2: Heatmap showing ChIP-seq sample similarity using Pearson correlation coefficient
</p>
</div>
<p>Here we can see a perfect example of why quality control is important.
<strong>CTCF</strong> is a zinc finger protein which co-localizes with the Cohesin complex.
<strong>SMC3</strong> is a sub unit of the Cohesin complex, and we would therefore expect to
see that the <strong>SMC3</strong> signal profile has high correlation with the <strong>CTCF</strong> signal profile.
This is true for the second biological replicate of <strong>SMC3</strong>, while the first
replicate (SMC3_r1) clusters with the input samples. This indicates that the
sample likely has low enrichment.
We can see that the ChIP and Input samples form separate cluster. This implies
that the ChIP samples have an enrichment of fragments.
Additionally, we see that the biological replicates of other experiments
cluster together.</p>
</div>
<div id="visualization-the-genome-browser" class="section level3">
<h3><span class="header-section-number">9.6.2</span> Visualization the Genome Browser</h3>
<p>One of the first steps in any ChIP-seq analysis should be looking at the
data. By looking at the data we get an intuition about the quality of the
experiment, and start seeing preliminary correlations between the samples, which
we can use to guide our analysis.
This can be achieved either by plotting signal profiles around
regions of interest, or by loading data into a genome browser
(such as IGV, or UCSC genome browsers).</p>
<p>Genome browsers are standalone applications which represent the genome
as a one dimensional (1D) coordinate systems. The browsers enable
simultaneous visualization and comparison of multiple types of annotations and experimental data.</p>
<p>Genome browsers can visualize most of the commonly used bioinformatics data formats:
bam, bed, wig, bigWig …
The easiest way to access our data would be to load the .bam files into the browser.
This will show us the sequence, and position of every mapped read. If we want to
view multiple samples in parallel, loading every mapped read can be restrictive -
it takes up a lot of computational resources, and the amount of information
makes the visual comparison hard to do.
We would like to convert our data so that we get a compressed visualization,
which would show us the main properties of our samples, namely, the quality, and
the location of the enrichment.
This is achieved by summarizing the read enrichment into a signal profile -
the whole experiment is converted into a numeric vector - a coverage vector.
The vector contains the information of how many reads overlap each position
in the genome.</p>
<p>We will proceed as follows:
Firstly we will import a <strong>.bam</strong> file into <strong>R</strong>. Then we will calculate
the signal profile (construct the coverage vector), and finally, we export the
vector as a <strong>.bigWig</strong> file.</p>
<p>First we select one of the ChIP samples.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># list the bam files in the directory</span>
<span class="co"># The &#39;$&#39; sign tells the pattern recognizer to omit bam.bai files</span>
bam_files =<span class="st"> </span><span class="kw">list.files</span>(data_path, <span class="dt">full.names=</span><span class="ot">TRUE</span>, <span class="dt">pattern=</span><span class="st">&#39;bam$&#39;</span>)

<span class="co"># select the first bam file</span>
chip_file =<span class="st"> </span>bam_files[<span class="dv">1</span>]</code></pre>
<p>We will use the <strong>readGAlignemnts</strong> function from the <strong>GenomicAlignemnts</strong>
package to load the reads into <strong>R</strong>, and then the <strong>granges</strong> function
to convert them into a <strong>GRanges</strong> object.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the genomic alignments package</span>
<span class="kw">library</span>(GenomicAlignments)

<span class="co"># read the ChIP reads into R</span>
reads =<span class="st"> </span><span class="kw">readGAlignments</span>(chip_file)

<span class="co"># the reads need to be converted to a granges object</span>
reads =<span class="st"> </span><span class="kw">granges</span>(reads)</code></pre>
<p>Because DNA fragments are being sequenced from their ends (both the 3’ and 5’ end),
the read enrichment does not correspond to the exact location of the bound protein.
Rather, reads end to form clusters of enrichment upstream and downstream of the true binding location.
To correct for this we use a small hack - before we create the signal profiles,
we will extend the reads towards their <strong>3’</strong> end. The reads are extended to
form fragments of 200 base pairs. This is an empiric measure, which
corresponds to the average fragment size of the Illumina sample preparation kit.
The exact average fragment size will differ from 200 base pairs, but if the
deviation is not large (i.e. more than 200 base pairs),
it will not affect the visual properties of our samples.</p>
<p>Read extension is done using the <strong>resize</strong> function. The function
takes two arguments:</p>
<ol style="list-style-type: decimal">
<li><p>width - resulting fragment width</p></li>
<li><p>fix - which position of the fragment should not be changed (if fix is set to start,
the reads will be extended towards the <strong>3’</strong> end, if fix is set to end, they will
be extended towards the <strong>5’</strong> end)</p></li>
</ol>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extends the reads towards the 3&#39; end</span>
reads =<span class="st"> </span><span class="kw">resize</span>(reads, <span class="dt">width=</span><span class="dv">200</span>, <span class="dt">fix=</span><span class="st">&#39;start&#39;</span>)

<span class="co"># keeps only chromosome 21</span>
reads =<span class="st"> </span><span class="kw">keepSeqlevels</span>(reads, <span class="st">&#39;chr21&#39;</span>, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>)</code></pre>
<p>Conversion of reads into coverage vectors is done with the <strong>coverage</strong>
function.
The function takes only one argument (<strong>width</strong>), which corresponds to chromosome sizes.
For this purpose we can use the, previously created, <strong>seqlengths</strong> variable.
The <strong>coverage</strong> function converts the reads into a compressed <strong>Rle</strong> object</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># convert the reads into a signal profile</span>
cov =<span class="st"> </span><span class="kw">coverage</span>(reads, <span class="dt">width =</span> seqlengths)</code></pre>
<pre><code>## RleList of length 1
## $chr21
## integer-Rle of length 46709983 with 199419 runs
##   Lengths: 5038228     200   63546 ...     200   27856
##   Values :       0       1       0 ...       1       0</code></pre>
<p>The name of the output file is created by changing the file suffix from <em>.bam</em> to <em>.bigWig</em></p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># change the file extension from .bam to .bigWig</span>
output_file =<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;.bam&#39;</span>,<span class="st">&#39;.bigWig&#39;</span>, chip_file)</code></pre>
<p>Now we can use the <strong>export.bw</strong> function from the rtracklayer package to
write the bigWig file.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the rtracklayer package</span>
<span class="kw">library</span>(rtracklayer)

<span class="co"># export the bigWig output file</span>
<span class="kw">export.bw</span>(cov, <span class="st">&#39;output_file&#39;</span>)</code></pre>
<div id="gviz" class="section level4">
<h4><span class="header-section-number">9.6.2.1</span> Gviz</h4>
<p>We can create Genome browser like visualizations using the <strong>Gviz</strong> package.
Gviz is a too which enables exhaustive customized visualization of
genomics experiments.
The basic usage principle is to define tracks, where each track can represent
genomic annotation, or a signal profile; subsequently we define the order
of the tracks and plot them.</p>
<p>Here we will define two tracks, a genome axis, which will show the position
along the human chromosome 21; and a signal track from our CTCF experiment.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Gviz)
<span class="co"># define the genome axis track</span>
axis   =<span class="st"> </span><span class="kw">GenomeAxisTrack</span>(<span class="kw">GRanges</span>(<span class="st">&#39;chr21&#39;</span>, <span class="kw">IRanges</span>(<span class="dv">1</span>, <span class="dt">width=</span>seqlengths)))

<span class="co"># convert the signal into genomic ranges and define the signal track</span>
gcov   =<span class="st"> </span><span class="kw">as</span>(cov, <span class="st">&#39;GRanges&#39;</span>)
dtrack =<span class="st"> </span><span class="kw">DataTrack</span>(gcov, <span class="dt">name =</span> <span class="st">&quot;CTCF&quot;</span>, <span class="dt">type=</span><span class="st">&#39;l&#39;</span>)

<span class="co"># define the track ordering</span>
track_list =<span class="st"> </span><span class="kw">list</span>(axis,dtrack)</code></pre>
<p>Tracks are plotted with the <strong>plotTracks</strong> function.
<strong>sizes</strong> argument needs to be the same size as the track_list, and defines the
relative size of each track.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot the list of browser tracks</span>
<span class="co"># sizes argument defines the relative sizes of tracks</span>
<span class="co"># background title defines the color for the track labels</span>
<span class="kw">plotTracks</span>(track_list, <span class="dt">sizes=</span><span class="kw">c</span>(.<span class="dv">1</span>,<span class="dv">1</span>), <span class="dt">background.title =</span> <span class="st">&quot;black&quot;</span>)</code></pre>
<div class="figure" style="text-align: center"><span id="fig:genome-browser-gviz-show"></span>
<img src="compgenomrReloaded_files/figure-html/genome-browser-gviz-show-1.png" alt="ChIP-seq signal visualized as a browser track using Gviz" width="672" />
<p class="caption">
FIGURE 9.3: ChIP-seq signal visualized as a browser track using Gviz
</p>
</div>
</div>
</div>
<div id="plus-and-minus-strand-cross-correlation" class="section level3">
<h3><span class="header-section-number">9.6.3</span> Plus and minus strand cross-correlation</h3>
<p>Cross-correlation between plus and minus strands is a method
which quantifies whether the DNA library was enriched for fragments of
a certain length.</p>
<p>Similarity between the plus and minus strands defined as the correlation of
the signal profiles for the reads that map to the <strong>+</strong> and the <strong>-</strong> strands.
The distribution of reads is shown on @ref(fig:Figure_BrowserScreenshot)</p>
<div class="figure" style="text-align: center"><span id="fig:Figure-BrowserScreenshot"></span>
<img src="Figures/BrowserScreenshot.png" alt="Browser screenshot of aligned reads for one ChIP, and control sample. ChIP samples have an assymetric distribution of reads - reads mapping to the + strand are located on the left side of the peak, while the reads mapping to the - strand are found on the right side of the peak" width="960" />
<p class="caption">
FIGURE 9.4: Browser screenshot of aligned reads for one ChIP, and control sample. ChIP samples have an assymetric distribution of reads - reads mapping to the + strand are located on the left side of the peak, while the reads mapping to the - strand are found on the right side of the peak
</p>
</div>
<p>Due to the sequencing properties, reads which correspond to
the <strong>5’</strong> fragment ends will map to the opposite strand then the reads
coming from the <strong>3’</strong> ends. Most often (depending on the sequencing protocol)
the reads from the <strong>5’</strong> fragment ends map to the <strong>+</strong> strand,
while the reads from the <strong>3’</strong> ends map to the <strong>-</strong> strand.</p>
<p>We calculate the cross-correlation, by shifting the signal on the <strong>+</strong> strand,
by a pre-defined amount (i.e. shift by 1 - 400 nucleotides), and calculating,
for each shift, the correlation between the <strong>+</strong>, and the <strong>-</strong> strands.
Subsequently we plot the correlation Vs shift, and locate the maximum value.
The maximum value should correspond to the average DNA fragment length which
was present in the library. This value tells us whether the ChIP enriched for
fragments of certain length (i.e. whether the ChIP was successful).</p>
<p>Due to the size of genomic data, it might be computationally prohibitive to
calculate the Pearson correlation between whole genome (or even whole chromosome)
signal profiles.
To get around this problem, we will resort to a trick - we will disregard the dynamic
range of the signal profiles, and only keep the information of which
genomic bases contained the ends of the fragments.
This is done by calculating the coverage vector of read starting position (separately
for each strand), and converting the coverage vector into a boolean vector.
The boolean vector contains the information of which genomic positions
contained the DNA fragment ends.</p>
<p>Similarity between two boolean vectors can be promptly computed using the Jaccard index.
Jaccard index is defined as an intersection between two boolean vectors, divided by their union
(fig:Figure_JaccardSimilarity).</p>
<div class="figure" style="text-align: center"><span id="fig:Figure-JaccardSimilarity"></span>
<img src="Figures/Jaccard.png" alt="Jaccard similarity is defined as the ration of the intersection and union of two sets." width="289" />
<p class="caption">
FIGURE 9.5: Jaccard similarity is defined as the ration of the intersection and union of two sets.
</p>
</div>
<p>Firstly we load the reads for one of the CTCF ChIP experiments.
Then we create signal profiles, separately for reads on the <strong>+</strong>, and <strong>-</strong>
strands.
Unlike before, we do not extend the reads to the average expected fragment
length (200 base pairs); we keep only the starting position of each read.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the reads</span>
reads =<span class="st"> </span><span class="kw">readGAlignments</span>(chip_file)
reads =<span class="st"> </span><span class="kw">granges</span>(reads)

<span class="co"># keep only the starting position of each read</span>
reads =<span class="st"> </span><span class="kw">resize</span>(reads, <span class="dt">width=</span><span class="dv">1</span>, <span class="dt">fix=</span><span class="st">&#39;start&#39;</span>)

reads =<span class="st"> </span><span class="kw">keepSeqlevels</span>(reads, <span class="st">&#39;chr21&#39;</span>, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>)</code></pre>
<p>Now we can calculate the coverage vector of read starting position.
The coverage vector is then automatically converted into a boolean vector by
asking which genomic positions have coverage &gt; 0.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate the coverage profile for plus and minus strand</span>
reads =<span class="st"> </span><span class="kw">split</span>(reads, <span class="kw">strand</span>(reads))

<span class="co"># coverage(x, width = seqlengths)[[1]] &gt; 0 calculates the coverage and</span>
<span class="co"># converts the coverage vector into a boolean</span>
cov   =<span class="st"> </span><span class="kw">lapply</span>(reads, <span class="cf">function</span>(x)<span class="kw">coverage</span>(x, <span class="dt">width =</span> seqlengths)[[<span class="dv">1</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)
cov   =<span class="st"> </span><span class="kw">lapply</span>(cov, as.vector)</code></pre>
<p>We will no shift the coverage vector from the plus strand by 1 - 400 base pairs,
and for each pair shift we will calculate the Jaccard index between the vectors
on the plus and minus strand.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># defines the shift range</span>
wsize =<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">400</span>

<span class="co"># defines the jaccard similarity</span>
jaccard =<span class="st"> </span><span class="cf">function</span>(x,y)<span class="kw">sum</span>((x <span class="op">&amp;</span><span class="st"> </span>y)) <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>((x <span class="op">|</span><span class="st"> </span>y))

<span class="co"># shifts the + vector by 1 - 400 nucleotides and calculates the correlation coefficient</span>
cc =<span class="st"> </span><span class="kw">shiftApply</span>(wsize, cov[[<span class="st">&#39;+&#39;</span>]], cov[[<span class="st">&#39;-&#39;</span>]], <span class="dt">FUN=</span>jaccard)

<span class="co"># converts the results into a data frame</span>
cc =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">fragment_size =</span> wsize, <span class="dt">cross_correlation =</span> cc)</code></pre>
<p>We can finally plot the shift Vs the correlation coefficient:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">ggplot</span>(<span class="dt">data =</span> cc, <span class="kw">aes</span>(fragment_size, cross_correlation)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw">which.max</span>(cc<span class="op">$</span>cross_correlation), 
               <span class="dt">size=</span><span class="dv">2</span>, <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">linetype=</span><span class="dv">2</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(
        <span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),
        <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),
        <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&#39;Shift in base pairs&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="st">&#39;Jaccard similarity&#39;</span>) </code></pre>
<div class="figure" style="text-align: center"><span id="fig:correlation-plot"></span>
<img src="compgenomrReloaded_files/figure-html/correlation-plot-1.png" alt="The figure shows the correlation coefficient between the ChIP-seq signal on + and - strands. The peak of the distribution designates the fragment size" width="672" />
<p class="caption">
FIGURE 9.6: The figure shows the correlation coefficient between the ChIP-seq signal on + and - strands. The peak of the distribution designates the fragment size
</p>
</div>
<p>The shift in base pairs which corresponds to the maximum value of the correlation coefficient
gives us an approximation to the expected average DNA fragment length.
Because this value is not 0, or monotonically decreasing, we can conclude
that there was substantial enrichment of certain fragments in the ChIP samples.</p>
</div>
<div id="gc-bias-quantification" class="section level3">
<h3><span class="header-section-number">9.6.4</span> GC bias quantification</h3>
<p>The PCR amplification procedure can cause a significant bias in the ChIP
experiments. The bias can be influenced by the DNA fragment size distribution,
sequence composition, hexamer distribution of PCR primers, and the number of cycles used
for the amplification.
One way how to determine whether some of the samples have significantly
different sequence composition is to look at whether regions with
differing GC composition were equally enriched in all experiments.</p>
<p>We will do the following: Firstly we will calculate the GC content of each
of the tilling windows, and then we will compare the GC content with the corresponding
cpm (count per million reads) value, for each tile.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># fetches the chromosome lengths and constructs the tiles</span>
<span class="kw">library</span>(GenomeInfoDb)
<span class="kw">library</span>(GenomicRanges)

hg_chrs        =<span class="st"> </span><span class="kw">fetchExtendedChromInfoFromUCSC</span>(<span class="st">&#39;hg38&#39;</span>)
hg_chrs        =<span class="st"> </span><span class="kw">subset</span>(hg_chrs, <span class="kw">grepl</span>(<span class="st">&#39;chr21$&#39;</span>,UCSC_seqlevel))
seqlengths     =<span class="st"> </span><span class="kw">with</span>(hg_chrs, <span class="kw">setNames</span>(UCSC_seqlength, UCSC_seqlevel))
tilling_window =<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">tileGenome</span>(seqlengths, <span class="dt">tilewidth=</span><span class="dv">1000</span>))</code></pre>
<p>We will extract the sequence information from the <strong>BSgenome.Hsapiens.UCSC.hg38</strong>
package. <strong>BSgenome</strong> are generic Bioconductor containers for genomic sequences.
Sequences are extracted from the <strong>BSgenome</strong> container using the <strong>getSeq</strong> function.
<strong>getSeq</strong> function takes as input the genome object, and the ranges with the
regions of interest - in our case, the tilling windows.
The function returns a <strong>DNAString</strong> object.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># loads the human genome sequence</span>
<span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg38)

<span class="co"># extracts the sequence from the human genome</span>
seq =<span class="st"> </span><span class="kw">getSeq</span>(BSgenome.Hsapiens.UCSC.hg38, tilling_window)</code></pre>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq
## [1]  1000 TTACCTTGTGCCATACCTTCG...ACAGGGAGTTTCGCAATGCTT
## [2]  1000 CCTCATACAATGTCTGGAATC...GGGAGAGACTTTGAGGATATA
## [3]  1000 TTATTACGTGAAATTTGCCAA...AAATAGCCGGGCGTGGTGGCG
## [4]  1000 GGCGCCTGTAGTCCCAGCTAC...TTAGGGTTTTGTTGGGTTATC
## [5]  1000 TTCATGAAGATATAAATTAAT...TATATTAAAATATGGCATCTG
## [6]  1000 ATTGCCTCTTCTGAAATTAGG...AAAAATATCAATTATTTTACC</code></pre>
<p>To calculate the GC content, we will use the <strong>oligonucleotideFrequency</strong> on the
<strong>DNAString</strong> object. By setting the width parameter to 2 we will
calculate the <strong>dinucleotide</strong> frequency.
Each row in the resulting table will contain the number of all possible
dinucleotides observed in each tilling window.
Because we have tilling windows of same length we do not
necessarily need to normalize the counts by the window length.
If all of the windows do not have the same length (i.e. when looking
at the ChIP-seq peaks), then the normalization is a prerequisite.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculates the frequency of all possible dimers in our sequence set</span>
nuc =<span class="st"> </span><span class="kw">oligonucleotideFrequency</span>(seq, <span class="dt">width =</span> <span class="dv">2</span>)

<span class="co"># converts the matrix into a data.frame</span>
nuc =<span class="st"> </span><span class="kw">as.data.frame</span>(nuc)

<span class="co"># calculates the percentages, and rounds the number</span>
nuc =<span class="st"> </span><span class="kw">round</span>(nuc<span class="op">/</span><span class="dv">1000</span>,<span class="dv">3</span>)</code></pre>
<pre><code>##          AA    AC    AG    AT    CA    CC    CG    CT
## 20000 0.085 0.070 0.075 0.044 0.097 0.078 0.027 0.071
## 20001 0.096 0.049 0.052 0.123 0.063 0.026 0.001 0.059
## 20002 0.160 0.046 0.050 0.110 0.046 0.030 0.017 0.057
## 20003 0.103 0.045 0.065 0.074 0.058 0.046 0.011 0.070
## 20004 0.106 0.050 0.064 0.117 0.070 0.030 0.002 0.059
## 20005 0.130 0.057 0.068 0.078 0.079 0.053 0.007 0.061
##          GA    GC    GG    GT    TA    TC    TG    TT
## 20000 0.054 0.055 0.072 0.047 0.038 0.070 0.054 0.062
## 20001 0.047 0.021 0.025 0.055 0.115 0.052 0.070 0.145
## 20002 0.053 0.033 0.038 0.038 0.107 0.041 0.058 0.115
## 20003 0.052 0.036 0.045 0.060 0.074 0.059 0.071 0.130
## 20004 0.059 0.025 0.021 0.050 0.102 0.056 0.069 0.119
## 20005 0.056 0.043 0.031 0.045 0.067 0.048 0.069 0.107</code></pre>
<p>Now we can combine the GC frequency with the cpm values.
We will convert the cpm values to the log10 scale. To avoid
taking the log(0), we add a pseudo count of 1 to cpm.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># counts the number of reads per tilling window for each experiment</span>
so =<span class="st"> </span><span class="kw">summarizeOverlaps</span>(tilling_window, bam_files)

<span class="co"># converts the raw counts to cpm values</span>
counts  =<span class="st"> </span><span class="kw">assays</span>(so)[[<span class="dv">1</span>]]
cpm     =<span class="st"> </span><span class="kw">t</span>(<span class="kw">t</span>(counts)<span class="op">*</span>(<span class="dv">1000000</span><span class="op">/</span><span class="kw">colSums</span>(counts)))
cpm_log =<span class="st"> </span><span class="kw">log10</span>(cpm<span class="op">+</span><span class="dv">1</span>)</code></pre>
<p>Combine the cpm values with the GC content, and plot the results.</p>
<pre class="sourceCode r"><code class="sourceCode r">gc   =<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">data.frame</span>(cpm_log), <span class="dt">GC =</span> nuc[<span class="st">&#39;GC&#39;</span>])</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> gc, <span class="kw">aes</span>(GC, GM12878_hg38_CTCF_r1.chr21.bam)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">2</span>, <span class="dt">alpha=</span>.<span class="dv">3</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(
        <span class="dt">axis.text  =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),
        <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),
        <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&#39;GC content in one kilobase windows&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="st">&#39;log10( cpm + 1 )&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&#39;CTCF Replicate 1&#39;</span>)</code></pre>
<div class="figure" style="text-align: center"><span id="fig:gc-plot"></span>
<img src="compgenomrReloaded_files/figure-html/gc-plot-1.png" alt="GC content abundance in a ChIP-seq experiment" width="672" />
<p class="caption">
FIGURE 9.7: GC content abundance in a ChIP-seq experiment
</p>
</div>
<p>CPM Vs GC content plot gives us two important pieces of information.
The first one is whether there was a specific amplification of regions
with extremely high or extremely low GC content. This would be strong indication
that the either the PCR or the size selection procedure were not successfully
executed.
The second piece of information comes by comparison of plots
corresponding to multiple experiments. If different ChIP-samples have
highly diverging enrichment of different ChIP regions, this implies that
some of the samples were affected by unknown batch affects. Such effects
need to be taken into account in downstream analysis.</p>
<p>Firstly, we will reorder the columns of the data.frame using the gather
function from the tidyr package</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the tidyr package</span>
<span class="kw">library</span>(tidyr)

<span class="co"># gather converts a fat data.frame into a tall data.frame, which is</span>
<span class="co"># the format used by the ggplot package</span>
gcd =<span class="st"> </span><span class="kw">gather</span>(<span class="dt">data =</span> gc, experiment, cpm, <span class="op">-</span>GC)

<span class="co"># we select the ChIP files corresponding to the ctcf experiment</span>
gcd =<span class="st"> </span><span class="kw">subset</span>(gcd, <span class="kw">grepl</span>(<span class="st">&#39;CTCF&#39;</span>, experiment))

<span class="co"># remove the chr21 suffix</span>
gcd<span class="op">$</span>experiment =<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;chr21.&#39;</span>,<span class="st">&#39;&#39;</span>,gcd<span class="op">$</span>experiment)     

<span class="kw">ggplot</span>(<span class="dt">data =</span> gcd, <span class="kw">aes</span>(GC, <span class="kw">log10</span>(cpm<span class="op">+</span><span class="dv">1</span>))) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">2</span>, <span class="dt">alpha=</span>.<span class="dv">05</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>experiment, <span class="dt">nrow=</span><span class="dv">1</span>)<span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(
        <span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),
        <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),
        <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&#39;GC content in one kilobase windows&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="st">&#39;log10( cpm + 1 )&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&#39;CTCF Replicates 1 and 2&#39;</span>)</code></pre>
<p><img src="compgenomrReloaded_files/figure-html/gc.tidy-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="sequence-read-genomic-distribution" class="section level3">
<h3><span class="header-section-number">9.6.5</span> Sequence read genomic distribution</h3>
<p>The fourth way to look at the ChIP quality control is to visualize
the genomic distribution of reads in different functional genomic regions.
If the ChIP samples have the same distribution of reads as the Input samples,
this implies a lack of specific enrichment. Additionally, if we have
prior knowledge of where our proteins should be located, we can use
the visualization to judge how well the genomic distributions conform to our priors.
For example, the trimethylation of histone H3 on lysine 36 - H3K36me3 is associated
with elongating polymerase and productive transcription. If we performed a
successful ChIP experiment with an anti-H3K36me3 antibody, we would expect most of the reads
to fall within gene bodies (introns and exons).</p>
<div id="hierarchical-annotation-of-genomic-features" class="section level4">
<h4><span class="header-section-number">9.6.5.1</span> Hierarchical annotation of genomic features</h4>
<p>Overlapping genomic features (a transcription start site of one
gene might be in an intron of another gene) will cause an ambiguity during
the read annotation. If a read overlaps more than one functional category, we are not
certain which category it should be assigned to.
To solve the problem of multiple assignments, we need to construct a set of annotation rules.
A heuristic solution is to organize the genomic annotation into a
hierarchy which will imply prioritization.
We can then look, for each read, which functional categories it overlaps, and
if it within multiple categories, we assign the read to the topmost category.
As an example, let’s say that we have 4 genomic categories:</p>
<p>TSS (transcription start sites), exon, intron, intergenic, with the following</p>
<p>hierarchy:</p>
<p>TSS -&gt; exon -&gt; intron -&gt; intergenic</p>
<p>If a read overlaps a TSS and an intron, it will be annotates as TSS @ref(fig:Figure_Hierarchical_Annotation).</p>
<div class="figure" style="text-align: center"><span id="fig:Figure-Hierarchical-Annotation"></span>
<img src="Figures/Hierarchical_Annotation.png" alt="Principle of hierarchical annotation. The region of interest is annotated as the topmost ranked category that it overlaps. In this case, our region overlaps a TSS, an exon, and a intergenic regions. Because the TSS has the topmost rank, it is annotated as a TSS." width="454" />
<p class="caption">
FIGURE 9.8: Principle of hierarchical annotation. The region of interest is annotated as the topmost ranked category that it overlaps. In this case, our region overlaps a TSS, an exon, and a intergenic regions. Because the TSS has the topmost rank, it is annotated as a TSS.
</p>
</div>
<p>Now we will construct the set of functional genomic regions, and annotate
the reads.</p>
</div>
<div id="finding-annotations" class="section level4">
<h4><span class="header-section-number">9.6.5.2</span> Finding Annotations</h4>
<p>There are multiple sources of genomic annotation. UCSC, Genbank, and
ENSEMBL databases represent stable resources, from which the annotation
can be easily obtained.</p>
<p><strong>AnnotationHub</strong> is a Bioconductor based online resource which contains a large
amount of experiments from various sources.
We will use the AnnotationHub to download the location of genes corresponding
to the hg38 genome.</p>
<p>The hub is accessed in the following way:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the AnnotationHub package</span>
<span class="kw">library</span>(AnnotationHub)

<span class="co"># connect to the hub object</span>
hub =<span class="st"> </span><span class="kw">AnnotationHub</span>()</code></pre>
<p>The <strong>hub</strong> variable contains the programming interface towards the online database.
We can use the <strong>query</strong> function to find out the ID of the
ENSEMBL gene annotation.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># query the hub for the human annotation</span>
AnnotationHub<span class="op">::</span><span class="kw">query</span>(hub, <span class="kw">c</span>(<span class="st">&#39;ENSEMBL&#39;</span>,<span class="st">&#39;Homo&#39;</span>,<span class="st">&#39;GRCh38&#39;</span>,<span class="st">&#39;chr&#39;</span>,<span class="st">&#39;gtf&#39;</span>))</code></pre>
<pre><code>## AnnotationHub with 22 records
## # snapshotDate(): 2018-10-24 
## # $dataprovider: Ensembl
## # $species: Homo sapiens
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome,
## #   description, coordinate_1_based, maintainer,
## #   rdatadateadded, preparerclass, tags,
## #   rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g.,
## #   &#39;object[[&quot;AH50842&quot;]]&#39; 
## 
##             title                                     
##   AH50842 | Homo_sapiens.GRCh38.84.chr.gtf            
##   AH50843 | Homo_sapiens.GRCh38.84.chr_patch_hapl_s...
##   AH51012 | Homo_sapiens.GRCh38.85.chr.gtf            
##   AH51013 | Homo_sapiens.GRCh38.85.chr_patch_hapl_s...
##   AH51953 | Homo_sapiens.GRCh38.86.chr.gtf            
##   ...       ...                                       
##   AH61127 | Homo_sapiens.GRCh38.92.chr_patch_hapl_s...
##   AH63755 | Homo_sapiens.GRCh38.93.chr.gtf            
##   AH63756 | Homo_sapiens.GRCh38.93.chr_patch_hapl_s...
##   AH64629 | Homo_sapiens.GRCh38.94.chr.gtf            
##   AH64630 | Homo_sapiens.GRCh38.94.chr_patch_hapl_s...</code></pre>
<p>We are interested in the version GRCh38.92, which is available under <strong>AH61126</strong>.
To download the data from the hub, we use the <strong>[[</strong> operator on the
hub API.
We will download the annotation in the <strong>.gtf</strong> format, into a <strong>GRanges</strong> object.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># retrieve the human gene annotation</span>
gtf =<span class="st"> </span>hub[[<span class="st">&#39;AH61126&#39;</span>]]</code></pre>
<pre><code>## GRanges object with 6 ranges and 22 metadata columns:
##       seqnames      ranges strand |   source
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   [1]        1 11869-14409      + |   havana
##   [2]        1 11869-14409      + |   havana
##   [3]        1 11869-12227      + |   havana
##   [4]        1 12613-12721      + |   havana
##   [5]        1 13221-14409      + |   havana
##   [6]        1 12010-13670      + |   havana
##             type     score     phase         gene_id
##         &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;     &lt;character&gt;
##   [1]       gene      &lt;NA&gt;      &lt;NA&gt; ENSG00000223972
##   [2] transcript      &lt;NA&gt;      &lt;NA&gt; ENSG00000223972
##   [3]       exon      &lt;NA&gt;      &lt;NA&gt; ENSG00000223972
##   [4]       exon      &lt;NA&gt;      &lt;NA&gt; ENSG00000223972
##   [5]       exon      &lt;NA&gt;      &lt;NA&gt; ENSG00000223972
##   [6] transcript      &lt;NA&gt;      &lt;NA&gt; ENSG00000223972
##       gene_version   gene_name gene_source
##        &lt;character&gt; &lt;character&gt; &lt;character&gt;
##   [1]            5     DDX11L1      havana
##   [2]            5     DDX11L1      havana
##   [3]            5     DDX11L1      havana
##   [4]            5     DDX11L1      havana
##   [5]            5     DDX11L1      havana
##   [6]            5     DDX11L1      havana
##                             gene_biotype
##                              &lt;character&gt;
##   [1] transcribed_unprocessed_pseudogene
##   [2] transcribed_unprocessed_pseudogene
##   [3] transcribed_unprocessed_pseudogene
##   [4] transcribed_unprocessed_pseudogene
##   [5] transcribed_unprocessed_pseudogene
##   [6] transcribed_unprocessed_pseudogene
##         transcript_id transcript_version
##           &lt;character&gt;        &lt;character&gt;
##   [1]            &lt;NA&gt;               &lt;NA&gt;
##   [2] ENST00000456328                  2
##   [3] ENST00000456328                  2
##   [4] ENST00000456328                  2
##   [5] ENST00000456328                  2
##   [6] ENST00000450305                  2
##       transcript_name transcript_source
##           &lt;character&gt;       &lt;character&gt;
##   [1]            &lt;NA&gt;              &lt;NA&gt;
##   [2]     DDX11L1-202            havana
##   [3]     DDX11L1-202            havana
##   [4]     DDX11L1-202            havana
##   [5]     DDX11L1-202            havana
##   [6]     DDX11L1-201            havana
##                       transcript_biotype         tag
##                              &lt;character&gt; &lt;character&gt;
##   [1]                               &lt;NA&gt;        &lt;NA&gt;
##   [2]               processed_transcript       basic
##   [3]               processed_transcript       basic
##   [4]               processed_transcript       basic
##   [5]               processed_transcript       basic
##   [6] transcribed_unprocessed_pseudogene       basic
##       transcript_support_level exon_number
##                    &lt;character&gt; &lt;character&gt;
##   [1]                     &lt;NA&gt;        &lt;NA&gt;
##   [2]                        1        &lt;NA&gt;
##   [3]                        1           1
##   [4]                        1           2
##   [5]                        1           3
##   [6]                       NA        &lt;NA&gt;
##               exon_id exon_version  protein_id
##           &lt;character&gt;  &lt;character&gt; &lt;character&gt;
##   [1]            &lt;NA&gt;         &lt;NA&gt;        &lt;NA&gt;
##   [2]            &lt;NA&gt;         &lt;NA&gt;        &lt;NA&gt;
##   [3] ENSE00002234944            1        &lt;NA&gt;
##   [4] ENSE00003582793            1        &lt;NA&gt;
##   [5] ENSE00002312635            1        &lt;NA&gt;
##   [6]            &lt;NA&gt;         &lt;NA&gt;        &lt;NA&gt;
##       protein_version     ccds_id
##           &lt;character&gt; &lt;character&gt;
##   [1]            &lt;NA&gt;        &lt;NA&gt;
##   [2]            &lt;NA&gt;        &lt;NA&gt;
##   [3]            &lt;NA&gt;        &lt;NA&gt;
##   [4]            &lt;NA&gt;        &lt;NA&gt;
##   [5]            &lt;NA&gt;        &lt;NA&gt;
##   [6]            &lt;NA&gt;        &lt;NA&gt;
##   -------
##   seqinfo: 25 sequences (1 circular) from GRCh38 genome; no seqlengths</code></pre>
<p>By default the ENSEMBL project labels chromosomes using numeric identifiers (i.e. 1,2,3 … X),
without the <strong>chr</strong> prefix.
We need to therefore append the prefix to the chromosome names (seqlevels).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># paste the chr prefix to chromosome names</span>
<span class="kw">seqlevels</span>(gtf, <span class="dt">pruning.mode=</span><span class="st">&#39;coarse&#39;</span>) =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;chr&#39;</span>, <span class="kw">seqlevels</span>(gtf))</code></pre>
<p>And finally we subset only regions which correspond to chromosome 21.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># keep only chromosome 21</span>
gtf =<span class="st"> </span>gtf[<span class="kw">seqnames</span>(gtf) <span class="op">==</span><span class="st"> &#39;chr21&#39;</span>]</code></pre>
</div>
<div id="constructing-genomic-annotation" class="section level4">
<h4><span class="header-section-number">9.6.5.3</span> Constructing genomic annotation</h4>
<p>Once we have downloaded the annotation we can define the functional hierarchy.
We will use the previously mentioned ordering:</p>
<p>TSS -&gt; exon -&gt; intron -&gt; intergenic</p>
<p>With TSS having the highest priority and the intergenic regions having the lowest priority.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># construct a GRangesList with human annotation</span>
annotation_list =<span class="st"> </span><span class="kw">GRangesList</span>(
    
    <span class="co"># promoters function extends the gtf around the TSS by an upstream and downstream amounts</span>
    <span class="dt">tss    =</span> <span class="kw">promoters</span>(<span class="kw">subset</span>(gtf, type<span class="op">==</span><span class="st">&#39;gene&#39;</span>), <span class="dt">upstream=</span><span class="dv">1000</span>, <span class="dt">downstream=</span><span class="dv">1000</span>),
    <span class="dt">exon   =</span> <span class="kw">subset</span>(gtf, type<span class="op">==</span><span class="st">&#39;exon&#39;</span>),
    <span class="dt">intron =</span> <span class="kw">subset</span>(gtf, type<span class="op">==</span><span class="st">&#39;gene&#39;</span>)
)</code></pre>
</div>
<div id="annotating-reads" class="section level4">
<h4><span class="header-section-number">9.6.5.4</span> Annotating reads</h4>
<p>To annotate the reads we will define a function that takes as input a <strong>.bam</strong> file,
and an annotation list, and returns the frequency of reads in each genomic category.
We will then loop over all of the <strong>.bam</strong> files to annotate each experiment.</p>
<p><strong>annotateReads</strong> function works in the following way:</p>
<ol style="list-style-type: decimal">
<li><p>Load the <strong>.bam</strong> file</p></li>
<li><p>Find overlaps between the reads and the annotation categories</p></li>
<li><p>Arrange the annotated reads based on the hierarchy, and remove duplicated assignments</p></li>
<li><p>Count the number of reads in each category.</p></li>
</ol>
<p>The crucial step to understand here is using the <strong>arrange</strong> and <strong>filter</strong> functions
to keep only one annotated category per read.</p>
<pre class="sourceCode r"><code class="sourceCode r">annotateReads =<span class="st"> </span><span class="cf">function</span>(bam_file, annotation_list){
    
    <span class="kw">library</span>(dplyr)
    <span class="kw">message</span>(<span class="kw">basename</span>(bam_file))
    
    <span class="co"># load the reads into R</span>
    bam    =<span class="st"> </span><span class="kw">readGAlignments</span>(bam_file)
    
    <span class="co"># find overlaps between reads and annotation</span>
    result =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">suppressWarnings</span>(<span class="kw">findOverlaps</span>(bam, annotation_list))) 
    result<span class="op">$</span>annotation =<span class="st"> </span><span class="kw">names</span>(annotation_list)[result<span class="op">$</span>subjectHits]
        
    <span class="co"># order the overlaps based on the hierarchy</span>
    result =<span class="st"> </span>result[<span class="kw">order</span>(result<span class="op">$</span>subjectHits),]
        
    <span class="co"># select only one category per read</span>
    result =<span class="st"> </span><span class="kw">subset</span>(result, <span class="op">!</span><span class="kw">duplicated</span>(queryHits))
        
    <span class="co"># count the number of reads in each category</span>
    <span class="co"># group the result data frame by the corresponding category</span>
    result =<span class="st"> </span><span class="kw">group_by</span>(<span class="dt">.data=</span>result, annotation)
    
    <span class="co"># count the number of reads in each category</span>
    result =<span class="st"> </span><span class="kw">summarise</span>(<span class="dt">.data =</span> result, <span class="dt">counts =</span> <span class="kw">length</span>(annotation))
    
    <span class="co"># classify all reads which are outside of the annotation as intergenic</span>
    result =<span class="st"> </span><span class="kw">rbind</span>(result, <span class="kw">data.frame</span>(
                    <span class="dt">annotation =</span> <span class="st">&#39;intergenic&#39;</span>,
                    <span class="dt">counts     =</span> <span class="kw">length</span>(bam) <span class="op">-</span><span class="st"> </span><span class="kw">sum</span>(result<span class="op">$</span>counts)))
    
    <span class="co"># calculate the frequency</span>
    result<span class="op">$</span>frequency  =<span class="st"> </span><span class="kw">with</span>(result, <span class="kw">round</span>(counts<span class="op">/</span><span class="kw">sum</span>(counts),<span class="dv">2</span>))

    <span class="co"># append the experiment name</span>
    result<span class="op">$</span>experiment =<span class="st"> </span><span class="kw">basename</span>(bam_file)
        
    <span class="kw">return</span>(result)
}</code></pre>
<p>We execute the annotation function on all files.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># list all bam files in the folder</span>
bam_files   =<span class="st"> </span><span class="kw">list.files</span>(data_path, <span class="dt">full.names=</span><span class="ot">TRUE</span>, <span class="dt">pattern=</span><span class="st">&#39;bam$&#39;</span>)

<span class="co"># calculate the read distribution for every file</span>
annot_reads_list =<span class="st"> </span><span class="kw">lapply</span>(bam_files, <span class="cf">function</span>(x)<span class="kw">annotateReads</span>(x, annotation_list))</code></pre>
<p>And plot the results.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># collapse the per-file read distributions into one data.frame</span>
annot_reads_df =<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">bind_rows</span>(annot_reads_list)

<span class="co"># format the experiment names</span>
annot_reads_df<span class="op">$</span>experiment =<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;.chr21.bam&#39;</span>,<span class="st">&#39;&#39;</span>,annot_reads_df<span class="op">$</span>experiment)
annot_reads_df<span class="op">$</span>experiment =<span class="st"> </span><span class="kw">sub</span>(<span class="st">&#39;GM12878_hg38_&#39;</span>,<span class="st">&#39;&#39;</span>,annot_reads_df<span class="op">$</span>experiment)


<span class="kw">ggplot</span>(<span class="dt">data =</span> annot_reads_df, <span class="kw">aes</span>(experiment, frequency, <span class="dt">fill=</span>annotation)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&#39;identity&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_brewer</span>(<span class="dt">palette=</span><span class="st">&#39;Set2&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(
        <span class="dt">axis.text =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>, <span class="dt">face=</span><span class="st">&#39;bold&#39;</span>),
        <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">14</span>,<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>),
        <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),
        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust =</span> <span class="dv">1</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&#39;Sample&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="st">&#39;Percentage of reads&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ggtitle</span>(<span class="st">&#39;Percentage of reads in annotation&#39;</span>)</code></pre>
<div class="figure" style="text-align: center"><span id="fig:read-annot-plot"></span>
<img src="compgenomrReloaded_files/figure-html/read-annot-plot-1.png" alt="ChIP-seq signal distribution over functional genomic features" width="672" />
<p class="caption">
FIGURE 9.9: ChIP-seq signal distribution over functional genomic features
</p>
</div>
<p>We can observe a slight increase of H3K36me3 on the exons and introns, and H3K4me3
on the TSS. Interestingly, both replicates of the ZNF143 transcription factor show increased
read abundance around the TSS.</p>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Gu_2016">
<p>Gu, Zuguang, Roland Eils, and Matthias Schlesner. 2016b. “Complex Heatmaps Reveal Patterns and Correlations in Multidimensional Genomic Data.” <em>Bioinformatics</em>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="computational-analysis-of-chip-seq-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="peak-calling.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/lunr.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/compgenomr/book/edit/master/09-chip-seq-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["compgenomrReloaded.pdf"],
"toc": {
"collapse": "none"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
