<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Computational Genomics with R</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="book files for computatiation genomcis with R">
  <meta name="generator" content="bookdown 0.0.71 and GitBook 2.6.7">

  <meta property="og:title" content="Computational Genomics with R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="book files for computatiation genomcis with R" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Computational Genomics with R" />
  
  <meta name="twitter:description" content="book files for computatiation genomcis with R" />
  

<meta name="author" content="Altuna Akalin">

<meta name="date" content="2016-05-10">

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="statistics-for-genomics.html">


<script src="book_assets/jquery-2.2.3/jquery.min.js"></script>
<link href="book_assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Prerequisites</a></li>
<li class="chapter" data-level="2" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html"><i class="fa fa-check"></i><b>2</b> Introduction to Genomics</a></li>
<li class="chapter" data-level="3" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>3</b> Introduction to R</a><ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#the-setup"><i class="fa fa-check"></i><b>3.1</b> The Setup</a><ul>
<li class="chapter" data-level="3.1.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#installing-packages"><i class="fa fa-check"></i><b>3.1.1</b> Installing packages</a></li>
<li class="chapter" data-level="3.1.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#installing-packages-in-custom-locations"><i class="fa fa-check"></i><b>3.1.2</b> Installing packages in custom locations</a></li>
<li class="chapter" data-level="3.1.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#getting-help-on-functions-and-packages"><i class="fa fa-check"></i><b>3.1.3</b> Getting help on functions and packages</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#computations-in-r"><i class="fa fa-check"></i><b>3.2</b> Computations in R</a></li>
<li class="chapter" data-level="3.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-structures"><i class="fa fa-check"></i><b>3.3</b> Data structures</a><ul>
<li class="chapter" data-level="3.3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>3.3.1</b> Vectors</a></li>
<li class="chapter" data-level="3.3.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#matrices"><i class="fa fa-check"></i><b>3.3.2</b> Matrices</a></li>
<li class="chapter" data-level="3.3.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-frames"><i class="fa fa-check"></i><b>3.3.3</b> Data Frames</a></li>
<li class="chapter" data-level="3.3.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#lists"><i class="fa fa-check"></i><b>3.3.4</b> Lists</a></li>
<li class="chapter" data-level="3.3.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#factors"><i class="fa fa-check"></i><b>3.3.5</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-types"><i class="fa fa-check"></i><b>3.4</b> Data types</a></li>
<li class="chapter" data-level="3.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#reading-and-writing-data"><i class="fa fa-check"></i><b>3.5</b> Reading and writing data</a></li>
<li class="chapter" data-level="3.6" data-path="introduction-to-r.html"><a href="introduction-to-r.html#plotting-in-r"><i class="fa fa-check"></i><b>3.6</b> Plotting in R</a></li>
<li class="chapter" data-level="3.7" data-path="introduction-to-r.html"><a href="introduction-to-r.html#functions-and-control-structures-for-ifelse-etc."><i class="fa fa-check"></i><b>3.7</b> Functions and control structures (for, if/else etc.)</a><ul>
<li class="chapter" data-level="3.7.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#user-defined-functions"><i class="fa fa-check"></i><b>3.7.1</b> User defined functions</a></li>
<li class="chapter" data-level="3.7.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#loops-and-looping-structures-in-r"><i class="fa fa-check"></i><b>3.7.2</b> Loops and looping structures in R</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="introduction-to-r.html"><a href="introduction-to-r.html#session-info"><i class="fa fa-check"></i><b>3.8</b> Session info</a></li>
<li class="chapter" data-level="3.9" data-path="introduction-to-r.html"><a href="introduction-to-r.html#exercises"><i class="fa fa-check"></i><b>3.9</b> Exercises</a><ul>
<li class="chapter" data-level="3.9.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#computations-in-r-1"><i class="fa fa-check"></i><b>3.9.1</b> Computations in R</a></li>
<li class="chapter" data-level="3.9.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-structures-in-r"><i class="fa fa-check"></i><b>3.9.2</b> Data structures in R</a></li>
<li class="chapter" data-level="3.9.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#reading-in-and-writing-data-out-in-r"><i class="fa fa-check"></i><b>3.9.3</b> Reading in and writing data out in R</a></li>
<li class="chapter" data-level="3.9.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#plotting-in-r-1"><i class="fa fa-check"></i><b>3.9.4</b> Plotting in R</a></li>
<li class="chapter" data-level="3.9.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#functions-and-control-structures-for-ifelse-etc.-1"><i class="fa fa-check"></i><b>3.9.5</b> Functions and control structures (for, if/else etc.)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html"><i class="fa fa-check"></i><b>4</b> Statistics for Genomics</a><ul>
<li class="chapter" data-level="4.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions"><i class="fa fa-check"></i><b>4.1</b> How to summarize collection of data points: The idea behind statistical distributions</a><ul>
<li class="chapter" data-level="4.1.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#describing-the-central-tendency-mean-and-median"><i class="fa fa-check"></i><b>4.1.1</b> Describing the central tendency: mean and median</a></li>
<li class="chapter" data-level="4.1.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#describing-the-spread-measurements-of-variation"><i class="fa fa-check"></i><b>4.1.2</b> Describing the spread: measurements of variation</a></li>
<li class="chapter" data-level="4.1.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#precision-of-estimates-confidence-intervals"><i class="fa fa-check"></i><b>4.1.3</b> Precision of estimates: Confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-test-for-differences-in-samples"><i class="fa fa-check"></i><b>4.2</b> How to test for differences in samples</a><ul>
<li class="chapter" data-level="4.2.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#randomization-based-testing-for-difference-of-the-means"><i class="fa fa-check"></i><b>4.2.1</b> randomization based testing for difference of the means</a></li>
<li class="chapter" data-level="4.2.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#using-t-test-for-difference-of-the-means-between-two-samples"><i class="fa fa-check"></i><b>4.2.2</b> Using t-test for difference of the means between two samples</a></li>
<li class="chapter" data-level="4.2.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#multiple-testing-correction"><i class="fa fa-check"></i><b>4.2.3</b> multiple testing correction</a></li>
<li class="chapter" data-level="4.2.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#moderated-t-tests-using-information-from-multiple-comparisons"><i class="fa fa-check"></i><b>4.2.4</b> moderated t-tests: using information from multiple comparisons</a></li>
<li class="chapter" data-level="4.2.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#want-to-know-more"><i class="fa fa-check"></i><b>4.2.5</b> Want to know more…</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#relationship-between-variables-linear-models-and-correlation"><i class="fa fa-check"></i><b>4.3</b> Relationship between variables: linear models and correlation</a><ul>
<li class="chapter" data-level="4.3.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-fit-a-line"><i class="fa fa-check"></i><b>4.3.1</b> How to fit a line</a></li>
<li class="chapter" data-level="4.3.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-estimate-the-error-of-the-coefficients"><i class="fa fa-check"></i><b>4.3.2</b> How to estimate the error of the coefficients</a></li>
<li class="chapter" data-level="4.3.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#accuracy-of-the-model"><i class="fa fa-check"></i><b>4.3.3</b> Accuracy of the model</a></li>
<li class="chapter" data-level="4.3.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#regression-with-categorical-variables"><i class="fa fa-check"></i><b>4.3.4</b> Regression with categorical variables</a></li>
<li class="chapter" data-level="4.3.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#regression-pitfalls"><i class="fa fa-check"></i><b>4.3.5</b> Regression pitfalls</a></li>
<li class="chapter" data-level="4.3.6" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#want-to-know-more-1"><i class="fa fa-check"></i><b>4.3.6</b> Want to know more…</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#roadmap-for-future"><i class="fa fa-check"></i><b>4.4</b> Roadmap for future</a><ul>
<li class="chapter" data-level="4.4.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#clustering"><i class="fa fa-check"></i><b>4.4.1</b> Clustering</a></li>
<li class="chapter" data-level="4.4.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#dimension-reduction-with-pca"><i class="fa fa-check"></i><b>4.4.2</b> Dimension reduction with PCA</a></li>
<li class="chapter" data-level="4.4.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#classification"><i class="fa fa-check"></i><b>4.4.3</b> Classification</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#exercises-1"><i class="fa fa-check"></i><b>4.5</b> Exercises</a><ul>
<li class="chapter" data-level="4.5.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1"><i class="fa fa-check"></i><b>4.5.1</b> How to summarize collection of data points: The idea behind statistical distributions</a></li>
<li class="chapter" data-level="4.5.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-test-for-differences-in-samples-1"><i class="fa fa-check"></i><b>4.5.2</b> How to test for differences in samples</a></li>
<li class="chapter" data-level="4.5.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#relationship-between-variables-linear-models-and-correlation-1"><i class="fa fa-check"></i><b>4.5.3</b> Relationship between variables: linear models and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html"><i class="fa fa-check"></i><b>5</b> Genomic Intervals and R</a><ul>
<li class="chapter" data-level="5.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#operations-on-genomic-intervals-with-genomicranges-package"><i class="fa fa-check"></i><b>5.1</b> Operations on Genomic Intervals with GenomicRanges package</a><ul>
<li class="chapter" data-level="5.1.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>5.1.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="5.1.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>5.1.2</b> Getting genomic regions into R as GRanges objects</a></li>
<li class="chapter" data-level="5.1.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>5.1.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-high-throughput-sequencing-reads"><i class="fa fa-check"></i><b>5.2</b> Dealing with high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="5.2.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#quality-check-on-sequencing-reads-and-mapping-reads-to-the-genome"><i class="fa fa-check"></i><b>5.2.1</b> Quality check on sequencing reads and mapping reads to the genome</a></li>
<li class="chapter" data-level="5.2.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>5.2.2</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-continuous-scores-over-the-genome"><i class="fa fa-check"></i><b>5.3</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="5.3.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>5.3.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-and-summarizing-genomic-intervals"><i class="fa fa-check"></i><b>5.4</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="5.4.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-intervals-in-locus-of-interest"><i class="fa fa-check"></i><b>5.4.1</b> Visualizing intervals in locus of interest</a></li>
<li class="chapter" data-level="5.4.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#summaries-of-genomic-intervals"><i class="fa fa-check"></i><b>5.4.2</b> Summaries of genomic intervals</a></li>
<li class="chapter" data-level="5.4.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>5.4.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#session-info-1"><i class="fa fa-check"></i><b>5.5</b> Session info</a></li>
<li class="chapter" data-level="5.6" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#exercises-2"><i class="fa fa-check"></i><b>5.6</b> Exercises</a><ul>
<li class="chapter" data-level="5.6.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#the-setup-1"><i class="fa fa-check"></i><b>5.6.1</b> The setup</a></li>
<li class="chapter" data-level="5.6.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#operations-on-genomic-intervals-with-genomicranges-package-1"><i class="fa fa-check"></i><b>5.6.2</b> Operations on Genomic Intervals with GenomicRanges package</a></li>
<li class="chapter" data-level="5.6.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-mapped-high-throughput-sequencing-reads"><i class="fa fa-check"></i><b>5.6.3</b> Dealing with mapped high-throughput sequencing reads</a></li>
<li class="chapter" data-level="5.6.4" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-contiguous-scores-over-the-genome"><i class="fa fa-check"></i><b>5.6.4</b> Dealing with contiguous scores over the genome</a></li>
<li class="chapter" data-level="5.6.5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-and-summarizing-genomic-intervals-1"><i class="fa fa-check"></i><b>5.6.5</b> Visualizing and summarizing genomic intervals</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Genomics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="genomic-intervals-and-r" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Genomic Intervals and R</h1>
<p>A considerable time in computational genomics is spent on overlapping different features of the genome. Each feature can be represented with a genomic interval within the chromosomal coordinate system. In addition, each interval can carry different sorts of information. An interval may for instance represent exon coordinates or a transcription factor binding site. On the other hand, you can have base-pair resolution, continuous scores over the genome such as read coverage or scores that could be associated with only certain bases such as in the case of CpG methylation (See Figure 1). Typically, you will need to overlap intervals on interest with other features of the genome, again represented as intervals. For example, you may want to overlap transcription factor binding sites with CpG islands or promoters to quantify what percentage of binding sites overlap with your regions of interest. Overlapping mapped reads from high-throughput sequencing experiments with genomic features such as exons, promoters, enhancers can also be classified as operations on genomic intervals. You can think of a million other ways that involves overlapping two sets of different features on the genome. This chapter aims to show how to do analysis involving operations on genomic intervals.</p>
<div class="figure">
<img src="GenomicIntervals_data/nonR_Figures/genomeIntervalSummary.png" alt="" />

</div>
<p>Figure 1: Summary of genomic intervals with different kinds of information</p>
<div id="operations-on-genomic-intervals-with-genomicranges-package" class="section level2">
<h2><span class="header-section-number">5.1</span> Operations on Genomic Intervals with GenomicRanges package</h2>
<p>The <a href="http://bioconductor.org">Bioconductor</a> project has a dedicated package called <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html"><strong>GenomicRanges</strong></a> to deal with genomic intervals. In this section, we will provide use cases involving operations on genomic intervals. The main reason we will stick to this package is that it provides tools to do overlap operations. However package requires that users operate on specific data types that are conceptually similar to a tabular data structure implemented in a way that makes overlapping and related operations easier. The main object we will be using is called <code>GRanges</code> object and we will also see some other related objects from the <code>GenomicRanges</code> package.</p>
<div id="how-to-create-and-manipulate-a-granges-object" class="section level3">
<h3><span class="header-section-number">5.1.1</span> How to create and manipulate a GRanges object</h3>
<p><code>GRanges</code> (from GenomicRanges package) is the main object that holds the genomic intervals and extra information about those intervals. Here we will show how to create one. Conceptually, it is similar to a data frame and some operations such as using <strong><a href="genomic-intervals-and-r.html#section-107"></a></strong> notation to subset the table will work also on GRanges, but keep in mind that not everything that works for data frames will work on GRanges objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicRanges)
gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>,<span class="st">&quot;chr2&quot;</span>,<span class="st">&quot;chr2&quot;</span>),
           <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">150</span>,<span class="dv">200</span>),<span class="dt">end=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">300</span>)),
           <span class="dt">strand=</span><span class="kw">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>,<span class="st">&quot;-&quot;</span>)
)
gr</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames     ranges strand
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1 [ 50, 100]      +
##   [2]     chr2 [150, 200]      -
##   [3]     chr2 [200, 300]      -
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset like a data frame</span>
gr[<span class="dv">1</span>:<span class="dv">2</span>,]</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames     ranges strand
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1 [ 50, 100]      +
##   [2]     chr2 [150, 200]      -
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>As you can see it looks a bit like a data frame. Also, note that the peculiar second argument “ranges” which basically contains start and end positions of the genomic intervals. However, you can not just give start and end positions you actually have to provide another object of IRanges. Do not let this confuse you, GRanges actually depends on another object that is very similar to itself called <code>IRanges</code> and you have to provide the “ranges” argument as an IRanges object. In its simplest for, an IRanges object can be constructed by providing start and end positions to IRanges() function. Think of it as something you just have to provide in order to construct the GRanges object.</p>
<p>GRanges can also contain other information about the genomic interval such as scores, names, etc. You can provide extra information at the time of the construction or you can add it later. Here is how you can do those:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>,<span class="st">&quot;chr2&quot;</span>,<span class="st">&quot;chr2&quot;</span>),
           <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">150</span>,<span class="dv">200</span>),<span class="dt">end=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">300</span>)),
           <span class="dt">names=</span><span class="kw">c</span>(<span class="st">&quot;id1&quot;</span>,<span class="st">&quot;id3&quot;</span>,<span class="st">&quot;id2&quot;</span>),
           <span class="dt">scores=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">90</span>,<span class="dv">50</span>)
)
<span class="co"># or add it later (replaces the existing meta data)</span>
<span class="kw">mcols</span>(gr)=<span class="kw">DataFrame</span>(<span class="dt">name2=</span><span class="kw">c</span>(<span class="st">&quot;pax6&quot;</span>,<span class="st">&quot;meis1&quot;</span>,<span class="st">&quot;zic4&quot;</span>),
                    <span class="dt">score2=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>))

gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>,<span class="st">&quot;chr2&quot;</span>,<span class="st">&quot;chr2&quot;</span>),
           <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">150</span>,<span class="dv">200</span>),<span class="dt">end=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">300</span>)),
           <span class="dt">names=</span><span class="kw">c</span>(<span class="st">&quot;id1&quot;</span>,<span class="st">&quot;id3&quot;</span>,<span class="st">&quot;id2&quot;</span>),
           <span class="dt">scores=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">90</span>,<span class="dv">50</span>)
)

<span class="co"># or appends to existing meta data</span>
<span class="kw">mcols</span>(gr)=<span class="kw">cbind</span>(<span class="kw">mcols</span>(gr),
                          <span class="kw">DataFrame</span>(<span class="dt">name2=</span><span class="kw">c</span>(<span class="st">&quot;pax6&quot;</span>,<span class="st">&quot;meis1&quot;</span>,<span class="st">&quot;zic4&quot;</span>)) )
gr</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames     ranges strand |       names    scores       name2
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
##   [1]     chr1 [ 50, 100]      * |         id1       100        pax6
##   [2]     chr2 [150, 200]      * |         id3        90       meis1
##   [3]     chr2 [200, 300]      * |         id2        50        zic4
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># elementMetadata() and values() do the same things</span>
<span class="kw">elementMetadata</span>(gr)</code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##         names    scores       name2
##   &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
## 1         id1       100        pax6
## 2         id3        90       meis1
## 3         id2        50        zic4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">values</span>(gr)</code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##         names    scores       name2
##   &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
## 1         id1       100        pax6
## 2         id3        90       meis1
## 3         id2        50        zic4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># you may also add metadata using the $ operator, as for data frames</span>
gr$name3 =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;C&quot;</span>, <span class="st">&quot;B&quot;</span>)
gr</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 4 metadata columns:
##       seqnames     ranges strand |       names    scores       name2
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
##   [1]     chr1 [ 50, 100]      * |         id1       100        pax6
##   [2]     chr2 [150, 200]      * |         id3        90       meis1
##   [3]     chr2 [200, 300]      * |         id2        50        zic4
##             name3
##       &lt;character&gt;
##   [1]           A
##   [2]           C
##   [3]           B
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="getting-genomic-regions-into-r-as-granges-objects" class="section level3">
<h3><span class="header-section-number">5.1.2</span> Getting genomic regions into R as GRanges objects</h3>
<p>There are multiple ways you can read in your genomic features into R and create a GRanges object. Most genomic interval data comes as a tabular format that has the basic information about the location of the interval and some other information. We already showed how to read BED files as data frame. Now we will show how to convert it to GRanges object. This is one way of doing it, but there are more convenient ways described further in the text.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read CpGi data set</span>
cpgi.df =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;GenomicIntervals_data/data/cpgi.hg19.chr21.bed&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>,
                     <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>) 
<span class="co"># remove chr names with &quot;_&quot;</span>
cpgi.df =cpgi.df [<span class="kw">grep</span>(<span class="st">&quot;_&quot;</span>,cpgi.df[,<span class="dv">1</span>],<span class="dt">invert=</span><span class="ot">TRUE</span>),]

cpgi.gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span>cpgi.df[,<span class="dv">1</span>],
                <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span>cpgi.df[,<span class="dv">2</span>],
                              <span class="dt">end=</span>cpgi.df[,<span class="dv">3</span>]))</code></pre></div>
<p>You may need to do some pre-processing before/after reading in the BED file. Below is an example of getting transcription start sites from BED files containing refseq transcript locations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read refseq file</span>
ref.df =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;GenomicIntervals_data/data/refseq.hg19.chr21.bed&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>,
                     <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>) 
ref.gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span>ref.df[,<span class="dv">1</span>],
               <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span>ref.df[,<span class="dv">2</span>],
                              <span class="dt">end=</span>ref.df[,<span class="dv">3</span>]),
               <span class="dt">strand=</span>ref.df[,<span class="dv">6</span>],<span class="dt">name=</span>ref.df[,<span class="dv">4</span>])
<span class="co"># get TSS</span>
tss.gr=ref.gr
<span class="co"># end of the + strand genes must be equalized to start pos</span>
<span class="kw">end</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;+&quot;</span>,])  =<span class="kw">start</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;+&quot;</span>,])
<span class="co"># startof the - strand genes must be equalized to end pos</span>
<span class="kw">start</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;-&quot;</span>,])=<span class="kw">end</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;-&quot;</span>,])
<span class="co"># remove duplicated TSSes ie alternative transcripts</span>
<span class="co"># this keeps the first instance and removes duplicates</span>
tss.gr=tss.gr[!<span class="kw">duplicated</span>(tss.gr),]</code></pre></div>
<p>Another way of doing this is from a BED file is to use <code>readTranscriptfeatures()</code> function from the <code>genomation</code> package.</p>
<p>Reading the genomic features as text files and converting to GRanges is not the only way to create GRanges object. With the help of the <a href="http://www.bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a> package we can directly import BED files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(rtracklayer)
<span class="kw">import.bed</span>(<span class="st">&quot;GenomicIntervals_data/data/refseq.hg19.chr21.bed&quot;</span>)</code></pre></div>
<p>Next, we will show how to use other methods to automatically obtain the data in GRanges format from online databases. But you will not be able to use these methods for every data set so it is good to now how to read data from flat files as well. We will use rtracklayer package to download data from <a href="http://genome-euro.ucsc.edu/cgi-bin/hgGateway?hgsid=197855328_moRAG0Dqe1qlcOEtNDAt09E1e6ab">UCSC browser</a>. We will download CpG islands as GRanges objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(rtracklayer)
session &lt;-<span class="st"> </span><span class="kw">browserSession</span>(<span class="st">&quot;UCSC&quot;</span>,<span class="dt">url =</span> <span class="st">&#39;http://genome-euro.ucsc.edu/cgi-bin/&#39;</span>)
<span class="kw">genome</span>(session) &lt;-<span class="st"> &quot;mm9&quot;</span>
## choose CpG island track on chr12
query &lt;-<span class="st"> </span><span class="kw">ucscTableQuery</span>(session, <span class="dt">track=</span><span class="st">&quot;CpG Islands&quot;</span>,<span class="dt">table=</span><span class="st">&quot;cpgIslandExt&quot;</span>,
        <span class="dt">range=</span><span class="kw">GRangesForUCSCGenome</span>(<span class="st">&quot;mm9&quot;</span>, <span class="st">&quot;chr12&quot;</span>))
## get the GRanges object for the track
<span class="kw">track</span>(query)</code></pre></div>
<div id="frequently-used-file-formats-and-how-to-read-them-into-r-as-a-table" class="section level4">
<h4><span class="header-section-number">5.1.2.1</span> Frequently used file formats and how to read them into R as a table</h4>
<p>There are multiple file formats in genomics but some of them you will see more frequently than others. We already mentioned some of them. Here is a list of files and functions to read them into R as <code>GRanges</code> objects or something coercible to GRanges objects.</p>
<p><strong>BED</strong>: These are used and popularized by UCSC browser, and can hold a variety of information including exon/intron structure of transcripts in a single line.</p>
<ul>
<li><code>genomation::readBed()</code></li>
<li><code>genomation::readTranscriptFeatures()</code> good for getting intron/exon/promoters from BED12 files</li>
<li><code>rtracklayer::import.bed()</code></li>
</ul>
<p><strong>GFF</strong>: GFF format is a tabular text format for genomic features similar to BED. However, it is a more flexible format than BED, which makes it harder to parse at times. Many gene annotation files are in this format.</p>
<ul>
<li><code>genomation::gffToGranges()</code></li>
<li><code>rtracklayer::impot.gff()</code></li>
</ul>
<p><strong>BAM</strong>: BAM format is compressed and indexed tabular file format designed for sequencing reads.</p>
<ul>
<li><code>GenomicAlignments::readGAlignments</code></li>
<li><code>Rsamtools::scanBam</code> returns a data frame with columns from SAM/BAM file.</li>
</ul>
<p><strong>BigWig</strong>: This is used to for storing scores associated with genomic intervals. It is an indexed format. Similar to BAM, this makes it easier to query and only necessary portions of the file could be loaded into memory.</p>
<ul>
<li><code>rtracklayer::import.bw()</code></li>
</ul>
<p><strong>Generic Text files</strong>: This represents any text file with the minimal information of chromosome, start and end coordinates.</p>
<ul>
<li><code>genomation::readGeneric()</code></li>
</ul>
<p><strong>Tabix/Bcf</strong>: These are tabular file formats indexed and compressed similar to BAM. The following functions return lists rather than tabular data structures. These formats are mostly used to store genomic variation data such as SNPs and indels.</p>
<ul>
<li><code>Rsamtools::scanTabix</code></li>
<li><code>Rsamtools::scanBcf</code></li>
</ul>
</div>
</div>
<div id="finding-regions-that-dodo-not-overlap-with-another-set-of-regions" class="section level3">
<h3><span class="header-section-number">5.1.3</span> Finding regions that do/do not overlap with another set of regions</h3>
<p>This is one of the most common tasks in genomics. Usually, you have a set of regions that you are interested in and you want to see if they overlap with another set of regions or see how many of them overlap. A good example is transcription factor binding sites determined by <a href="http://en.wikipedia.org/wiki/ChIP-sequencing">ChIP-seq</a> experiments. In these types of experiments and followed analysis, one usually ends up with genomic regions that are bound by transcription factors. One of the standard next questions would be to annotate binding sites with genomic annotations such as promoter,exon,intron and/or CpG islands. Below is a demonstration of how transcription factor binding sites can be annotated using CpG islands. First, we will get the subset of binding sites that overlap with the CpG islands. In this case, binding sites are ChIP-seq peaks.</p>
<p>We can find the subset of peaks that overlap with the CpG islands using the <strong>subsetByoverlaps()</strong> function. You will also see another way of converting <strong>data frames</strong> to <strong>GRanges</strong>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(genomation)
pk1.gr=<span class="kw">readBroadPeak</span>(<span class="st">&quot;GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz&quot;</span>)

pk1.gr=pk1.gr[<span class="kw">seqnames</span>(pk1.gr)==<span class="st">&quot;chr21&quot;</span>,]
<span class="co"># get the peaks that overlap with CpG</span>
<span class="co"># islands</span>
<span class="kw">subsetByOverlaps</span>(pk1.gr,cpgi.gr)</code></pre></div>
<pre><code>## GRanges object with 44 ranges and 5 metadata columns:
##        seqnames               ranges strand |        name     score
##           &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;
##    [1]    chr21 [ 9825359,  9826582]      * |   peak14562        56
##    [2]    chr21 [ 9968468,  9968984]      * |   peak14593       947
##    [3]    chr21 [15755367, 15755956]      * |   peak14828        90
##    [4]    chr21 [19191578, 19192525]      * |   peak14840       290
##    [5]    chr21 [26979618, 26980048]      * |   peak14854        32
##    ...      ...                  ...    ... .         ...       ...
##   [40]    chr21 [46237463, 46237809]      * |   peak15034        32
##   [41]    chr21 [46707701, 46708084]      * |   peak15037        67
##   [42]    chr21 [46961551, 46961875]      * |   peak15039        38
##   [43]    chr21 [47743586, 47744125]      * |   peak15050       353
##   [44]    chr21 [47878411, 47878891]      * |   peak15052       104
##        signalValue    pvalue    qvalue
##          &lt;numeric&gt; &lt;integer&gt; &lt;integer&gt;
##    [1]      183.11        -1        -1
##    [2]     3064.92        -1        -1
##    [3]      291.90        -1        -1
##    [4]      940.03        -1        -1
##    [5]      104.67        -1        -1
##    ...         ...       ...       ...
##   [40]      106.36        -1        -1
##   [41]      217.02        -1        -1
##   [42]      124.31        -1        -1
##   [43]     1141.58        -1        -1
##   [44]      338.78        -1        -1
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
<p>For each CpG island, we can count the number of peaks that overlap with a given CpG island with countOverlaps().</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#count the peaks that</span>
<span class="co"># overlap with CpG islands</span>
counts=<span class="kw">countOverlaps</span>(pk1.gr,cpgi.gr)
<span class="kw">head</span>(counts)</code></pre></div>
<pre><code>## [1] 0 0 0 0 0 0</code></pre>
<p>The <code>findOverlaps()</code> function can be used to see one-to-one overlaps between peaks and CpG islands. It returns a matrix showing which peak overlaps with which CpGi island.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">findOverlaps</span>(pk1.gr,cpgi.gr)</code></pre></div>
<pre><code>## Hits object with 45 hits and 0 metadata columns:
##        queryHits subjectHits
##        &lt;integer&gt;   &lt;integer&gt;
##    [1]       123           1
##    [2]       154           3
##    [3]       389           8
##    [4]       401          13
##    [5]       415          16
##    ...       ...         ...
##   [41]       595         155
##   [42]       598         166
##   [43]       600         176
##   [44]       611         192
##   [45]       613         200
##   -------
##   queryLength: 620 / subjectLength: 205</code></pre>
<p>Another interesting thing would be to look at the distances to nearest CpG islands for each peak. In addition, just finding the nearest CpG island could also be interesting. Often times, you will need to find nearest TSS or gene to your regions of interest, and the code below is handy for doing that.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># find nearest CpGi to each TSS</span>
n.ind=<span class="kw">nearest</span>(pk1.gr,cpgi.gr)
<span class="co"># get distance to nearest</span>
dists=<span class="kw">distanceToNearest</span>(pk1.gr,cpgi.gr,<span class="dt">select=</span><span class="st">&quot;arbitrary&quot;</span>)
dists</code></pre></div>
<pre><code>## Hits object with 620 hits and 1 metadata column:
##         queryHits subjectHits |  distance
##         &lt;integer&gt;   &lt;integer&gt; | &lt;integer&gt;
##     [1]         1           1 |    384188
##     [2]         2           1 |    382968
##     [3]         3           1 |    381052
##     [4]         4           1 |    379311
##     [5]         5           1 |    376978
##     ...       ...         ... .       ...
##   [616]       616         205 |     26211
##   [617]       617         205 |     27401
##   [618]       618         205 |     30467
##   [619]       619         205 |     31610
##   [620]       620         205 |     34089
##   -------
##   queryLength: 620 / subjectLength: 205</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># histogram of the distances to nearest TSS</span>
dist2plot=<span class="kw">mcols</span>(dists)[,<span class="dv">1</span>]
<span class="kw">hist</span>(<span class="kw">log10</span>(dist2plot),<span class="dt">xlab=</span><span class="st">&quot;log10(dist to nearest TSS)&quot;</span>,
     <span class="dt">main=</span><span class="st">&quot;Distances&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:findNearest"></span>
<img src="compgenomr_files/figure-html/findNearest-1.png" alt="Figure  2: histogram of distances" width="672" />
<p class="caption">
Figure 5.1: Figure 2: histogram of distances
</p>
</div>
</div>
</div>
<div id="dealing-with-high-throughput-sequencing-reads" class="section level2">
<h2><span class="header-section-number">5.2</span> Dealing with high-throughput sequencing reads</h2>
<p>In recent years, advances in sequencing technology helped researchers sequence the genome deeper than ever. The reads from sequencing machines are usually aligned to the genome and the next task is to quantify the enrichment of those aligned reads in the regions of interest. You may want to count how many reads overlapping with your promoter set of interest or you may want to quantify RNA-seq reads overlapping with exons. This is similar to operations on genomic intervals which are described previously. If you can read all your alignments into the memory and create a GRanges object, you can apply the previously described operations. However, most of the time we can not read all mapped reads into the memory, so we have to use specialized tools to query and quantify alignments on a given set of regions. One of the most common alignment formats is SAM/BAM format, most aligners will produce SAM/BAM output or you will be able to convert your specific alignment format to SAM/BAM format. The BAM format is a binary version of the human readable SAM format. The SAM format has specific columns that contain different kind of information about the alignment such as mismatches, qualities etc. (see <a href="http://samtools.sourceforge.net/SAM1.pdf" class="uri">http://samtools.sourceforge.net/SAM1.pdf</a> for SAM format specification).</p>
<div id="quality-check-on-sequencing-reads-and-mapping-reads-to-the-genome" class="section level3">
<h3><span class="header-section-number">5.2.1</span> Quality check on sequencing reads and mapping reads to the genome</h3>
<p>The sequencing technologies usually produce basecalls with varying quality. In addition, there could be sample-prep specific issues in your sequencing run, such as adapter contamination. It is standard procedure to check the quality of the reads before further analysis. Checking the quality and making some decisions for the downstream analysis can influence the outcome of your analysis. For a long time, quality check and mapping tasks were outside the R domain. However, certain packages in Bioconductor currently can deal with those tasks. Although, we will not go into detail we will mention some packages that can help with quality check and mapping.</p>
<p>Read quality checking is possible with Bioconductor packages: <a href="http://www.bioconductor.org/packages/release/bioc/html/qrqc.html">qcrc</a>, <a href="http://www.bioconductor.org/packages/release/bioc/html/Rsubread.html">Rsubread</a> and <a href="http://www.bioconductor.org/packages/release/bioc/html/QuasR.html">QuasR</a>. All the packages seem to have interface to C for fast quality score calculations or I/O operations. For the read mapping, QuasR uses Rbowtie package and produces BAM files (see below for short intro on BAM) and Rsubread employs its own mapping algorithm and can also produce BAM files.</p>
</div>
<div id="counting-mapped-reads-for-a-set-of-regions" class="section level3">
<h3><span class="header-section-number">5.2.2</span> Counting mapped reads for a set of regions</h3>
<p>Rsamtools package has functions to query BAM files. The function we will use in the first example is countBam which takes input of the BAM file and param argument. “param” argument takes a ScanBamParam object. The object is instantiated using <strong>ScanBamParam()</strong> and contains parameters for scanning the BAM file. The example below is a simple example where ScanBamParam() only includes regions of interest, promoters on chr21.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># regions of interest</span>
<span class="co"># promoters on chr21</span>
promoter.gr=tss.gr
<span class="kw">start</span>(promoter.gr)=<span class="kw">start</span>(promoter.gr)-<span class="dv">1000</span>
<span class="kw">end</span>(promoter.gr)  =<span class="kw">end</span>(promoter.gr)+<span class="dv">1000</span>
promoter.gr=promoter.gr[<span class="kw">seqnames</span>(promoter.gr)==<span class="st">&quot;chr21&quot;</span>]

<span class="kw">library</span>(Rsamtools)
bamfile=<span class="st">&quot;GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam&quot;</span>

<span class="co"># get reads for regions of interest from the bam file</span>
param &lt;-<span class="st"> </span><span class="kw">ScanBamParam</span>(<span class="dt">which=</span>promoter.gr)
counts=<span class="kw">countBam</span>(bamfile, <span class="dt">param=</span>param)</code></pre></div>
<p>Alternatively, aligned reads can be read in using GenomicAlignments package (which on this occasion relies on RSamtools package).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicAlignments)
alns &lt;-<span class="st"> </span><span class="kw">readGAlignments</span>(bamfile, <span class="dt">param=</span>param)</code></pre></div>
</div>
</div>
<div id="dealing-with-continuous-scores-over-the-genome" class="section level2">
<h2><span class="header-section-number">5.3</span> Dealing with continuous scores over the genome</h2>
<p>Most high-throughput data can be viewed as a continuous score over the bases of the genome. In case of RNA-seq or ChIP-seq experiments the data can be represented as read coverage values per genomic base position. In addition, other information (not necessarily from high-throughput experiments) can be represented this way. The GC content and conservation scores per base are prime examples of other data sets that can be represented as scores. This sort of data can be stored as a generic text file or can have special formats such as Wig (stands for wiggle) from UCSC, or the bigWig format is which is indexed binary format of the wig files. The bigWig format is great for data that covers large fraction of the genome with varying scores, because the file is much smaller than regular text files that have the same information and it can be queried easier since it is indexed.</p>
<p>In R/Bioconductor, the continuous data can also be represented in a compressed format, in a format called Rle vector, which stands for run-length encoded vector. This gives superior memory performance over regular vectors because repeating consecutive values are represented as one value in the Rle vector (See Figure 2). <img src="GenomicIntervals_data/nonR_Figures/Rle_demo.png" alt="Rle" /> Figure 2: Rle encoding explained</p>
<p>Typically, for genome-wide data you will have a RleList object which is a list of Rle vectors per chromosome. You can obtain such vectors by reading the reads in and calling <em>coverage()</em> function from <em>GenomicRanges</em> package. Let’s try that on the above data set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">covs=<span class="kw">coverage</span>(alns) <span class="co"># get coverage vectors</span>
covs</code></pre></div>
<pre><code>## RleList of length 24
## $chr1
## integer-Rle of length 249250621 with 1 run
##   Lengths: 249250621
##   Values :         0
## 
## $chr2
## integer-Rle of length 243199373 with 1 run
##   Lengths: 243199373
##   Values :         0
## 
## $chr3
## integer-Rle of length 198022430 with 1 run
##   Lengths: 198022430
##   Values :         0
## 
## $chr4
## integer-Rle of length 191154276 with 1 run
##   Lengths: 191154276
##   Values :         0
## 
## $chr5
## integer-Rle of length 180915260 with 1 run
##   Lengths: 180915260
##   Values :         0
## 
## ...
## &lt;19 more elements&gt;</code></pre>
<p>Alternatively, you can get the coverage from the Bam file directly. Below, we are getting the coverage directly from the Bam file for our previously defined promoters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">covs=<span class="kw">coverage</span>(bamfile, <span class="dt">param=</span>param) <span class="co"># get coverage vectors</span></code></pre></div>
<p>One of the most common ways of storing score data is, as mentioned, wig or bigWig format. Most of the ENCODE project data can be downloaded in bigWig format. In addition, conservation scores can also be downloaded as wig/bigWig format. You can import bigWig files into R using <em>import()</em> function from <em>rtracklayer</em> package. However, it is generally not advisable to read the whole bigWig file in memory as it was the case with BAM files. Usually, you will be interested in only a fraction of the genome, such as promoters, exons etc. So it is best you extract the data for those regions and read those into memory rather than the whole file. Below we read the a bigWig file only for promoters. The operation returns an <em>GRanges</em> object with score column which indicates the scores in the BigWig file per genomic region.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rtracklayer)
<span class="co"># File from ENCODE ChIP-seq tracks</span>
bwFile=<span class="st">&quot;GenomicIntervals_data/data/wgEncodeHaibTfbsA549.chr21.bw&quot;</span>
bw.gr=<span class="kw">import</span>(bwFile, <span class="dt">which=</span>promoter.gr) <span class="co"># get coverage vectors</span>
bw.gr</code></pre></div>
<pre><code>## GRanges object with 9205 ranges and 1 metadata column:
##          seqnames               ranges strand |     score
##             &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##      [1]    chr21   [9825456, 9825457]      * |         1
##      [2]    chr21   [9825458, 9825464]      * |         2
##      [3]    chr21   [9825465, 9825466]      * |         4
##      [4]    chr21   [9825467, 9825470]      * |         5
##      [5]    chr21   [9825471, 9825471]      * |         6
##      ...      ...                  ...    ... .       ...
##   [9201]    chr21 [48055809, 48055856]      * |         2
##   [9202]    chr21 [48055857, 48055858]      * |         1
##   [9203]    chr21 [48055872, 48055921]      * |         1
##   [9204]    chr21 [48055944, 48055993]      * |         1
##   [9205]    chr21 [48056069, 48056118]      * |         1
##   -------
##   seqinfo: 1 sequence from an unspecified genome</code></pre>
<p>Following this we can create an <code>RleList</code> object from the GRanges with <code>coverage</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cov.bw=<span class="kw">coverage</span>(bw.gr,<span class="dt">weight =</span> <span class="st">&quot;score&quot;</span>)

<span class="co"># or get this directly from</span>
cov.bw=<span class="kw">import</span>(bwFile, <span class="dt">which=</span>promoter.gr,<span class="dt">as =</span> <span class="st">&quot;RleList&quot;</span>)</code></pre></div>
<div id="extracting-subsections-of-rle-and-rlelist-objects" class="section level3">
<h3><span class="header-section-number">5.3.1</span> Extracting subsections of Rle and RleList objects</h3>
<p>Frequently, we will need to extract subsections of the Rle vectors or RleList objects. We will need to do this to visualize that subsection or get some statistics out of those sections. For example, we could be interested in average coverage per base for the regions we are interested in. We have to extract those regions from RleList object and apply summary statistics. Below, we show how to extract subsections of RleList object. We are extracting promoter regions from ChIP-seq read coverage RleList. Following that, we will plot the one of the promoters associated coverage values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myViews=<span class="kw">Views</span>(cov.bw,<span class="kw">as</span>(promoter.gr,<span class="st">&quot;RangesList&quot;</span>)) <span class="co"># get subsets of coverage</span>
<span class="co"># there is a views object for each chromosome</span>
myViews</code></pre></div>
<pre><code>## RleViewsList of length 1
## names(1): chr21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myViews[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>## Views on a 48129895-length Rle subject
## 
## views:
##          start      end width
##   [1] 42218039 42220039  2001 [2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   [2] 17441841 17443841  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   [3] 17565698 17567698  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   [4] 30395937 30397937  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   [5] 27542138 27544138  2001 [1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 1 ...]
##   [6] 27511708 27513708  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   [7] 32930290 32932290  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   [8] 27542446 27544446  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   [9] 28338439 28340439  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
##   ...      ...      ...   ... ...
## [370] 47517032 47519032  2001 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...]
## [371] 47648157 47650157  2001 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...]
## [372] 47603373 47605373  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
## [373] 47647738 47649738  2001 [2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 ...]
## [374] 47704236 47706236  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
## [375] 47742785 47744785  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
## [376] 47881383 47883383  2001 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...]
## [377] 48054506 48056506  2001 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...]
## [378] 48024035 48026035  2001 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ...]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the coverage vector from the 5th view and plot</span>
<span class="kw">plot</span>(myViews[[<span class="dv">1</span>]][[<span class="dv">5</span>]],<span class="dt">type=</span><span class="st">&quot;l&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:getViews"></span>
<img src="compgenomr_files/figure-html/getViews-1.png" alt="Figure  4: Coverage vector extracted from RleList via Views() function is plotted as a line plot." width="672" />
<p class="caption">
Figure 5.2: Figure 4: Coverage vector extracted from RleList via Views() function is plotted as a line plot.
</p>
</div>
<p>Next, we are interested in average coverage per base for the promoters using summary functions that works on Views object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the mean of the views</span>
<span class="kw">head</span>(
  <span class="kw">viewMeans</span>(myViews[[<span class="dv">1</span>]])
)</code></pre></div>
<pre><code>## [1] 0.2258871 0.3498251 1.2243878 0.4997501 2.0904548 0.6996502</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the max of the views</span>
<span class="kw">head</span>(
  <span class="kw">viewMaxs</span>(myViews[[<span class="dv">1</span>]])
)</code></pre></div>
<pre><code>## [1]  2  4 12  4 21  6</code></pre>
</div>
</div>
<div id="visualizing-and-summarizing-genomic-intervals" class="section level2">
<h2><span class="header-section-number">5.4</span> Visualizing and summarizing genomic intervals</h2>
<p>Data integration and visualization is corner stone of genomic data analysis. Below, we will show different ways of integrating and visualizing genomic intervals.</p>
<div id="visualizing-intervals-in-locus-of-interest" class="section level3">
<h3><span class="header-section-number">5.4.1</span> Visualizing intervals in locus of interest</h3>
<p>Often times, we will be interested in particular genomic locus and try to visualize different genomic datasets over that locus. This is similar to looking at the data over one of the genome browsers. Below we will display genes, GpG islands and read coverage from a ChIP-seq experiment using Gviz package.For Gviz, we first need to set the tracks to display. The tracks can be in various formats. They can be R objects such as IRanges,GRanges and data.frame, or they can be in flat file formats such as BigWig,BED and BAM. After the tracks are set, we can display them with <code>plotTracks</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Gviz)
<span class="co"># set tracks to display</span>

<span class="co"># set CpG island track</span>
cpgi.track=<span class="kw">AnnotationTrack</span>(cpgi.gr,
                           <span class="dt">name =</span> <span class="st">&quot;CpG&quot;</span>)

<span class="co"># set gene track</span>
<span class="co"># we will get this from EBI Biomart webservice</span>
gene.track &lt;-<span class="st"> </span><span class="kw">BiomartGeneRegionTrack</span>(<span class="dt">genome =</span> <span class="st">&quot;hg19&quot;</span>,
                                    <span class="dt">chromosome =</span> <span class="st">&quot;chr21&quot;</span>, 
                                    <span class="dt">start =</span> <span class="dv">27698681</span>, <span class="dt">end =</span> <span class="dv">28083310</span>,
                                    <span class="dt">name =</span> <span class="st">&quot;ENSEMBL&quot;</span>)


<span class="co"># set track for ChIP-seq coverage</span>
cov.track=<span class="kw">DataTrack</span>(<span class="st">&quot;GenomicIntervals_data/data/wgEncodeHaibTfbsA549.chr21.bw&quot;</span>,<span class="dt">type =</span> <span class="st">&quot;l&quot;</span>,
                    <span class="dt">name=</span><span class="st">&quot;coverage&quot;</span>)
  
<span class="co"># call the display function plotTracks</span>
track.list=<span class="kw">list</span>(cpgi.track,gene.track,cov.track)
<span class="kw">plotTracks</span>(track.list,<span class="dt">from=</span><span class="dv">27698681</span>,<span class="dt">to=</span><span class="dv">28083310</span>,<span class="dt">chromsome=</span><span class="st">&quot;chr21&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-166"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-166-1.png" alt="Figure  3: tracks visualized using Gviz" width="672" />
<p class="caption">
Figure 5.3: Figure 3: tracks visualized using Gviz
</p>
</div>
</div>
<div id="summaries-of-genomic-intervals" class="section level3">
<h3><span class="header-section-number">5.4.2</span> Summaries of genomic intervals</h3>
<p>Looking at data one region at a time could be inefficient. One can summarize different data sets over thousands of regions of interest and identify patterns. This summaries can include different data types such as motifs, read coverage and other scores associated with genomic intervals. The <code>genomation</code> package can summarize and help identify patterns in the datasets. The datasets can have different kinds of information and multiple file types can be used such as BED, GFF, BAM and bigWig. We will look at H3K4me3 ChIP-seq and DNAse-seq signals from H1 embryonic stem cell line. H3K4me3 is usually associated with promoters and regions with high DNAse-seq signal are associated with accessible regions, that means mostly regulatory regions. We will summarize those datasets around the transcription start sites (TSS) of genes on chromosome 20 of human hg19 assembly. We will first read the genes and extract the region around TSS, 500bp upstream and downstream. We will then create a matrix of ChIP-seq scores for those regions, each row will represent a region around a specific TSS and columns will be the scores per base. We will then plot, average enrichment values around the TSSes of genes on chromosome 20.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get transcription start sites on chr20</span>
<span class="kw">library</span>(genomation)
feat=<span class="kw">readTranscriptFeatures</span>(<span class="st">&quot;GenomicIntervals_data/data/refseq.hg19.chr20.bed&quot;</span>,<span class="dt">remove.unusual =</span> <span class="ot">TRUE</span>,
                            <span class="dt">up.flank =</span> <span class="dv">500</span>, <span class="dt">down.flank =</span> <span class="dv">500</span>)
prom=feat$promoters <span class="co"># get promoters from the features</span>


<span class="co"># get for H3K4me3 values around TSSes</span>
<span class="co"># we use strand.aware=TRUE so - strands will</span>
<span class="co"># be reversed</span>
sm=<span class="kw">ScoreMatrix</span>(<span class="st">&quot;GenomicIntervals_data/data/H1.ESC.H3K4me3.chr20.bw&quot;</span>,prom,
               <span class="dt">type=</span><span class="st">&quot;bigWig&quot;</span>,<span class="dt">strand.aware =</span> <span class="ot">TRUE</span>)


<span class="co"># look for the average enrichment</span>
<span class="kw">plotMeta</span>(sm, <span class="dt">profile.names =</span> <span class="st">&quot;H3K4me3&quot;</span>, <span class="dt">xcoords =</span> <span class="kw">c</span>(-<span class="dv">500</span>,<span class="dv">500</span>),
         <span class="dt">ylab=</span><span class="st">&quot;H3K4me3 enrichment&quot;</span>,<span class="dt">dispersion =</span> <span class="st">&quot;se&quot;</span>,
         <span class="dt">xlab=</span><span class="st">&quot;bases around TSS&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-167"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-167-1.png" alt="Figure  4: meta region plot using genomation" width="672" />
<p class="caption">
Figure 5.4: Figure 4: meta region plot using genomation
</p>
</div>
<p>The pattern we see is expected, there is a dip just around TSS and signal is more intense on the downstream of the TSS. We can also plot a heatmap where each row is a region around TSS and color coded by enrichment. This can show us not only the general pattern as in the meta-region plot but also how many of the regions produce such a pattern.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatMatrix</span>(sm,<span class="dt">order=</span><span class="ot">TRUE</span>,<span class="dt">xcoords =</span> <span class="kw">c</span>(-<span class="dv">500</span>,<span class="dv">500</span>),<span class="dt">xlab=</span><span class="st">&quot;bases around TSS&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-168"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-168-1.png" alt="Figure  5: Heatmap of enrichment of H3K4me2 around TSS" width="672" />
<p class="caption">
Figure 5.5: Figure 5: Heatmap of enrichment of H3K4me2 around TSS
</p>
</div>
<p>Here we saw that about half of the regions do not have any signal. In addition it seems the multi-modal profile we have observed earlier is more complicated. Certain regions seems to have signal on both sides of the TSS, whereas others have signal mostly on the downstream side.</p>
<p>Normally, there would be more than one experiment or we can integrate datasets from public repositories. In this case, we can see how different signals look like on the regions we are interested in. Now, we will also use DNAse-seq data and create a list of matrices with our datasets and plot the average profile of the signals from both datasets.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sml=<span class="kw">ScoreMatrixList</span>(<span class="kw">c</span>(<span class="dt">H3K4me3=</span><span class="st">&quot;GenomicIntervals_data/data/H1.ESC.H3K4me3.chr20.bw&quot;</span>,
                      <span class="dt">DNAse=</span><span class="st">&quot;GenomicIntervals_data/data/H1.ESC.dnase.chr20.bw&quot;</span>),prom,
                      <span class="dt">type=</span><span class="st">&quot;bigWig&quot;</span>,<span class="dt">strand.aware =</span> <span class="ot">TRUE</span>)
<span class="kw">plotMeta</span>(sml)</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-169"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-169-1.png" alt="Figure  6: Average profiles of DNAse and H3K4me3 ChIP-seq" width="672" />
<p class="caption">
Figure 5.6: Figure 6: Average profiles of DNAse and H3K4me3 ChIP-seq
</p>
</div>
<p>We should now look at the heatmaps side by side and we should also cluster the rows based on their similarity. We will be using <code>multiHeatMatrix</code> since we have multiple ScoreMatrix objects in the list. In this case, we will also use <code>winsorize</code> argument to limit extreme values, every score above 95th percentile will be equalized the the value of the 95th percentile. In addition, <code>heatMatrix</code> and <code>multiHeatMatrix</code> can cluster the rows. Below, we will be using k-means clustering with 3 clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1029</span>)
<span class="kw">multiHeatMatrix</span>(sml,<span class="dt">order=</span><span class="ot">TRUE</span>,<span class="dt">xcoords =</span> <span class="kw">c</span>(-<span class="dv">500</span>,<span class="dv">500</span>),
                <span class="dt">xlab=</span><span class="st">&quot;bases around TSS&quot;</span>,<span class="dt">winsorize =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">95</span>),
                <span class="dt">matrix.main =</span> <span class="kw">c</span>(<span class="st">&quot;H3K4me3&quot;</span>,<span class="st">&quot;DNAse&quot;</span>),
                <span class="dt">column.scale=</span><span class="ot">TRUE</span>,
                <span class="dt">clustfun=</span>function(x) <span class="kw">kmeans</span>(x, <span class="dt">centers=</span><span class="dv">3</span>)$cluster)</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-170"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-170-1.png" alt="Figure  7: Heatmaps of H3K4me3 & DNAse data" width="672" />
<p class="caption">
Figure 5.7: Figure 7: Heatmaps of H3K4me3 &amp; DNAse data
</p>
</div>
<p>This revealed a different picture than we have observed before. Almost half of the promoters have no signal for DNAse or H3K4me3; these regions are probably not active and associated genes are not expressed. For regions with H3K4me3 signal, there are two major patterns. One pattern where both downstream and upstream of the TSS are enriched. On the other pattern, mostly downstream of the TSS is enriched.</p>
</div>
<div id="making-karyograms-and-circos-plots" class="section level3">
<h3><span class="header-section-number">5.4.3</span> Making karyograms and circos plots</h3>
<p>Chromosomal karyograms and circos plots are beneficial for displaying data over the whole genome of chromosomes of interest. Although,the information that can be displayed over these large regions are usually not very clear and only large trends can be discerned by eye, such as loss of methylation in large regions or genome-wide. Below, we are showing how to use <code>ggbio</code> package for plotting. This package has a slightly different syntax than base graphics. The syntax follows <a href="http://www.springer.com/de/book/9780387245447">grammar of graphics</a> logic. It is a deconstructed way of thinking about the plot. You add your data and apply mappings and transformations in order to achieve the final output. In ggbio, things are relatively easy since a high-level function <code>autoplot</code> function will recognize most of the datatypes and guess the most appropriate plot type. You can change it is behavior by applying low-level functions. We first get the sizes of chromosomes and make a karyogram template.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggbio)
<span class="kw">data</span>(ideoCyto, <span class="dt">package =</span> <span class="st">&quot;biovizBase&quot;</span>)
p &lt;-<span class="st"> </span><span class="kw">autoplot</span>(<span class="kw">seqinfo</span>(ideoCyto$hg19), <span class="dt">layout =</span> <span class="st">&quot;karyogram&quot;</span>)</code></pre></div>
<p>Next, we would like to plot CpG islands on this Karyogram. We simply do this by adding a layer with <code>layout_karyogram</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read CpG islands from a generic text file</span>
cpgi.gr=genomation::<span class="kw">readGeneric</span>(<span class="st">&quot;GenomicIntervals_data/data/CpGi.hg19.table.txt&quot;</span>, 
                        <span class="dt">chr =</span> <span class="dv">1</span>, <span class="dt">start =</span> <span class="dv">2</span>, <span class="dt">end =</span> <span class="dv">3</span>,<span class="dt">header=</span><span class="ot">TRUE</span>, 
                        <span class="dt">keep.all.metadata =</span><span class="ot">TRUE</span>,<span class="dt">remove.unusual=</span><span class="ot">TRUE</span> )

p +<span class="st"> </span><span class="kw">layout_karyogram</span>(cpgi.gr)</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-172"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-172-1.png" alt="Figure  9: Karyogram of CpG islands" width="672" />
<p class="caption">
Figure 5.8: Figure 9: Karyogram of CpG islands
</p>
</div>
<p>Next, we would like to plot some data over the chromosomes. This could be ChIP-seq signal or any other signal over the genome, we will use CpG island scores from the data set we read earlier. We will plot a point proportional to “obsExp” column in the data set. We use <code>ylim</code> argument to squish the chromosomal rectangles and plot on top of those. <code>aes</code> argument defines how the data is mapped to geometry. In this case, it says the points will have x coordinate from CpG island start positions and y coordinate from obsExp score of CpG islands.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p +<span class="st"> </span><span class="kw">layout_karyogram</span>(cpgi.gr, <span class="dt">geom =</span> <span class="st">&quot;point&quot;</span>, <span class="dt">size =</span> <span class="fl">0.5</span>, 
                         <span class="kw">aes</span>(<span class="dt">x =</span> start, <span class="dt">y =</span> obsExp),
                         <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">11</span>,<span class="dv">40</span>), <span class="dt">color =</span> <span class="st">&quot;red&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-173"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-173-1.png" alt="Figure  9: Karyogram of CpG islands" width="672" />
<p class="caption">
Figure 5.9: Figure 9: Karyogram of CpG islands
</p>
</div>
<p>Another way to depict regions or quantitative signals on the chromosomes is circos plots. These are circular plots usually used for showing chromosomal rearrangements, but can also be used for depicting signals.<code>ggbio</code> package can produce all kinds of circos plots. Below, we will show how to use that for our CpG island score example.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># set the chromsome in a circle</span>
<span class="co"># color set to white to look transparent </span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() +<span class="st"> </span><span class="kw">layout_circle</span>(ideoCyto$hg19, <span class="dt">geom =</span> <span class="st">&quot;ideo&quot;</span>, <span class="dt">fill =</span> <span class="st">&quot;white&quot;</span>,
                              <span class="dt">colour=</span><span class="st">&quot;white&quot;</span>,<span class="dt">cytoband =</span> <span class="ot">TRUE</span>,
                              <span class="dt">radius =</span> <span class="dv">39</span>, <span class="dt">trackWidth =</span> <span class="dv">2</span>)
<span class="co"># plot the scores as points   </span>
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">layout_circle</span>(cpgi.gr, <span class="dt">geom =</span> <span class="st">&quot;point&quot;</span>, <span class="dt">grid=</span><span class="ot">TRUE</span>,
                           <span class="dt">size =</span> <span class="dv">1</span>, <span class="kw">aes</span>(<span class="dt">y =</span> obsExp),<span class="dt">color=</span><span class="st">&quot;red&quot;</span>,
                       <span class="dt">radius =</span> <span class="dv">42</span>, <span class="dt">trackWidth =</span> <span class="dv">10</span>)
<span class="co"># set the chromosome names</span>
p &lt;-<span class="st"> </span>p +<span class="st"> </span><span class="kw">layout_circle</span>(<span class="kw">as</span>(<span class="kw">seqinfo</span>(ideoCyto$hg19),<span class="st">&quot;GRanges&quot;</span>), 
                       <span class="dt">geom =</span> <span class="st">&quot;text&quot;</span>, <span class="kw">aes</span>(<span class="dt">label =</span> seqnames), 
                      <span class="dt">vjust =</span> <span class="dv">0</span>, <span class="dt">radius =</span> <span class="dv">55</span>, <span class="dt">trackWidth =</span> <span class="dv">7</span>,
                      <span class="dt">size=</span><span class="dv">3</span>) 

<span class="co"># display the plot</span>
p</code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-174"></span>
<img src="compgenomr_files/figure-html/unnamed-chunk-174-1.png" alt="Figure  10: circos plot for CpG islands scores" width="672" />
<p class="caption">
Figure 5.10: Figure 10: circos plot for CpG islands scores
</p>
</div>
</div>
</div>
<div id="session-info-1" class="section level2">
<h2><span class="header-section-number">5.5</span> Session info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.0 (2016-05-03)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.11.4 (El Capitan)
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
##  [1] grid      stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] ggbio_1.20.0               Gviz_1.16.0               
##  [3] rtracklayer_1.32.0         GenomicAlignments_1.8.0   
##  [5] SummarizedExperiment_1.2.0 Biobase_2.32.0            
##  [7] Rsamtools_1.24.0           Biostrings_2.40.0         
##  [9] XVector_0.12.0             genomation_1.4.0          
## [11] GenomicRanges_1.24.0       GenomeInfoDb_1.8.0        
## [13] IRanges_2.6.0              S4Vectors_0.10.0          
## [15] BiocGenerics_0.18.0        scales_0.4.0              
## [17] plot3D_1.1                 matrixStats_0.50.2        
## [19] qvalue_2.4.0               mosaic_0.13.0             
## [21] mosaicData_0.13.0          car_2.1-2                 
## [23] ggplot2_2.1.0              lattice_0.20-33           
## [25] dplyr_0.4.3                captioner_2.2.3           
## [27] knitr_1.12.28             
## 
## loaded via a namespace (and not attached):
##  [1] minqa_1.2.4                   colorspace_1.2-6             
##  [3] biovizBase_1.20.0             ggdendro_0.1-20              
##  [5] dichromat_2.0-0               MatrixModels_0.4-1           
##  [7] interactiveDisplayBase_1.10.0 AnnotationDbi_1.34.0         
##  [9] codetools_0.2-14              splines_3.3.0                
## [11] impute_1.46.0                 Formula_1.2-1                
## [13] nloptr_1.0.4                  seqPattern_1.4.0             
## [15] pbkrtest_0.4-6                gridBase_0.4-7               
## [17] cluster_2.0.4                 graph_1.50.0                 
## [19] shiny_0.13.2                  readr_0.2.2                  
## [21] httr_1.1.0                    assertthat_0.1               
## [23] Matrix_1.2-6                  lazyeval_0.1.10              
## [25] formatR_1.4                   acepack_1.3-3.3              
## [27] htmltools_0.3.5               quantreg_5.21                
## [29] tools_3.3.0                   misc3d_0.8-4                 
## [31] gtable_0.2.0                  reshape2_1.4.1               
## [33] Rcpp_0.12.4.5                 nlme_3.1-127                 
## [35] stringr_1.0.0                 lme4_1.1-12                  
## [37] mime_0.4                      miniUI_0.1.1                 
## [39] ensembldb_1.4.0               XML_3.98-1.4                 
## [41] AnnotationHub_2.4.0           zlibbioc_1.18.0              
## [43] MASS_7.3-45                   BSgenome_1.40.0              
## [45] VariantAnnotation_1.18.0      BiocInstaller_1.22.1         
## [47] RBGL_1.48.0                   SparseM_1.7                  
## [49] RColorBrewer_1.1-2            yaml_2.1.13                  
## [51] gridExtra_2.2.1               biomaRt_2.28.0               
## [53] rpart_4.1-10                  reshape_0.8.5                
## [55] latticeExtra_0.6-28           stringi_1.0-1                
## [57] RSQLite_1.0.0                 plotrix_3.6-1                
## [59] GenomicFeatures_1.24.0        BiocParallel_1.6.0           
## [61] chron_2.3-47                  bitops_1.0-6                 
## [63] evaluate_0.9                  labeling_0.3                 
## [65] GGally_1.0.1                  plyr_1.8.3                   
## [67] magrittr_1.5                  bookdown_0.0.71              
## [69] R6_2.1.2                      Hmisc_3.17-4                 
## [71] DBI_0.4-1                     foreign_0.8-66               
## [73] mgcv_1.8-12                   survival_2.39-3              
## [75] RCurl_1.95-4.8                nnet_7.3-12                  
## [77] KernSmooth_2.23-15            OrganismDbi_1.14.0           
## [79] rmarkdown_0.9.6               data.table_1.9.6             
## [81] digest_0.6.9                  xtable_1.8-2                 
## [83] httpuv_1.3.3                  munsell_0.4.3</code></pre>
</div>
<div id="exercises-2" class="section level2">
<h2><span class="header-section-number">5.6</span> Exercises</h2>
<div id="the-setup-1" class="section level3">
<h3><span class="header-section-number">5.6.1</span> The setup</h3>
<p>Set your working directory to the source file location in RStudio top menu:</p>
<p>‘Session &gt; Set Working Directory &gt; To Source File Location’.</p>
<p>if you have done it correctly you should see this script when you type dir()</p>
<pre><code>&gt; dir()
[1] &quot;GenomicInterval.exercises.html&quot;</code></pre>
<p>The data for the exercises is located at <code>GenomicIntervals_data/data</code> folder.</p>
<p>Run the following to see the data files.</p>
<pre><code>dir(&quot;GenomicIntervals_data/data&quot;)</code></pre>
</div>
<div id="operations-on-genomic-intervals-with-genomicranges-package-1" class="section level3">
<h3><span class="header-section-number">5.6.2</span> Operations on Genomic Intervals with GenomicRanges package</h3>
<div id="section-95" class="section level4">
<h4><span class="header-section-number">5.6.2.1</span> </h4>
<p>Create a GRanges object using the information in the table below:</p>
<table>
<thead>
<tr class="header">
<th align="left">chr</th>
<th align="left">start</th>
<th align="left">end</th>
<th align="left">strand</th>
<th align="left">score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">chr1</td>
<td align="left">10000</td>
<td align="left">10300</td>
<td align="left">+</td>
<td align="left">10</td>
</tr>
<tr class="even">
<td align="left">chr1</td>
<td align="left">11100</td>
<td align="left">11500</td>
<td align="left">-</td>
<td align="left">20</td>
</tr>
<tr class="odd">
<td align="left">chr2</td>
<td align="left">20000</td>
<td align="left">20030</td>
<td align="left">+</td>
<td align="left">15</td>
</tr>
</tbody>
</table>
</div>
<div id="section-96" class="section level4">
<h4><span class="header-section-number">5.6.2.2</span> </h4>
<p>use <code>start()</code>, <code>end()</code>, <code>strand()</code>,<code>seqnames()</code> and <code>width()</code> functions on the GRanges object you created. Figure out what they are doing. Can you get a subset of GRanges object for intervals that are only on + strand? If you can do that, try getting intervals that are on chr1. <em>HINT:</em> GRanges objects can be subset using <a href="genomic-intervals-and-r.html#section-107"></a> operator similar to data.frames but you may need to use <code>start()</code>, <code>end()</code> and <code>strand()</code>,<code>seqnames()</code> within the <a href="genomic-intervals-and-r.html#section-107"></a>.</p>
</div>
<div id="section-97" class="section level4">
<h4><span class="header-section-number">5.6.2.3</span> </h4>
<p>Import mouse (mm9 assembly) CpG islands and refseq transcripts for chr12 from UCSC browser as <code>GRanges</code> objects using <code>rtracklayer</code> functions. HINT: Check the lecture material and modify the code there as necessary. If that somehow does not work, go to UCSC browser and download it as a BED file. The trackname for Refseq genes is “RefSeq Genes” and table name is “refGene”.</p>
</div>
<div id="section-98" class="section level4">
<h4><span class="header-section-number">5.6.2.4</span> </h4>
<p>Following from the exercise above, get the promoters of Refseq transcripts (-1000bp and +1000 bp of the TSS) and calculate what percentage of them overlap with CpG islands. HINT: You have to get the promoter coordinates and use <code>findOverlaps()</code> or <code>subsetByOverlaps()</code> from <code>GenomicRanges</code> package. To get promoters, type <code>?promoters</code> on the R console and see how to use that function to get promoters or calculate their coordinates as shown in the lecture material.</p>
</div>
<div id="section-99" class="section level4">
<h4><span class="header-section-number">5.6.2.5</span> </h4>
<p>Plot the distribution of CpG island lengths for CpG islands that overlap with the promoters.</p>
</div>
<div id="section-100" class="section level4">
<h4><span class="header-section-number">5.6.2.6</span> </h4>
<p>Get canonical peaks for SP1 (peaks that are in both replicates) on chr21. Peaks for each replicate are located in <code>GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz</code> and <code>GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep2.broadPeak.gz</code> files. HINT: You need to use <code>findOverlaps()</code> or <code>subsetByOverlaps()</code> to get the subset of peaks that occur in both replicates. You can try to read *broadPeak.gz files using genomation function <code>readBroadPeak</code>, broadPeak is just an extended BED format. EXTRA credit: Try use <code>coverage()</code> and slice<code>()</code> functions to get canonical peaks.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(genomation)
rep1=<span class="kw">readBroadPeak</span>(<span class="st">&quot;GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz&quot;</span>)
rep2=<span class="kw">readBroadPeak</span>(<span class="st">&quot;GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep2.broadPeak.gz&quot;</span>)</code></pre></div>
</div>
</div>
<div id="dealing-with-mapped-high-throughput-sequencing-reads" class="section level3">
<h3><span class="header-section-number">5.6.3</span> Dealing with mapped high-throughput sequencing reads</h3>
<div id="section-101" class="section level4">
<h4><span class="header-section-number">5.6.3.1</span> </h4>
<p>Count the reads overlapping with canonical Sp1 peaks using the BAM file for one of the replicates: <code>GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam</code>. <strong>HINT</strong>: Use functions from <code>GenomicAlignments</code>, see <a href="GenomicIntervals_data/lectures/genomicIntervalsTutorial.html">lecture notes</a>.</p>
</div>
</div>
<div id="dealing-with-contiguous-scores-over-the-genome" class="section level3">
<h3><span class="header-section-number">5.6.4</span> Dealing with contiguous scores over the genome</h3>
<div id="section-102" class="section level4">
<h4><span class="header-section-number">5.6.4.1</span> </h4>
<p>Extract Views object for the promoters on chr20 from <code>GenomicIntervals_data/data/H1.ESC.H3K4me1.chr20.bw</code> file. Plot the first “View” as a line plot. <strong>HINT</strong>: see <a href="GenomicIntervals_data/lectures/genomicIntervalsTutorial.html">lecture notes</a>, adapt the code from there.</p>
</div>
<div id="section-103" class="section level4">
<h4><span class="header-section-number">5.6.4.2</span> </h4>
<p>Make a histogram of the maximum signal for the Views in the object you extracted above. You can use any of the view summary functions or use lapply() and write your own summary function.</p>
</div>
<div id="section-104" class="section level4">
<h4><span class="header-section-number">5.6.4.3</span> </h4>
<p>Get the genomic positions of maximum signal in each view and make a GRanges object. <strong>HINT</strong>: See ?viewRangeMaxs help page. Try to make a GRanges object out of the returned object.</p>
</div>
</div>
<div id="visualizing-and-summarizing-genomic-intervals-1" class="section level3">
<h3><span class="header-section-number">5.6.5</span> Visualizing and summarizing genomic intervals</h3>
<div id="section-105" class="section level4">
<h4><span class="header-section-number">5.6.5.1</span> </h4>
<p>Extract -500,+500 bp regions around TSSes on chr21, there are refseq files in the <code>GenomicIntervals_data/data</code> folder or you can pull the data out of UCSC browser. Use SP1 ChIP-seq data (<code>GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam</code> ) to create an average profile of read coverage around TSSes. Following that, visualize the read coverage with a heatmap. <strong>HINT</strong>: All of these possible using <code>genomation</code> package functions.</p>
</div>
<div id="section-106" class="section level4">
<h4><span class="header-section-number">5.6.5.2</span> </h4>
<p>Extract -500,+500 bp regions around TSSes on chr20. Use H3K4me3 (<code>GenomicIntervals_data/data/H1.ESC.H3K4me3.chr20.bw</code>) and H3K27ac (<code>GenomicIntervals_data/data/H1.ESC.H3K27ac.chr20.bw</code>) ChIP-seq enrichment data in the data folder and create heatmaps and average signal profiles for regions around the TSSes.</p>
</div>
<div id="section-107" class="section level4">
<h4><span class="header-section-number">5.6.5.3</span> </h4>
<p>Visualize one of the -500,+500 bp regions around TSS using <code>Gviz</code> functions. You should visualize both H3K4me3 and H3K27ac and the gene models.</p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="statistics-for-genomics.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>


<script src="book_assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="book_assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="book_assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
