<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Computational Genomics with R</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="book files for computatiation genomcis with R">
  <meta name="generator" content="bookdown 0.0.71 and GitBook 2.6.7">

  <meta property="og:title" content="Computational Genomics with R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="book files for computatiation genomcis with R" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Computational Genomics with R" />
  
  <meta name="twitter:description" content="book files for computatiation genomcis with R" />
  

<meta name="author" content="Altuna Akalin">

<meta name="date" content="2016-05-10">

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="introduction-to-r.html">

<script src="book_assets/jquery-2.2.3/jquery.min.js"></script>
<link href="book_assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#why-r"><i class="fa fa-check"></i><b>1.1</b> Why R ?</a><ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#data-munging-pre-processing"><i class="fa fa-check"></i><b>1.1.1</b> Data munging (pre-processing)</a></li>
<li class="chapter" data-level="1.1.2" data-path="index.html"><a href="index.html#general-data-anaylsis-and-exploration"><i class="fa fa-check"></i><b>1.1.2</b> General data anaylsis and exploration</a></li>
<li class="chapter" data-level="1.1.3" data-path="index.html"><a href="index.html#visualization"><i class="fa fa-check"></i><b>1.1.3</b> Visualization</a></li>
<li class="chapter" data-level="1.1.4" data-path="index.html"><a href="index.html#dealing-with-genomic-intervals"><i class="fa fa-check"></i><b>1.1.4</b> Dealing with genomic intervals</a></li>
<li class="chapter" data-level="1.1.5" data-path="index.html"><a href="index.html#application-of-other-bioinformatics-specific-algorithms"><i class="fa fa-check"></i><b>1.1.5</b> Application of other bioinformatics specific algorithms</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#who-is-this-book-for"><i class="fa fa-check"></i><b>1.2</b> Who is this book for?</a><ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#what-will-you-get-out-of-this"><i class="fa fa-check"></i><b>1.2.1</b> What will you get out of this?</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i><b>1.3</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html"><i class="fa fa-check"></i><b>2</b> Introduction to Genomics</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#dna-genes-and-central-dogma"><i class="fa fa-check"></i><b>2.1</b> DNA, genes and central dogma</a><ul>
<li class="chapter" data-level="2.1.1" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#gene-transcription-and-its-implications"><i class="fa fa-check"></i><b>2.1.1</b> Gene transcription and its implications</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#gene-regulation"><i class="fa fa-check"></i><b>2.2</b> Gene regulation</a><ul>
<li class="chapter" data-level="2.2.1" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#core-promoters"><i class="fa fa-check"></i><b>2.2.1</b> Core Promoters</a></li>
<li class="chapter" data-level="2.2.2" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#proximal-and-long-range-distal-regulation"><i class="fa fa-check"></i><b>2.2.2</b> Proximal and long-range (distal) regulation</a></li>
<li class="chapter" data-level="2.2.3" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#epigenetic-regulation"><i class="fa fa-check"></i><b>2.2.3</b> Epigenetic regulation</a></li>
<li class="chapter" data-level="2.2.4" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#regulation-by-ncrnas"><i class="fa fa-check"></i><b>2.2.4</b> Regulation by ncRNAs</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#high-throughput-experimental-methods-in-genomics"><i class="fa fa-check"></i><b>2.3</b> High-throughput experimental methods in genomics</a><ul>
<li class="chapter" data-level="2.3.1" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#end-sequencing"><i class="fa fa-check"></i><b>2.3.1</b> 5’ end sequencing:</a></li>
<li class="chapter" data-level="2.3.2" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#chromatin-immunoprecipitation-chip-followed-by-microarray-hybridization-or-sequencing"><i class="fa fa-check"></i><b>2.3.2</b> Chromatin immunoprecipitation (ChIP) followed by microarray hybridization or sequencing:</a></li>
<li class="chapter" data-level="2.3.3" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#dnase-hypersensitive-sites"><i class="fa fa-check"></i><b>2.3.3</b> DNAse hypersensitive sites:</a></li>
<li class="chapter" data-level="2.3.4" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#bisulfite-sequencing"><i class="fa fa-check"></i><b>2.3.4</b> Bisulfite sequencing:</a></li>
<li class="chapter" data-level="2.3.5" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html#chromosome-conformation-capture-based-technologies"><i class="fa fa-check"></i><b>2.3.5</b> Chromosome-conformation capture based technologies:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>3</b> Introduction to R</a><ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#the-setup"><i class="fa fa-check"></i><b>3.1</b> The Setup</a><ul>
<li class="chapter" data-level="3.1.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#installing-packages"><i class="fa fa-check"></i><b>3.1.1</b> Installing packages</a></li>
<li class="chapter" data-level="3.1.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#installing-packages-in-custom-locations"><i class="fa fa-check"></i><b>3.1.2</b> Installing packages in custom locations</a></li>
<li class="chapter" data-level="3.1.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#getting-help-on-functions-and-packages"><i class="fa fa-check"></i><b>3.1.3</b> Getting help on functions and packages</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#computations-in-r"><i class="fa fa-check"></i><b>3.2</b> Computations in R</a></li>
<li class="chapter" data-level="3.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-structures"><i class="fa fa-check"></i><b>3.3</b> Data structures</a><ul>
<li class="chapter" data-level="3.3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>3.3.1</b> Vectors</a></li>
<li class="chapter" data-level="3.3.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#matrices"><i class="fa fa-check"></i><b>3.3.2</b> Matrices</a></li>
<li class="chapter" data-level="3.3.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-frames"><i class="fa fa-check"></i><b>3.3.3</b> Data Frames</a></li>
<li class="chapter" data-level="3.3.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#lists"><i class="fa fa-check"></i><b>3.3.4</b> Lists</a></li>
<li class="chapter" data-level="3.3.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#factors"><i class="fa fa-check"></i><b>3.3.5</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-types"><i class="fa fa-check"></i><b>3.4</b> Data types</a></li>
<li class="chapter" data-level="3.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#reading-and-writing-data"><i class="fa fa-check"></i><b>3.5</b> Reading and writing data</a></li>
<li class="chapter" data-level="3.6" data-path="introduction-to-r.html"><a href="introduction-to-r.html#plotting-in-r"><i class="fa fa-check"></i><b>3.6</b> Plotting in R</a></li>
<li class="chapter" data-level="3.7" data-path="introduction-to-r.html"><a href="introduction-to-r.html#functions-and-control-structures-for-ifelse-etc."><i class="fa fa-check"></i><b>3.7</b> Functions and control structures (for, if/else etc.)</a><ul>
<li class="chapter" data-level="3.7.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#user-defined-functions"><i class="fa fa-check"></i><b>3.7.1</b> User defined functions</a></li>
<li class="chapter" data-level="3.7.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#loops-and-looping-structures-in-r"><i class="fa fa-check"></i><b>3.7.2</b> Loops and looping structures in R</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="introduction-to-r.html"><a href="introduction-to-r.html#session-info"><i class="fa fa-check"></i><b>3.8</b> Session info</a></li>
<li class="chapter" data-level="3.9" data-path="introduction-to-r.html"><a href="introduction-to-r.html#exercises"><i class="fa fa-check"></i><b>3.9</b> Exercises</a><ul>
<li class="chapter" data-level="3.9.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#computations-in-r-1"><i class="fa fa-check"></i><b>3.9.1</b> Computations in R</a></li>
<li class="chapter" data-level="3.9.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-structures-in-r"><i class="fa fa-check"></i><b>3.9.2</b> Data structures in R</a></li>
<li class="chapter" data-level="3.9.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#reading-in-and-writing-data-out-in-r"><i class="fa fa-check"></i><b>3.9.3</b> Reading in and writing data out in R</a></li>
<li class="chapter" data-level="3.9.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#plotting-in-r-1"><i class="fa fa-check"></i><b>3.9.4</b> Plotting in R</a></li>
<li class="chapter" data-level="3.9.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#functions-and-control-structures-for-ifelse-etc.-1"><i class="fa fa-check"></i><b>3.9.5</b> Functions and control structures (for, if/else etc.)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html"><i class="fa fa-check"></i><b>4</b> Statistics for Genomics</a><ul>
<li class="chapter" data-level="4.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions"><i class="fa fa-check"></i><b>4.1</b> How to summarize collection of data points: The idea behind statistical distributions</a><ul>
<li class="chapter" data-level="4.1.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#describing-the-central-tendency-mean-and-median"><i class="fa fa-check"></i><b>4.1.1</b> Describing the central tendency: mean and median</a></li>
<li class="chapter" data-level="4.1.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#describing-the-spread-measurements-of-variation"><i class="fa fa-check"></i><b>4.1.2</b> Describing the spread: measurements of variation</a></li>
<li class="chapter" data-level="4.1.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#precision-of-estimates-confidence-intervals"><i class="fa fa-check"></i><b>4.1.3</b> Precision of estimates: Confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-test-for-differences-in-samples"><i class="fa fa-check"></i><b>4.2</b> How to test for differences in samples</a><ul>
<li class="chapter" data-level="4.2.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#randomization-based-testing-for-difference-of-the-means"><i class="fa fa-check"></i><b>4.2.1</b> randomization based testing for difference of the means</a></li>
<li class="chapter" data-level="4.2.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#using-t-test-for-difference-of-the-means-between-two-samples"><i class="fa fa-check"></i><b>4.2.2</b> Using t-test for difference of the means between two samples</a></li>
<li class="chapter" data-level="4.2.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#multiple-testing-correction"><i class="fa fa-check"></i><b>4.2.3</b> multiple testing correction</a></li>
<li class="chapter" data-level="4.2.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#moderated-t-tests-using-information-from-multiple-comparisons"><i class="fa fa-check"></i><b>4.2.4</b> moderated t-tests: using information from multiple comparisons</a></li>
<li class="chapter" data-level="4.2.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#want-to-know-more"><i class="fa fa-check"></i><b>4.2.5</b> Want to know more…</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#relationship-between-variables-linear-models-and-correlation"><i class="fa fa-check"></i><b>4.3</b> Relationship between variables: linear models and correlation</a><ul>
<li class="chapter" data-level="4.3.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-fit-a-line"><i class="fa fa-check"></i><b>4.3.1</b> How to fit a line</a></li>
<li class="chapter" data-level="4.3.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-estimate-the-error-of-the-coefficients"><i class="fa fa-check"></i><b>4.3.2</b> How to estimate the error of the coefficients</a></li>
<li class="chapter" data-level="4.3.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#accuracy-of-the-model"><i class="fa fa-check"></i><b>4.3.3</b> Accuracy of the model</a></li>
<li class="chapter" data-level="4.3.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#regression-with-categorical-variables"><i class="fa fa-check"></i><b>4.3.4</b> Regression with categorical variables</a></li>
<li class="chapter" data-level="4.3.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#regression-pitfalls"><i class="fa fa-check"></i><b>4.3.5</b> Regression pitfalls</a></li>
<li class="chapter" data-level="4.3.6" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#want-to-know-more-1"><i class="fa fa-check"></i><b>4.3.6</b> Want to know more…</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#roadmap-for-future"><i class="fa fa-check"></i><b>4.4</b> Roadmap for future</a><ul>
<li class="chapter" data-level="4.4.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#clustering"><i class="fa fa-check"></i><b>4.4.1</b> Clustering</a></li>
<li class="chapter" data-level="4.4.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#dimension-reduction-with-pca"><i class="fa fa-check"></i><b>4.4.2</b> Dimension reduction with PCA</a></li>
<li class="chapter" data-level="4.4.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#classification"><i class="fa fa-check"></i><b>4.4.3</b> Classification</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#exercises-1"><i class="fa fa-check"></i><b>4.5</b> Exercises</a><ul>
<li class="chapter" data-level="4.5.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1"><i class="fa fa-check"></i><b>4.5.1</b> How to summarize collection of data points: The idea behind statistical distributions</a></li>
<li class="chapter" data-level="4.5.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-test-for-differences-in-samples-1"><i class="fa fa-check"></i><b>4.5.2</b> How to test for differences in samples</a></li>
<li class="chapter" data-level="4.5.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#relationship-between-variables-linear-models-and-correlation-1"><i class="fa fa-check"></i><b>4.5.3</b> Relationship between variables: linear models and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html"><i class="fa fa-check"></i><b>5</b> Genomic Intervals and R</a><ul>
<li class="chapter" data-level="5.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#operations-on-genomic-intervals-with-genomicranges-package"><i class="fa fa-check"></i><b>5.1</b> Operations on Genomic Intervals with GenomicRanges package</a><ul>
<li class="chapter" data-level="5.1.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>5.1.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="5.1.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>5.1.2</b> Getting genomic regions into R as GRanges objects</a></li>
<li class="chapter" data-level="5.1.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>5.1.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-high-throughput-sequencing-reads"><i class="fa fa-check"></i><b>5.2</b> Dealing with high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="5.2.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#quality-check-on-sequencing-reads-and-mapping-reads-to-the-genome"><i class="fa fa-check"></i><b>5.2.1</b> Quality check on sequencing reads and mapping reads to the genome</a></li>
<li class="chapter" data-level="5.2.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>5.2.2</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-continuous-scores-over-the-genome"><i class="fa fa-check"></i><b>5.3</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="5.3.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>5.3.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-and-summarizing-genomic-intervals"><i class="fa fa-check"></i><b>5.4</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="5.4.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-intervals-in-locus-of-interest"><i class="fa fa-check"></i><b>5.4.1</b> Visualizing intervals in locus of interest</a></li>
<li class="chapter" data-level="5.4.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#summaries-of-genomic-intervals"><i class="fa fa-check"></i><b>5.4.2</b> Summaries of genomic intervals</a></li>
<li class="chapter" data-level="5.4.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>5.4.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#session-info-1"><i class="fa fa-check"></i><b>5.5</b> Session info</a></li>
<li class="chapter" data-level="5.6" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#exercises-2"><i class="fa fa-check"></i><b>5.6</b> Exercises</a><ul>
<li class="chapter" data-level="5.6.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#the-setup-1"><i class="fa fa-check"></i><b>5.6.1</b> The setup</a></li>
<li class="chapter" data-level="5.6.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#operations-on-genomic-intervals-with-genomicranges-package-1"><i class="fa fa-check"></i><b>5.6.2</b> Operations on Genomic Intervals with GenomicRanges package</a></li>
<li class="chapter" data-level="5.6.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-mapped-high-throughput-sequencing-reads"><i class="fa fa-check"></i><b>5.6.3</b> Dealing with mapped high-throughput sequencing reads</a></li>
<li class="chapter" data-level="5.6.4" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-contiguous-scores-over-the-genome"><i class="fa fa-check"></i><b>5.6.4</b> Dealing with contiguous scores over the genome</a></li>
<li class="chapter" data-level="5.6.5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-and-summarizing-genomic-intervals-1"><i class="fa fa-check"></i><b>5.6.5</b> Visualizing and summarizing genomic intervals</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Genomics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-genomics" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Introduction to Genomics</h1>
<div id="dna-genes-and-central-dogma" class="section level2">
<h2><span class="header-section-number">2.1</span> DNA, genes and central dogma</h2>
<p>All living organisms store and transmit information necessary for their development and functioning. This information can be coded into a string - a molecule of DNA, a polymer made of recurring units called nucleotides. The four nucleotides in DNA molecules, Adenine, Guanine, Cytosine and Thymine (coded as four letters: A, C, G, and T) in a specific sequence, store the information for life. DNA is organized in a double-helix form where two complementary polymers interlace with each other and twist into the familiar shape. In eukaryotic cells DNA is wrapped around proteins (histones) forming higher-order structures like nucleosomes, chromatins and chromosomes. There might be several chromosomes depending on the organisms. In most prokaryotes DNA is stored in a circular form. The full DNA sequence of an organism is called a genome.</p>
<p>In the genome, there are specific regions containing precise information that encodes for a physical products of genetic information. A region in the genome with this kind of information is traditionally called a “gene”. However, the precise definition of the gene is still developing. According to the classical textbooks in molecular biology, a gene is a segment of a DNA sequence corresponding to a single protein or to a single catalytic and structural RNA molecule [1]. A modern definition is: “A region (or regions) that includes all of the sequence elements necessary to encode a functional transcript” [2]. No matter how variable the definitions are, all agree on the fact that genes are basic units of heredity in all living organisms. All cells use their hereditary information in the same way most of the time; the DNA is replicated in order to transfer the information to new cells. First, genes are transcribed into messenger RNAs (mRNAs) in nucleus (in eukaryotes), followed by mRNAs (if the gene is protein coding) translation into proteins in the cytoplasm. This is essentially a process of information transfer between information carrying polymers; DNA, RNA and proteins, known as the “central dogma” of molecular biology.</p>
<div id="gene-transcription-and-its-implications" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Gene transcription and its implications</h3>
<p>The first step in a process of information transfer - a production of an RNA copy (of a part of the DNA sequence) - is called transcription. This task is carried out by the RNA polymerase enzyme. RNA polymerase-dependent initiation of transcription is enabled by the existence of a specific region in the sequence of DNA - a core promoter. Core promoters are regions of DNA that promote transcription and are found upstream from the start site of transcription. In eukaryotes, several proteins, called general transcription factors recognize and bind to core promoters and form a pre-initiation complex. RNA polymerases recognize these complexes and initiate synthesis of all RNAs [3]. After mRNA is produced it is often translated into proteins. However, non-coding RNA (ncRNAs) genes are processed and assume their functional structures after that. This set of processes is usually referred as gene expression.</p>
</div>
</div>
<div id="gene-regulation" class="section level2">
<h2><span class="header-section-number">2.2</span> Gene regulation</h2>
<p>The mechanisms regulating gene expression are essential for all living organisms as they dictate where and how much of a gene product (may it be protein or ncRNA) should be manufactured. Gene expression regulation, commonly referred to as gene regulation, drives cellular differentiation; a process during which different tissues and cell types are produced. It also helps cells maintain differentiated states of cells/tissues. As a product of this process, at the final stage of differentiation, different kinds of cells maintain different expression profiles although they contain the same genetic material.</p>
<p>The rate of transcription initiation is the primary regulatory element in gene regulation. The rate is controlled by core promoter elements as well as distant-acting regulatory elements such as enhancers. On top of that, processes like histone modifications and/or DNA methylation have a crucial regulatory impact on transcription. If a region is not accessible for the transcriptional machinery, e.g. in the case when chromatin structure is compacted due to the presence of specific histone modifications, or if the promoter DNA is methylated, transcription may not start at all. Last but the not least, gene activity is also controlled post-transcriptionally by ncRNAs such as microRNAs (miRNAs), as well as by cell signaling resulting in protein modification or altered protein-protein interactions.</p>
<div id="core-promoters" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Core Promoters</h3>
<p>Core promoters are the immediate neighboring regions around the transcription start site (TSS) that serves as a docking site for the transcriptional machinery and pre-initiation complex (PIC) assembly [4]. The textbook model for transcription initiation is as follows: The core promoter has a TATA motif (referred as TATA-box) 30 bp upstream of an initiator sequence (Inr), which also containsTSS. Firstly, transcription factor TFIID binds to the TATA-box. Next, general transcription factors are recruited and transcription is initiated on the initiator sequence (see [4-6] for reviews). Apart from the TATA-box and Inr, there are a number of sequence elements on the metazoan core promoters that are associated with transcription initiation and PIC assembly, such as downstream promoter elements (DPEs), the BRE elements and CpG islands. DPEs are found 28-32 bp downstream of the TSS in TATA-less promoters of Drosophila melanogaster, it generally co-occurs with the Inr element, and is thought to have a similar function to the TATA-box [7, 8]. The BRE element is recognized by TFIIB protein and lies upstream of the TATA-box [9]. CpG islands are CG dinucleotide-enriched segments of vertebrate genomes, despite the general depletion of CG dinucleotides in those genomes [10]. 50-70% of promoters in human genome are associated with CpG islands [11, 12].</p>
<p>Core promoter elements are not as universal as was previously thought. Gershenzon et al. [13] found Inr to be the most common element and TATA-boxes were present only in one eighth of analyzed promoters. Similarly, Cooper et al. [14] estimate a low occurrence of TATA-box in core promoters. In addition, as noted above, not every promoter is associated with a CpG island. All these indicate that the presence of a TATA-box is an exception associated with a particular subset of genes rather than the rule, and this conclusion directly contradicts the long-held textbook transcription initiation model. In addition to that contradiction, recent advances have shown that core promoter elements and PIC assembly may not be sufficient for transcription initiation and elongation. RNA polymerase can be paused on the initiation site and can be reactivated in a later stage, especially in development (reviewed in [15]).</p>
<p>Recent improvements in high-throughput sequencing techniques have enabled scientists to perform genome-wide analysis of TSSs, and thus helped them to further understand core promoter features in mammalian genomes. Carninci et al. [16] sequenced 5’ends of mRNAs from various tissues in mouse and human using a technology called CAGE (Cap analysis of gene expression). 5’-end mRNA sequencing of sufficient depth essentially shows all the 5’ capped transcription initiation events on the genome for the cells or tissues in the experiment. Authors concluded that the distribution of TSS sites on the promoters indicates the existence of two major classes of promoters. “Broad promoters” had multiple TSSs spanning around 50-100 bp, and usually none of those were dominant over the other ones. On the other hand, “sharp promoters” had a dominant TSS and most of the transcription initiation events originated from a very short region[17]. Sharp promoters were associated with TATA-boxes and tissue specific expression on the other hand majority of broad promoters were associated with CpG islands and expressed ubiquitously [17].</p>
</div>
<div id="proximal-and-long-range-distal-regulation" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Proximal and long-range (distal) regulation</h3>
<p>Context-specific control over transcription is provided by cis-regulatory elements outside the core promoters. Such elements include enhancers, silencers, insulators and locus control regions [4] (See Figure 1 for their hypothesized organization around a gene).</p>
<div id="proximal-promoters" class="section level4">
<h4><span class="header-section-number">2.2.2.1</span> Proximal promoters:</h4>
<p>Proximal promoter elements are typically right upstream of the core promoters and usually contain binding sites for activator transcription factors. Early studies on regulation showed that there are multiple functional elements near the TSSs [18]; this observation is also confirmed by the latest high-throughput genome wide experiments [19, 20]. In summary proximal promoter elements contain binding sites for transcription factors, they are located right upstream of core-promoter elements, and they provide additional control over gene expression.</p>
</div>
<div id="enhancers" class="section level4">
<h4><span class="header-section-number">2.2.2.2</span> Enhancers:</h4>
<p>Proximal regulation is not the only, nor the most important mode of gene regulation. In fact, the same recent high-throughput experiments cited above [19, 20] suggest that most of the transcription factor binding sites in the human or Drosophila genome are found in intergenic regions or in introns . This indicates the widespread usage of distant regulatory elements of metazoan genomes. A number of enhancers have been studied in depth. On a molecular function level, they are similar to proximal promoters; they contain binding sites for the same transcriptional activators and they basically enhance the gene expression [4]. However, they are often highly modular and several of them can affect the same promoter at the same time or in different time-points or tissues [21, 22]. In addition, their activity is independent of their orientation and their distance to the promoter they interact with [21, 23]. A number of studies showed that enhancers can act upon their target genes over several kilobases away [24, 25] (also reviewed in [26]). According to a popular model, enhancers achieve this by looping the DNA and coming to contact with their target genes [27].</p>
</div>
<div id="silencers" class="section level4">
<h4><span class="header-section-number">2.2.2.3</span> Silencers:</h4>
<p>Silencers are similar to enhancers; however their effect is opposite of enhancers on the transcription of the target gene, and results in decreasing their level of transcription. They contain binding sites for repressive transcription factors. Repressor transcription factors can either block the binding of an activator , directly compete for the same binding site, or induce a repressive chromatin state in which no activator binding is possible. Silencer effects, similar to those of enhancers, are independent of orientation and distance to target genes. In contradiction to this general view, in Drosophila there are two types of silencers, long-range and short-range. Short-range silencers are close to promoters and long-range silencers can silence multiple promoters or enhancers over kilobases away [28]. Like enhancers, silencers bound by repressors may also induce changes in DNA structure by looping and creating higher order structures [29]. One class of such repressor proteins, which is thought to initiate higher-order structures by looping, is Polycomb group proteins (PcGs). These proteins are thought to maintain a silenced state of the genes by interacting with Polycomb response elements (PREs) and promoters. Recently it has been shown that PcGs, PREs and promoters interact with each other even at a distance,and by these interactions they maintain a compacted chromosome structure for repressed genes [30, 31].</p>
</div>
<div id="insulators" class="section level4">
<h4><span class="header-section-number">2.2.2.4</span> Insulators:</h4>
<p>Insulator regions limit the effect of other regulatory elements to certain chromosomal boundaries; in other words, they create regulatory domains untainted by the regulatory elements in regions outside that domain. Insulators can block enhancer-promoter communication and/or prevent spreading of repressive chromatin domains. In vertebrates and insects, some of the well-studied insulators are bound by CTCF (CCCTC-binding factor) [32-34]. Genome-wide studies from different mammalian tissues confirm that CTCF binding is largely invariant of cell type, and CTCF motif locations are conserved in vertebrates [35]. At present, there are two models of explaining the insulator function; the most prevalent model claims insulators create physically separate domains by modifying chromosome structure. This is thought to be achieved by CTCF-driven chromatin looping andrecent evidence shows that CTCF can induce a higher-order chromosome structure through creating loops of chromatins [29, 36]. According to the second model, an insulator-bound activator cannot bind an enhancer; thus enhancer-blocking activity is achieved and insulators can also recruit active histone domain, creating an active domain for enhancers to function [37, 38].</p>
</div>
<div id="locus-control-regions" class="section level4">
<h4><span class="header-section-number">2.2.2.5</span> Locus control regions:</h4>
<p>Locus control regions (LCRs) are clusters of different regulatory elements that control entire set of genes on a locus. LCRs help genes achieve their temporal and/or tissue-specific expression programs. LCRs may be composed of multiple cis-regulatory elements, such as insulators, enhancers and they act upon their targets even from long distances [39]. However LCRs function with an orientation dependent manner, for example the activity of b-globin LCR is lost if inverted [40]. The mechanism of LCR function otherwise seems similar to other long-range regulators described above. The evidence is mounting in the direction of a model where DNA-looping creates a chromosomal structure in which target genes are clustered together, which seems to be essential for maintaining open chromatin domain [41, 42].</p>
</div>
</div>
<div id="epigenetic-regulation" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Epigenetic regulation</h3>
<p>Epigenetics in biology usually refers to constructions (chromatin structure, DNA methylation etc.) other than DNA sequence that influence gene regulation. In essence, epigenetic regulation is the regulation of DNA packing and structure, the consequence of which is gene expression regulation therefore the regulation of gene expression as a consequence. A typical example is that DNA packing inside the nucleus can directly influence gene expression [55]. There are two main mechanisms in epigenetic regulation: i) DNA methylation ii) histone modifications.</p>
<div id="dna-methylation" class="section level4">
<h4><span class="header-section-number">2.2.3.1</span> DNA methylation:</h4>
<p>DNA methylation is usually associated with gene silencing. DNA methyltransferase catalyzes the addition of a methyl group to cytosine of CpG dinucleotides (while in mammals the addition of methyl group is largely restricted to CpG dinucleotides, methylation can occur in other bases as well) [56]. This covalent modification either interferes with transcription factor binding on the region, or methyl-CpG binding proteins induce the spread of repressive chromatin domains, thus the gene is silenced if its promoter has methylated CG dinucleotides [57]. DNA methylation usually occurs in repeat sequences to repress transposable elements. DNA methylation is also related to a key core and proximal promoter element: CpG islands. CpG islands are usually unmethylated, however for some genes CpG island methylation accompanies their silenced expression [58]. For example, during X-chromosome inactivation many CpG islands are heavily methylated and the associated genes are silenced [58]. In addition, in embryonic stem cell differentiation pluripotency-associated genes are silenced due to DNA methylation [59].</p>
</div>
<div id="histone-modifications" class="section level4">
<h4><span class="header-section-number">2.2.3.2</span> Histone Modifications:</h4>
<p>Histones are proteins that constitute nucleosome. In eukaryotes, eight histones nucleosomes are wrapped around by DNA and build nucleoseome. They help super-coiling of DNA and inducing high-order structure called chromatin. In chromatin, DNA is either densely packed (called heterochromatin or closed chromatin), or it is loosely packed (called euchromatin or open chromatin) [60, 61]. Heterochromatin is thought to harbor inactive genes since DNA is densely packed and transcriptional machinery cannot access it. On the other hand, euchromatin is more accessible for transcriptional machinery and might therefore harbor active genes. Histones have long and unstructured N-terminal tails which can be covalently modified. The most studied modifications include acetylation, methylation and phosphorylation [60]. Using their tails, histones interact with neighboring nucleosomes and the modifications on the tail affect the nucleosomes affinity to bind DNA and therefore influence DNA packaging around nucleosomes. Different modifications on histones are used in different combinations to program the activity of the genes during differentiation. Histone modifications have a distinct nomenclature, for example: H3K4me3 means the lysine (K) on the 4th position of histone H3 is tri-methylated.</p>
<table>
<caption>Table 1 Histone modifications and their effects. If more than one histone modification has the same effect, they are separated by commas.</caption>
<thead>
<tr class="header">
<th align="left">Modifications</th>
<th align="left">Effect</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">H3K9ac</td>
<td align="left">Active promoters and enhancers</td>
</tr>
<tr class="even">
<td align="left">H3K14ac</td>
<td align="left">Active transcription</td>
</tr>
<tr class="odd">
<td align="left">H3K4me3/me2/me1,H3K27ac</td>
<td align="left">Active promoters and enhancers, H3K4me1 and H3K27ac is enhancer-specific</td>
</tr>
<tr class="even">
<td align="left">H3K36me3</td>
<td align="left">Active transcribed regions</td>
</tr>
<tr class="odd">
<td align="left">H3K27me3/me2/me1</td>
<td align="left">Silent promoters</td>
</tr>
<tr class="even">
<td align="left">H3K9me3/me2/me1</td>
<td align="left">Silent promoters</td>
</tr>
</tbody>
</table>
<p>Histone modifications are associated with a number of different transcription-related conditions; some of them are summarized in Table 1. Histone modifications can indicate where the regulatory regions are and they can also indicate activity of the genes. From a gene regulatory perspective, maybe the most important modifications are the ones associated with enhancers and promoters. Certain genes in mouse embryonic stem cells have both active H3K4me3 and inactive H3K27me3 modifications in their promoters. Surprisingly, most of these genes have high CpG content, are important for development and are shown to have paused RNA polymerase II [66, 67]. In addition, Heintzman et al. showed that H3K4me1 could predict tissue-specific active enhancers in human cells [68, 69]. The examples above demonstrate the capability of histone modifications in predicting regulatory potential.</p>
<p>Furthermore, certain proteins can influence chromatin structure by interacting with histones. Some of these proteins, like those of the Polycomb Group (PcG) and CTCF, are discussed above in the insulators and silencer sections. In vertebrates and insects, PcGs are responsible for maintaining the silent state of developmental genes, and trithorax group proteins (trxG) for maintaining their active state [70, 71]. PcGs and trxGs induce repressed or active states by catalyzing histone modifications or DNA methylation. Both the proteins bind PREs that can be on promoters or several kilobases away [30, 31, 71]. Another protein that induces histone modifications is CTCF. In b-globin locus, CTCF binding is shown to be associated with repressive H3K9/K27me2 modifications [72].</p>
</div>
</div>
<div id="regulation-by-ncrnas" class="section level3">
<h3><span class="header-section-number">2.2.4</span> Regulation by ncRNAs</h3>
<p>Recent years have witnessed an explosion in noncoding RNA (ncRNA)-related research. Many publications implicated ncRNAs as important regulatory elements (see [73-77] for examples). Plants and animals produce many different types of ncRNAs such as long non-coding RNAs (lncRNAs), small-interferring RNAs (siRNAs), microRNAs (miRNAs), promoter-associated RNAs (PARs) and small nucleolar RNAs (snoRNAs). lncRNAs are typically &gt;200 bp long, they are involved in epigenetic regulation by interacting with chromatin remodeling factors and they function in gene regulation [78]. siRNAs are short double-stranded RNAs which are involved in gene-regulation and transposon control, they silence their target genes by cooperating with Argonaute proteins [79]. miRNAs are short single-stranded RNA molecules that interact with their target genes by using their complementary sequence and mark them for quicker degradation [79]. PARs may regulate gene expression as well: they are ~18-200bp long ncRNAs originating from promoters of coding genes [80]. snoRNAs also shown to play roles in gene regulation, although they are mostly believed to guide ribosomal RNA modifications [81].</p>
</div>
</div>
<div id="high-throughput-experimental-methods-in-genomics" class="section level2">
<h2><span class="header-section-number">2.3</span> High-throughput experimental methods in genomics</h2>
<p>A number of high-throughput methods are regularly used in regulatory genomics. Their high quality and genome-wide yield makes these methods perfect for genome-scale studies. The following methods are some of the frequently used high-throughput methods to study gene regulation, and most of them rely heavily on computational methods for data analysis.</p>
<div id="end-sequencing" class="section level3">
<h3><span class="header-section-number">2.3.1</span> 5’ end sequencing:</h3>
<p>5’ end sequencing of mRNAs is one of the frequently used methods for promoter-profiling and genome-wide TSS identification. It is used to determine the precise sites for proximal and core promoters, and also to quantify expression coming from the promoters. These types of methods rely on accurate and complete cDNA synthesis from RNA followed by sequencing of 5’ ends of the transcripts and mapping to the reference genome. There are a number of research studies that used mainly CAGE technology to do 5’ end sequencing in various tissues in mouse and human [16, 17, 82]. In CAGE, a cap-trapper system is used to capture full-length RNAs while avoiding ribosomal and transfer RNAs. The system only captures capped mRNAs. The cap ensures mRNA stability needed when mRNA is translated, so mostly protein-coding mRNAs and functional ncRNAs will be captured by the protocol. Following cap-trapping, the 5’end tag of the transcript is sequenced and mapped to the genome.</p>
</div>
<div id="chromatin-immunoprecipitation-chip-followed-by-microarray-hybridization-or-sequencing" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Chromatin immunoprecipitation (ChIP) followed by microarray hybridization or sequencing:</h3>
<p>In ChIP, protein-DNA complexes are covalently bound to each-other using one of the various cross-linking techniques. Then the cells are lysed and DNA is broken into small pieces. This is followed by immunoprecipation where the specific antibody for the protein of interest is used to pull down the specific protein-DNA complexes. Then, protein and DNA are separated and the purified DNA can be hybridized to a variety of high-resolution microarrays (resulting in ChIP-chip method) or it can be sequenced (resulting in ChIP-Seq method) using new high-throughput sequencing techniques (see [53, 66, 68, 69] for examples). This technique can identify the genome-wide locations of various DNA-binding proteins including RNA polymerases, modified histones and transcription factors.</p>
</div>
<div id="dnase-hypersensitive-sites" class="section level3">
<h3><span class="header-section-number">2.3.3</span> DNAse hypersensitive sites:</h3>
<p>In eukaryotic genomes, nucleosome-free regions are thought to have open chromatin structures. Therefore they have better access to transcription factors. These nucleosome-free regions are traditionally detected with DNAse I hypersensitive site detection [83]. DNAse I is an endonuclease enzyme which cuts the DNA into smaller pieces. The genomic regionsthat are easily accessible for this enzyme are thought to be nucleosome-free regions. DNAse hypersensitive sites are often found in gene regulatory regions including promoters, enhancers and locus control regions [83, 84].</p>
</div>
<div id="bisulfite-sequencing" class="section level3">
<h3><span class="header-section-number">2.3.4</span> Bisulfite sequencing:</h3>
<p>This technique is used for detection of methylated DNA. Sodium bisulphite converts unmethylated cytosine to uracil while methylated cytosines are protected from conversion to uracil [85]. The converted DNA can be fragmented and sequenced using various high-throughput sequencing techniques. The results will be aligned to the genome using special alignments procedures and the genome-wide methylation patterns can be produced [86].</p>
</div>
<div id="chromosome-conformation-capture-based-technologies" class="section level3">
<h3><span class="header-section-number">2.3.5</span> Chromosome-conformation capture based technologies:</h3>
<p>Chromosome conformation capture (3C) is a high-throughput technique used to analyze the organization of chromosomes in a cell. 3C can tell which parts of the chromosomes are in close proximity [87]. Practically, this technique uncovers long-range regulatory interactions such as enhancer-promoter or silencer-promoter interactions. Recently, other variants of the 3C technique have been developed. 4C [88] and 5C [89] are the two main techniques that can discover interactions of a specific sequence with all the other regions in the genome. The higher throughput in 4C and 5C is achieved by using microarrays or high-throughput sequencing following regular 3C technique that captures DNA-DNA interactions. All of previously described 3C-based technologies can only reveal the genome wide-interactions of a single site (referred as “view-point” in the literature), however recently developed Hi-C technology can capture genome-wide many-to-many DNA interactions (although at a low resolution at present), enabling the creation of genome-wide contact maps can for the investigation of long-range interactions in DNA [90].</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="introduction-to-r.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="book_assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="book_assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="book_assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
