<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title></title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="">
  <meta name="generator" content="bookdown 0.0.71 and GitBook 2.6.7">

  <meta property="og:title" content="" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="genomic-intervals-and-r.html">
<link rel="next" href="dealing-with-high-throughput-sequencing-reads.html">

<script src="book_assets/jquery-2.2.3/jquery.min.js"></script>
<link href="book_assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html"><i class="fa fa-check"></i><b>1</b> Genomic Intervals and R</a></li>
<li class="chapter" data-level="2" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html"><i class="fa fa-check"></i><b>2</b> Operations on Genomic Intervals with GenomicRanges package</a><ul>
<li class="chapter" data-level="2.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>2.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="2.2" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>2.2</b> Getting genomic regions into R as GRanges objects</a><ul>
<li class="chapter" data-level="2.2.1" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#frequently-used-file-formats-and-how-to-read-them-into-r-as-a-table"><i class="fa fa-check"></i><b>2.2.1</b> Frequently used file formats and how to read them into R as a table</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="operations-on-genomic-intervals-with-genomicranges-package.html"><a href="operations-on-genomic-intervals-with-genomicranges-package.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>2.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="dealing-with-high-throughput-sequencing-reads.html"><a href="dealing-with-high-throughput-sequencing-reads.html"><i class="fa fa-check"></i><b>3</b> Dealing with high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="3.1" data-path="dealing-with-high-throughput-sequencing-reads.html"><a href="dealing-with-high-throughput-sequencing-reads.html#quality-check-on-sequencing-reads-and-mapping-reads-to-the-genome"><i class="fa fa-check"></i><b>3.1</b> Quality check on sequencing reads and mapping reads to the genome</a></li>
<li class="chapter" data-level="3.2" data-path="dealing-with-high-throughput-sequencing-reads.html"><a href="dealing-with-high-throughput-sequencing-reads.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>3.2</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html"><i class="fa fa-check"></i><b>4</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="4.1" data-path="dealing-with-continuous-scores-over-the-genome.html"><a href="dealing-with-continuous-scores-over-the-genome.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>4.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html"><i class="fa fa-check"></i><b>5</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="5.1" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#visualizing-intervals-in-locus-of-interest"><i class="fa fa-check"></i><b>5.1</b> Visualizing intervals in locus of interest</a></li>
<li class="chapter" data-level="5.2" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#summaries-of-genomic-intervals"><i class="fa fa-check"></i><b>5.2</b> Summaries of genomic intervals</a></li>
<li class="chapter" data-level="5.3" data-path="visualizing-and-summarizing-genomic-intervals.html"><a href="visualizing-and-summarizing-genomic-intervals.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>5.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="session-info.html"><a href="session-info.html"><i class="fa fa-check"></i><b>6</b> Session info</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="operations-on-genomic-intervals-with-genomicranges-package" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Operations on Genomic Intervals with GenomicRanges package</h1>
<p>The <a href="http://bioconductor.org">Bioconductor</a> project has a dedicated package called <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html"><strong>GenomicRanges</strong></a> to deal with genomic intervals. In this section, we will provide use cases involving operations on genomic intervals. The main reason we will stick to this package is that it provides tools to do overlap operations. However package requires that users operate on specific data types that are conceptually similar to a tabular data structure implemented in a way that makes overlapping and related operations easier. The main object we will be using is called <code>GRanges</code> object and we will also see some other related objects from the <code>GenomicRanges</code> package.</p>
<div id="how-to-create-and-manipulate-a-granges-object" class="section level2">
<h2><span class="header-section-number">2.1</span> How to create and manipulate a GRanges object</h2>
<p><code>GRanges</code> (from GenomicRanges package) is the main object that holds the genomic intervals and extra information about those intervals. Here we will show how to create one. Conceptually, it is similar to a data frame and some operations such as using <strong>[ ]</strong> notation to subset the table will work also on GRanges, but keep in mind that not everything that works for data frames will work on GRanges objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicRanges)
gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>,<span class="st">&quot;chr2&quot;</span>,<span class="st">&quot;chr2&quot;</span>),
           <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">150</span>,<span class="dv">200</span>),<span class="dt">end=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">300</span>)),
           <span class="dt">strand=</span><span class="kw">c</span>(<span class="st">&quot;+&quot;</span>,<span class="st">&quot;-&quot;</span>,<span class="st">&quot;-&quot;</span>)
)
gr</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames     ranges strand
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1 [ 50, 100]      +
##   [2]     chr2 [150, 200]      -
##   [3]     chr2 [200, 300]      -
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># subset like a data frame</span>
gr[<span class="dv">1</span>:<span class="dv">2</span>,]</code></pre></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames     ranges strand
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1 [ 50, 100]      +
##   [2]     chr2 [150, 200]      -
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>As you can see it looks a bit like a data frame. Also, note that the peculiar second argument “ranges” which basically contains start and end positions of the genomic intervals. However, you can not just give start and end positions you actually have to provide another object of IRanges. Do not let this confuse you, GRanges actually depends on another object that is very similar to itself called <code>IRanges</code> and you have to provide the “ranges” argument as an IRanges object. In its simplest for, an IRanges object can be constructed by providing start and end positions to IRanges() function. Think of it as something you just have to provide in order to construct the GRanges object.</p>
<p>GRanges can also contain other information about the genomic interval such as scores, names, etc. You can provide extra information at the time of the construction or you can add it later. Here is how you can do those:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>,<span class="st">&quot;chr2&quot;</span>,<span class="st">&quot;chr2&quot;</span>),
           <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">150</span>,<span class="dv">200</span>),<span class="dt">end=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">300</span>)),
           <span class="dt">names=</span><span class="kw">c</span>(<span class="st">&quot;id1&quot;</span>,<span class="st">&quot;id3&quot;</span>,<span class="st">&quot;id2&quot;</span>),
           <span class="dt">scores=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">90</span>,<span class="dv">50</span>)
)
<span class="co"># or add it later (replaces the existing meta data)</span>
<span class="kw">mcols</span>(gr)=<span class="kw">DataFrame</span>(<span class="dt">name2=</span><span class="kw">c</span>(<span class="st">&quot;pax6&quot;</span>,<span class="st">&quot;meis1&quot;</span>,<span class="st">&quot;zic4&quot;</span>),
                    <span class="dt">score2=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>))

gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>,<span class="st">&quot;chr2&quot;</span>,<span class="st">&quot;chr2&quot;</span>),
           <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span><span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">150</span>,<span class="dv">200</span>),<span class="dt">end=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">200</span>,<span class="dv">300</span>)),
           <span class="dt">names=</span><span class="kw">c</span>(<span class="st">&quot;id1&quot;</span>,<span class="st">&quot;id3&quot;</span>,<span class="st">&quot;id2&quot;</span>),
           <span class="dt">scores=</span><span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">90</span>,<span class="dv">50</span>)
)

<span class="co"># or appends to existing meta data</span>
<span class="kw">mcols</span>(gr)=<span class="kw">cbind</span>(<span class="kw">mcols</span>(gr),
                          <span class="kw">DataFrame</span>(<span class="dt">name2=</span><span class="kw">c</span>(<span class="st">&quot;pax6&quot;</span>,<span class="st">&quot;meis1&quot;</span>,<span class="st">&quot;zic4&quot;</span>)) )
gr</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames     ranges strand |       names    scores       name2
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
##   [1]     chr1 [ 50, 100]      * |         id1       100        pax6
##   [2]     chr2 [150, 200]      * |         id3        90       meis1
##   [3]     chr2 [200, 300]      * |         id2        50        zic4
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># elementMetadata() and values() do the same things</span>
<span class="kw">elementMetadata</span>(gr)</code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##         names    scores       name2
##   &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
## 1         id1       100        pax6
## 2         id3        90       meis1
## 3         id2        50        zic4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">values</span>(gr)</code></pre></div>
<pre><code>## DataFrame with 3 rows and 3 columns
##         names    scores       name2
##   &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
## 1         id1       100        pax6
## 2         id3        90       meis1
## 3         id2        50        zic4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># you may also add metadata using the $ operator, as for data frames</span>
gr$name3 =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;C&quot;</span>, <span class="st">&quot;B&quot;</span>)
gr</code></pre></div>
<pre><code>## GRanges object with 3 ranges and 4 metadata columns:
##       seqnames     ranges strand |       names    scores       name2
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt; &lt;character&gt;
##   [1]     chr1 [ 50, 100]      * |         id1       100        pax6
##   [2]     chr2 [150, 200]      * |         id3        90       meis1
##   [3]     chr2 [200, 300]      * |         id2        50        zic4
##             name3
##       &lt;character&gt;
##   [1]           A
##   [2]           C
##   [3]           B
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="getting-genomic-regions-into-r-as-granges-objects" class="section level2">
<h2><span class="header-section-number">2.2</span> Getting genomic regions into R as GRanges objects</h2>
<p>There are multiple ways you can read in your genomic features into R and create a GRanges object. Most genomic interval data comes as a tabular format that has the basic information about the location of the interval and some other information. We already showed how to read BED files as data frame. Now we will show how to convert it to GRanges object. This is one way of doing it, but there are more convenient ways described further in the text.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read CpGi data set</span>
cpgi.df =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;GenomicIntervals_data/data/cpgi.hg19.chr21.bed&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>,
                     <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>) 
<span class="co"># remove chr names with &quot;_&quot;</span>
cpgi.df =cpgi.df [<span class="kw">grep</span>(<span class="st">&quot;_&quot;</span>,cpgi.df[,<span class="dv">1</span>],<span class="dt">invert=</span><span class="ot">TRUE</span>),]

cpgi.gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span>cpgi.df[,<span class="dv">1</span>],
                <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span>cpgi.df[,<span class="dv">2</span>],
                              <span class="dt">end=</span>cpgi.df[,<span class="dv">3</span>]))</code></pre></div>
<p>You may need to do some pre-processing before/after reading in the BED file. Below is an example of getting transcription start sites from BED files containing refseq transcript locations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read refseq file</span>
ref.df =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;GenomicIntervals_data/data/refseq.hg19.chr21.bed&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>,
                     <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>) 
ref.gr=<span class="kw">GRanges</span>(<span class="dt">seqnames=</span>ref.df[,<span class="dv">1</span>],
               <span class="dt">ranges=</span><span class="kw">IRanges</span>(<span class="dt">start=</span>ref.df[,<span class="dv">2</span>],
                              <span class="dt">end=</span>ref.df[,<span class="dv">3</span>]),
               <span class="dt">strand=</span>ref.df[,<span class="dv">6</span>],<span class="dt">name=</span>ref.df[,<span class="dv">4</span>])
<span class="co"># get TSS</span>
tss.gr=ref.gr
<span class="co"># end of the + strand genes must be equalized to start pos</span>
<span class="kw">end</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;+&quot;</span>,])  =<span class="kw">start</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;+&quot;</span>,])
<span class="co"># startof the - strand genes must be equalized to end pos</span>
<span class="kw">start</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;-&quot;</span>,])=<span class="kw">end</span>(tss.gr[<span class="kw">strand</span>(tss.gr)==<span class="st">&quot;-&quot;</span>,])
<span class="co"># remove duplicated TSSes ie alternative transcripts</span>
<span class="co"># this keeps the first instance and removes duplicates</span>
tss.gr=tss.gr[!<span class="kw">duplicated</span>(tss.gr),]</code></pre></div>
<p>Another way of doing this is from a BED file is to use <code>readTranscriptfeatures()</code> function from the <code>genomation</code> package.</p>
<p>Reading the genomic features as text files and converting to GRanges is not the only way to create GRanges object. With the help of the <a href="http://www.bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a> package we can directly import BED files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(rtracklayer)
<span class="kw">import.bed</span>(<span class="st">&quot;GenomicIntervals_data/data/refseq.hg19.chr21.bed&quot;</span>)</code></pre></div>
<p>Next, we will show how to use other methods to automatically obtain the data in GRanges format from online databases. But you will not be able to use these methods for every data set so it is good to now how to read data from flat files as well. We will use rtracklayer package to download data from <a href="http://genome-euro.ucsc.edu/cgi-bin/hgGateway?hgsid=197855328_moRAG0Dqe1qlcOEtNDAt09E1e6ab">UCSC browser</a>. We will download CpG islands as GRanges objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(rtracklayer)
session &lt;-<span class="st"> </span><span class="kw">browserSession</span>(<span class="st">&quot;UCSC&quot;</span>,<span class="dt">url =</span> <span class="st">&#39;http://genome-euro.ucsc.edu/cgi-bin/&#39;</span>)
<span class="kw">genome</span>(session) &lt;-<span class="st"> &quot;mm9&quot;</span>
## choose CpG island track on chr12
query &lt;-<span class="st"> </span><span class="kw">ucscTableQuery</span>(session, <span class="dt">track=</span><span class="st">&quot;CpG Islands&quot;</span>,<span class="dt">table=</span><span class="st">&quot;cpgIslandExt&quot;</span>,
        <span class="dt">range=</span><span class="kw">GRangesForUCSCGenome</span>(<span class="st">&quot;mm9&quot;</span>, <span class="st">&quot;chr12&quot;</span>))
## get the GRanges object for the track
<span class="kw">track</span>(query)</code></pre></div>
<div id="frequently-used-file-formats-and-how-to-read-them-into-r-as-a-table" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Frequently used file formats and how to read them into R as a table</h3>
<p>There are multiple file formats in genomics but some of them you will see more frequently than others. We already mentioned some of them. Here is a list of files and functions to read them into R as <code>GRanges</code> objects or something coercible to GRanges objects.</p>
<p><strong>BED</strong>: These are used and popularized by UCSC browser, and can hold a variety of information including exon/intron structure of transcripts in a single line.</p>
<ul>
<li><code>genomation::readBed()</code></li>
<li><code>genomation::readTranscriptFeatures()</code> good for getting intron/exon/promoters from BED12 files</li>
<li><code>rtracklayer::import.bed()</code></li>
</ul>
<p><strong>GFF</strong>: GFF format is a tabular text format for genomic features similar to BED. However, it is a more flexible format than BED, which makes it harder to parse at times. Many gene annotation files are in this format.</p>
<ul>
<li><code>genomation::gffToGranges()</code></li>
<li><code>rtracklayer::impot.gff()</code></li>
</ul>
<p><strong>BAM</strong>: BAM format is compressed and indexed tabular file format designed for sequencing reads.</p>
<ul>
<li><code>GenomicAlignments::readGAlignments</code></li>
<li><code>Rsamtools::scanBam</code> returns a data frame with columns from SAM/BAM file.</li>
</ul>
<p><strong>BigWig</strong>: This is used to for storing scores associated with genomic intervals. It is an indexed format. Similar to BAM, this makes it easier to query and only necessary portions of the file could be loaded into memory.</p>
<ul>
<li><code>rtracklayer::import.bw()</code></li>
</ul>
<p><strong>Generic Text files</strong>: This represents any text file with the minimal information of chromosome, start and end coordinates.</p>
<ul>
<li><code>genomation::readGeneric()</code></li>
</ul>
<p><strong>Tabix/Bcf</strong>: These are tabular file formats indexed and compressed similar to BAM. The following functions return lists rather than tabular data structures. These formats are mostly used to store genomic variation data such as SNPs and indels.</p>
<ul>
<li><code>Rsamtools::scanTabix</code></li>
<li><code>Rsamtools::scanBcf</code></li>
</ul>
</div>
</div>
<div id="finding-regions-that-dodo-not-overlap-with-another-set-of-regions" class="section level2">
<h2><span class="header-section-number">2.3</span> Finding regions that do/do not overlap with another set of regions</h2>
<p>This is one of the most common tasks in genomics. Usually, you have a set of regions that you are interested in and you want to see if they overlap with another set of regions or see how many of them overlap. A good example is transcription factor binding sites determined by <a href="http://en.wikipedia.org/wiki/ChIP-sequencing">ChIP-seq</a> experiments. In these types of experiments and followed analysis, one usually ends up with genomic regions that are bound by transcription factors. One of the standard next questions would be to annotate binding sites with genomic annotations such as promoter,exon,intron and/or CpG islands. Below is a demonstration of how transcription factor binding sites can be annotated using CpG islands. First, we will get the subset of binding sites that overlap with the CpG islands. In this case, binding sites are ChIP-seq peaks.</p>
<p>We can find the subset of peaks that overlap with the CpG islands using the <strong>subsetByoverlaps()</strong> function. You will also see another way of converting <strong>data frames</strong> to <strong>GRanges</strong>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(genomation)
pk1.gr=<span class="kw">readBroadPeak</span>(<span class="st">&quot;GenomicIntervals_data/data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz&quot;</span>)

pk1.gr=pk1.gr[<span class="kw">seqnames</span>(pk1.gr)==<span class="st">&quot;chr21&quot;</span>,]
<span class="co"># get the peaks that overlap with CpG</span>
<span class="co"># islands</span>
<span class="kw">subsetByOverlaps</span>(pk1.gr,cpgi.gr)</code></pre></div>
<pre><code>## GRanges object with 44 ranges and 5 metadata columns:
##        seqnames               ranges strand |        name     score
##           &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;
##    [1]    chr21 [ 9825359,  9826582]      * |   peak14562        56
##    [2]    chr21 [ 9968468,  9968984]      * |   peak14593       947
##    [3]    chr21 [15755367, 15755956]      * |   peak14828        90
##    [4]    chr21 [19191578, 19192525]      * |   peak14840       290
##    [5]    chr21 [26979618, 26980048]      * |   peak14854        32
##    ...      ...                  ...    ... .         ...       ...
##   [40]    chr21 [46237463, 46237809]      * |   peak15034        32
##   [41]    chr21 [46707701, 46708084]      * |   peak15037        67
##   [42]    chr21 [46961551, 46961875]      * |   peak15039        38
##   [43]    chr21 [47743586, 47744125]      * |   peak15050       353
##   [44]    chr21 [47878411, 47878891]      * |   peak15052       104
##        signalValue    pvalue    qvalue
##          &lt;numeric&gt; &lt;integer&gt; &lt;integer&gt;
##    [1]      183.11        -1        -1
##    [2]     3064.92        -1        -1
##    [3]      291.90        -1        -1
##    [4]      940.03        -1        -1
##    [5]      104.67        -1        -1
##    ...         ...       ...       ...
##   [40]      106.36        -1        -1
##   [41]      217.02        -1        -1
##   [42]      124.31        -1        -1
##   [43]     1141.58        -1        -1
##   [44]      338.78        -1        -1
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
<p>For each CpG island, we can count the number of peaks that overlap with a given CpG island with countOverlaps().</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#count the peaks that</span>
<span class="co"># overlap with CpG islands</span>
counts=<span class="kw">countOverlaps</span>(pk1.gr,cpgi.gr)
<span class="kw">head</span>(counts)</code></pre></div>
<pre><code>## [1] 0 0 0 0 0 0</code></pre>
<p>The <code>findOverlaps()</code> function can be used to see one-to-one overlaps between peaks and CpG islands. It returns a matrix showing which peak overlaps with which CpGi island.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">findOverlaps</span>(pk1.gr,cpgi.gr)</code></pre></div>
<pre><code>## Hits object with 45 hits and 0 metadata columns:
##        queryHits subjectHits
##        &lt;integer&gt;   &lt;integer&gt;
##    [1]       123           1
##    [2]       154           3
##    [3]       389           8
##    [4]       401          13
##    [5]       415          16
##    ...       ...         ...
##   [41]       595         155
##   [42]       598         166
##   [43]       600         176
##   [44]       611         192
##   [45]       613         200
##   -------
##   queryLength: 620 / subjectLength: 205</code></pre>
<p>Another interesting thing would be to look at the distances to nearest CpG islands for each peak. In addition, just finding the nearest CpG island could also be interesting. Often times, you will need to find nearest TSS or gene to your regions of interest, and the code below is handy for doing that.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># find nearest CpGi to each TSS</span>
n.ind=<span class="kw">nearest</span>(pk1.gr,cpgi.gr)
<span class="co"># get distance to nearest</span>
dists=<span class="kw">distanceToNearest</span>(pk1.gr,cpgi.gr,<span class="dt">select=</span><span class="st">&quot;arbitrary&quot;</span>)
dists</code></pre></div>
<pre><code>## Hits object with 620 hits and 1 metadata column:
##         queryHits subjectHits |  distance
##         &lt;integer&gt;   &lt;integer&gt; | &lt;integer&gt;
##     [1]         1           1 |    384188
##     [2]         2           1 |    382968
##     [3]         3           1 |    381052
##     [4]         4           1 |    379311
##     [5]         5           1 |    376978
##     ...       ...         ... .       ...
##   [616]       616         205 |     26211
##   [617]       617         205 |     27401
##   [618]       618         205 |     30467
##   [619]       619         205 |     31610
##   [620]       620         205 |     34089
##   -------
##   queryLength: 620 / subjectLength: 205</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># histogram of the distances to nearest TSS</span>
dist2plot=<span class="kw">mcols</span>(dists)[,<span class="dv">1</span>]
<span class="kw">hist</span>(<span class="kw">log10</span>(dist2plot),<span class="dt">xlab=</span><span class="st">&quot;log10(dist to nearest TSS)&quot;</span>,
     <span class="dt">main=</span><span class="st">&quot;Distances&quot;</span>)</code></pre></div>
<div class="figure"><span id="fig:findNearest"></span>
<img src="04-genomicIntervals_files/figure-html/findNearest-1.png" alt="Figure  2: histogram of distances" width="672" />
<p class="caption">
Figure 2.1: Figure 2: histogram of distances
</p>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="genomic-intervals-and-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="dealing-with-high-throughput-sequencing-reads.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="book_assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="book_assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="book_assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"search": false
});
});
</script>

</body>

</html>
