<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Computational Genomics with R</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="book files for computatiation genomcis with R">
  <meta name="generator" content="bookdown 0.0.71 and GitBook 2.6.7">

  <meta property="og:title" content="Computational Genomics with R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="book files for computatiation genomcis with R" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Computational Genomics with R" />
  
  <meta name="twitter:description" content="book files for computatiation genomcis with R" />
  

<meta name="author" content="Altuna Akalin">

<meta name="date" content="2016-05-09">

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction-to-genomics.html">
<link rel="next" href="statistics-for-genomics.html">

<script src="book_assets/jquery-2.2.3/jquery.min.js"></script>
<link href="book_assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="book_assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="introduction-to-genomics.html"><a href="introduction-to-genomics.html"><i class="fa fa-check"></i><b>1</b> Introduction to Genomics</a></li>
<li class="chapter" data-level="2" data-path="introduction-to-r.html"><a href="introduction-to-r.html"><i class="fa fa-check"></i><b>2</b> Introduction to R</a><ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#the-setup"><i class="fa fa-check"></i><b>2.1</b> The Setup</a><ul>
<li class="chapter" data-level="2.1.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#installing-packages"><i class="fa fa-check"></i><b>2.1.1</b> Installing packages</a></li>
<li class="chapter" data-level="2.1.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#installing-packages-in-custom-locations"><i class="fa fa-check"></i><b>2.1.2</b> Installing packages in custom locations</a></li>
<li class="chapter" data-level="2.1.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#getting-help-on-functions-and-packages"><i class="fa fa-check"></i><b>2.1.3</b> Getting help on functions and packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#computations-in-r"><i class="fa fa-check"></i><b>2.2</b> Computations in R</a></li>
<li class="chapter" data-level="2.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-structures"><i class="fa fa-check"></i><b>2.3</b> Data structures</a><ul>
<li class="chapter" data-level="2.3.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#vectors"><i class="fa fa-check"></i><b>2.3.1</b> Vectors</a></li>
<li class="chapter" data-level="2.3.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#matrices"><i class="fa fa-check"></i><b>2.3.2</b> Matrices</a></li>
<li class="chapter" data-level="2.3.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-frames"><i class="fa fa-check"></i><b>2.3.3</b> Data Frames</a></li>
<li class="chapter" data-level="2.3.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#lists"><i class="fa fa-check"></i><b>2.3.4</b> Lists</a></li>
<li class="chapter" data-level="2.3.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#factors"><i class="fa fa-check"></i><b>2.3.5</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-types"><i class="fa fa-check"></i><b>2.4</b> Data types</a></li>
<li class="chapter" data-level="2.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#reading-and-writing-data"><i class="fa fa-check"></i><b>2.5</b> Reading and writing data</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-r.html"><a href="introduction-to-r.html#plotting-in-r"><i class="fa fa-check"></i><b>2.6</b> Plotting in R</a></li>
<li class="chapter" data-level="2.7" data-path="introduction-to-r.html"><a href="introduction-to-r.html#functions-and-control-structures-for-ifelse-etc."><i class="fa fa-check"></i><b>2.7</b> Functions and control structures (for, if/else etc.)</a><ul>
<li class="chapter" data-level="2.7.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#user-defined-functions"><i class="fa fa-check"></i><b>2.7.1</b> User defined functions</a></li>
<li class="chapter" data-level="2.7.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#loops-and-looping-structures-in-r"><i class="fa fa-check"></i><b>2.7.2</b> Loops and looping structures in R</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="introduction-to-r.html"><a href="introduction-to-r.html#session-info"><i class="fa fa-check"></i><b>2.8</b> Session info</a></li>
<li class="chapter" data-level="2.9" data-path="introduction-to-r.html"><a href="introduction-to-r.html#exercises"><i class="fa fa-check"></i><b>2.9</b> Exercises</a><ul>
<li class="chapter" data-level="2.9.1" data-path="introduction-to-r.html"><a href="introduction-to-r.html#computations-in-r-1"><i class="fa fa-check"></i><b>2.9.1</b> Computations in R</a></li>
<li class="chapter" data-level="2.9.2" data-path="introduction-to-r.html"><a href="introduction-to-r.html#data-structures-in-r"><i class="fa fa-check"></i><b>2.9.2</b> Data structures in R</a></li>
<li class="chapter" data-level="2.9.3" data-path="introduction-to-r.html"><a href="introduction-to-r.html#reading-in-and-writing-data-out-in-r"><i class="fa fa-check"></i><b>2.9.3</b> Reading in and writing data out in R</a></li>
<li class="chapter" data-level="2.9.4" data-path="introduction-to-r.html"><a href="introduction-to-r.html#plotting-in-r-1"><i class="fa fa-check"></i><b>2.9.4</b> Plotting in R</a></li>
<li class="chapter" data-level="2.9.5" data-path="introduction-to-r.html"><a href="introduction-to-r.html#functions-and-control-structures-for-ifelse-etc.-1"><i class="fa fa-check"></i><b>2.9.5</b> Functions and control structures (for, if/else etc.)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html"><i class="fa fa-check"></i><b>3</b> Statistics for Genomics</a><ul>
<li class="chapter" data-level="3.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions"><i class="fa fa-check"></i><b>3.1</b> How to summarize collection of data points: The idea behind statistical distributions</a><ul>
<li class="chapter" data-level="3.1.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#describing-the-central-tendency-mean-and-median"><i class="fa fa-check"></i><b>3.1.1</b> Describing the central tendency: mean and median</a></li>
<li class="chapter" data-level="3.1.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#describing-the-spread-measurements-of-variation"><i class="fa fa-check"></i><b>3.1.2</b> Describing the spread: measurements of variation</a></li>
<li class="chapter" data-level="3.1.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#precision-of-estimates-confidence-intervals"><i class="fa fa-check"></i><b>3.1.3</b> Precision of estimates: Confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-test-for-differences-in-samples"><i class="fa fa-check"></i><b>3.2</b> How to test for differences in samples</a><ul>
<li class="chapter" data-level="3.2.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#randomization-based-testing-for-difference-of-the-means"><i class="fa fa-check"></i><b>3.2.1</b> randomization based testing for difference of the means</a></li>
<li class="chapter" data-level="3.2.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#using-t-test-for-difference-of-the-means-between-two-samples"><i class="fa fa-check"></i><b>3.2.2</b> Using t-test for difference of the means between two samples</a></li>
<li class="chapter" data-level="3.2.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#multiple-testing-correction"><i class="fa fa-check"></i><b>3.2.3</b> multiple testing correction</a></li>
<li class="chapter" data-level="3.2.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#moderated-t-tests-using-information-from-multiple-comparisons"><i class="fa fa-check"></i><b>3.2.4</b> moderated t-tests: using information from multiple comparisons</a></li>
<li class="chapter" data-level="3.2.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#want-to-know-more"><i class="fa fa-check"></i><b>3.2.5</b> Want to know more…</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#relationship-between-variables-linear-models-and-correlation"><i class="fa fa-check"></i><b>3.3</b> Relationship between variables: linear models and correlation</a><ul>
<li class="chapter" data-level="3.3.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-fit-a-line"><i class="fa fa-check"></i><b>3.3.1</b> How to fit a line</a></li>
<li class="chapter" data-level="3.3.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-estimate-the-error-of-the-coefficients"><i class="fa fa-check"></i><b>3.3.2</b> How to estimate the error of the coefficients</a></li>
<li class="chapter" data-level="3.3.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#accuracy-of-the-model"><i class="fa fa-check"></i><b>3.3.3</b> Accuracy of the model</a></li>
<li class="chapter" data-level="3.3.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#regression-with-categorical-variables"><i class="fa fa-check"></i><b>3.3.4</b> Regression with categorical variables</a></li>
<li class="chapter" data-level="3.3.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#regression-pitfalls"><i class="fa fa-check"></i><b>3.3.5</b> Regression pitfalls</a></li>
<li class="chapter" data-level="3.3.6" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#want-to-know-more-1"><i class="fa fa-check"></i><b>3.3.6</b> Want to know more…</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#roadmap-for-future"><i class="fa fa-check"></i><b>3.4</b> Roadmap for future</a><ul>
<li class="chapter" data-level="3.4.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#clustering"><i class="fa fa-check"></i><b>3.4.1</b> Clustering</a></li>
<li class="chapter" data-level="3.4.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#dimension-reduction-with-pca"><i class="fa fa-check"></i><b>3.4.2</b> Dimension reduction with PCA</a></li>
<li class="chapter" data-level="3.4.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#classification"><i class="fa fa-check"></i><b>3.4.3</b> Classification</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#exercises-1"><i class="fa fa-check"></i><b>3.5</b> Exercises</a><ul>
<li class="chapter" data-level="3.5.1" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-summarize-collection-of-data-points-the-idea-behind-statistical-distributions-1"><i class="fa fa-check"></i><b>3.5.1</b> How to summarize collection of data points: The idea behind statistical distributions</a></li>
<li class="chapter" data-level="3.5.2" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#how-to-test-for-differences-in-samples-1"><i class="fa fa-check"></i><b>3.5.2</b> How to test for differences in samples</a></li>
<li class="chapter" data-level="3.5.3" data-path="statistics-for-genomics.html"><a href="statistics-for-genomics.html#relationship-between-variables-linear-models-and-correlation-1"><i class="fa fa-check"></i><b>3.5.3</b> Relationship between variables: linear models and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html"><i class="fa fa-check"></i><b>4</b> Genomic Intervals and R</a><ul>
<li class="chapter" data-level="4.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#operations-on-genomic-intervals-with-genomicranges-package"><i class="fa fa-check"></i><b>4.1</b> Operations on Genomic Intervals with GenomicRanges package</a><ul>
<li class="chapter" data-level="4.1.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#how-to-create-and-manipulate-a-granges-object"><i class="fa fa-check"></i><b>4.1.1</b> How to create and manipulate a GRanges object</a></li>
<li class="chapter" data-level="4.1.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#getting-genomic-regions-into-r-as-granges-objects"><i class="fa fa-check"></i><b>4.1.2</b> Getting genomic regions into R as GRanges objects</a></li>
<li class="chapter" data-level="4.1.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#finding-regions-that-dodo-not-overlap-with-another-set-of-regions"><i class="fa fa-check"></i><b>4.1.3</b> Finding regions that do/do not overlap with another set of regions</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-high-throughput-sequencing-reads"><i class="fa fa-check"></i><b>4.2</b> Dealing with high-throughput sequencing reads</a><ul>
<li class="chapter" data-level="4.2.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#quality-check-on-sequencing-reads-and-mapping-reads-to-the-genome"><i class="fa fa-check"></i><b>4.2.1</b> Quality check on sequencing reads and mapping reads to the genome</a></li>
<li class="chapter" data-level="4.2.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#counting-mapped-reads-for-a-set-of-regions"><i class="fa fa-check"></i><b>4.2.2</b> Counting mapped reads for a set of regions</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-continuous-scores-over-the-genome"><i class="fa fa-check"></i><b>4.3</b> Dealing with continuous scores over the genome</a><ul>
<li class="chapter" data-level="4.3.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#extracting-subsections-of-rle-and-rlelist-objects"><i class="fa fa-check"></i><b>4.3.1</b> Extracting subsections of Rle and RleList objects</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-and-summarizing-genomic-intervals"><i class="fa fa-check"></i><b>4.4</b> Visualizing and summarizing genomic intervals</a><ul>
<li class="chapter" data-level="4.4.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-intervals-in-locus-of-interest"><i class="fa fa-check"></i><b>4.4.1</b> Visualizing intervals in locus of interest</a></li>
<li class="chapter" data-level="4.4.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#summaries-of-genomic-intervals"><i class="fa fa-check"></i><b>4.4.2</b> Summaries of genomic intervals</a></li>
<li class="chapter" data-level="4.4.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#making-karyograms-and-circos-plots"><i class="fa fa-check"></i><b>4.4.3</b> Making karyograms and circos plots</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#session-info-1"><i class="fa fa-check"></i><b>4.5</b> Session info</a></li>
<li class="chapter" data-level="4.6" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#exercises-2"><i class="fa fa-check"></i><b>4.6</b> Exercises</a><ul>
<li class="chapter" data-level="4.6.1" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#the-setup-1"><i class="fa fa-check"></i><b>4.6.1</b> The setup</a></li>
<li class="chapter" data-level="4.6.2" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#operations-on-genomic-intervals-with-genomicranges-package-1"><i class="fa fa-check"></i><b>4.6.2</b> Operations on Genomic Intervals with GenomicRanges package</a></li>
<li class="chapter" data-level="4.6.3" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-mapped-high-throughput-sequencing-reads"><i class="fa fa-check"></i><b>4.6.3</b> Dealing with mapped high-throughput sequencing reads</a></li>
<li class="chapter" data-level="4.6.4" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#dealing-with-contiguous-scores-over-the-genome"><i class="fa fa-check"></i><b>4.6.4</b> Dealing with contiguous scores over the genome</a></li>
<li class="chapter" data-level="4.6.5" data-path="genomic-intervals-and-r.html"><a href="genomic-intervals-and-r.html#visualizing-and-summarizing-genomic-intervals-1"><i class="fa fa-check"></i><b>4.6.5</b> Visualizing and summarizing genomic intervals</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Computational Genomics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-r" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Introduction to R</h1>
<p>R is a free statistical programming language that is popular among researchers and data miners to build software and analyze data1. In next sections, we will first get you started with the setup of environment for using R and then introduce some basic R operations and data structures that will be good to know if you do not have prior experience with R. If you need more in depth R introduction you will want to check out some beginner level books and online tutorials. This website has a bunch of resources listed: <a href="">http: //www.introductoryr.co.uk/R_Resources_for_Beginners.html</a></p>
<div id="the-setup" class="section level2">
<h2><span class="header-section-number">2.1</span> The Setup</h2>
<p>Download and install R <a href="http://cran.r-project.org/" class="uri">http://cran.r-project.org/</a> and RStudio <a href="http://www.rstudio.com/" class="uri">http://www.rstudio.com/</a> if you do not have them already. Rstudio is optional but it is a great tool if you are just starting to learn R. You will need specific data sets to run the codes in this document. Download the data.zip[URL to come] and extract it to your directory of choice. The folder name should be “data” and your R working directory should be level above the data folder. That means in your R console, when you type “dir(“data”)” you should be able to see the contents of the data folder. You can change your working directory by <em>setwd()</em> command and get your current working directory with <em>getwd()</em> command in R. In RStudio, you can click on the top menu and change the location of your working directory via user interface.</p>
<div id="installing-packages" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Installing packages</h3>
<p>R packages are add-ons to base R that help you achieve additional tasks that are not directly supported by base R. It is by the action of these extra functionality that R excels as a tool for computational genomics. Bioconductor project (<a href="http://bioconductor.org/" class="uri">http://bioconductor.org/</a>) is a dedicated package repository for computational biology related packages. However main package repository of R, called CRAN, has also computational biology related packages. In addition, R-Forge(<a href="http://r-forge.r-project.org/" class="uri">http://r-forge.r-project.org/</a>), GitHub(<a href="https://github" class="uri">https://github</a>. com/), and googlecode(<a href="http://code.google.com" class="uri">http://code.google.com</a>) are other locations where R packages might be hosted. You can install CRAN packages using install.packages(). (# is the comment character in R)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install package named &quot;randomForests&quot; from CRAN</span>
<span class="kw">install.packages</span>(<span class="st">&quot;randomForests&quot;</span>)</code></pre></div>
<p>You can install bioconductor packages with a specific installer script</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the installer package</span>
<span class="kw">source</span>(<span class="st">&quot;http://bioconductor.org/biocLite.R&quot;</span>)
<span class="co"># install bioconductor package &quot;rtracklayer&quot;</span>
<span class="kw">biocLite</span>(<span class="st">&quot;rtracklayer&quot;</span>)</code></pre></div>
<p>You can install packages from github using <em>install_github()</em> function from <strong>devtools</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;hadley/stringr&quot;</span>)</code></pre></div>
<p>Another way to install packages are from the source.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the source file</span>
<span class="kw">download.file</span>(<span class="st">&quot;http://goo.gl/3pvHYI&quot;</span>,
               <span class="dt">destfile=</span><span class="st">&quot;methylKit_0.5.7.tar.gz&quot;</span>)
<span class="co"># install the package from the source file</span>
<span class="kw">install.packages</span>(<span class="st">&quot;methylKit_0.5.7.tar.gz&quot;</span>,
                 <span class="dt">repos=</span><span class="ot">NULL</span>,<span class="dt">type=</span><span class="st">&quot;source&quot;</span>)
<span class="co"># delete the source file</span>
<span class="kw">unlink</span>(<span class="st">&quot;methylKit_0.5.7.tar.gz&quot;</span>)</code></pre></div>
<p>You can also update CRAN and Bioconductor packages.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># updating CRAN packages</span>
<span class="kw">update.packages</span>()

<span class="co"># updating bioconductor packages</span>
<span class="kw">source</span>(<span class="st">&quot;http://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;BiocUpgrade&quot;</span>)</code></pre></div>
</div>
<div id="installing-packages-in-custom-locations" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Installing packages in custom locations</h3>
<p>If you will be using R on servers or computing clusters rather than your personal computer it is unlikely that you will have administrator access to install packages. In that case, you can install packages in custom locations by telling R where to look for additional packages. This is done by setting up an .Renviron file in your home directory and add the following line:</p>
<pre><code>R_LIBS=~/Rlibs</code></pre>
<p>This tells R that “Rlibs” directory at your home directory will be the first choice of locations to look for packages and install packages (The directory name and location is up to you above is just an example). You should go and create that directory now. After that, start a fresh R session and start installing packages. From now on, packages will be installed to your local directory where you have read-write access.</p>
</div>
<div id="getting-help-on-functions-and-packages" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Getting help on functions and packages</h3>
<p>You can get help on functions by help() and help.search() functions. You can list the functions in a package with ls() function</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(MASS)
<span class="kw">ls</span>(<span class="st">&quot;package:MASS&quot;</span>) <span class="co"># functions in the package</span>
<span class="kw">ls</span>() <span class="co"># objects in your R enviroment</span>
<span class="co"># get help on hist() function</span>
?hist
<span class="kw">help</span>(<span class="st">&quot;hist&quot;</span>)
<span class="co"># search the word &quot;hist&quot; in help pages</span>
<span class="kw">help.search</span>(<span class="st">&quot;hist&quot;</span>)
??hist</code></pre></div>
<div id="more-help-needed" class="section level4">
<h4><span class="header-section-number">2.1.3.1</span> More help needed?</h4>
<p>In addition, check package vignettes for help and practical understanding of the functions. All Bionconductor packages have vignettes that walk you through example analysis. Google search will always be helpful as well, there are many blogs and web pages that have posts about R. R-help, Stackoverflow and R-bloggers are usually source of good and reliable information.</p>
</div>
</div>
</div>
<div id="computations-in-r" class="section level2">
<h2><span class="header-section-number">2.2</span> Computations in R</h2>
<p>R can be used as an ordinary calculator, some say it is an over-grown calculator. Here are some examples. Remember <strong>#</strong> is the comment character. The comments give details about the operations in case they are not clear.</p>
<pre class="r{}"><code>2 + 3 * 5       # Note the order of operations.
log(10)        # Natural logarithm with base e
5^2            # 5 raised to the second power
3/2            # Division
sqrt(16)      # Square root
abs(3-7)      # Absolute value of 3-7
pi             # The number
exp(2)        # exponential function
# This is a comment line

</code></pre>
</div>
<div id="data-structures" class="section level2">
<h2><span class="header-section-number">2.3</span> Data structures</h2>
<p>R has multiple data structures. If you are familiar with excel you can think of data structures as building blocks of a table and the table itself, and a table is similar to a sheet in excel. Most of the time you will deal with tabular data sets, you will manipulate them, take sub-sections of them. It is essential to know what are the common data structures in R and how they can be used. R deals with named data structures, this means you can give names to data structures and manipulate or operate on them using those names.</p>
<div id="vectors" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Vectors</h3>
<p>Vectors are one the core R data structures. It is basically a list of elements of the same type (numeric,character or logical). Later you will see that every column of a table will be represented as a vector. R handles vectors easily and intuitively. You can create vectors with c() function, however that is not the only way. The operations on vectors will propagate to all the elements of the vectors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x&lt;-<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">10</span>,<span class="dv">5</span>)    <span class="co">#create a vector named x with 5 components</span>
x =<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">10</span>,<span class="dv">5</span>)  
x</code></pre></div>
<pre><code>## [1]  1  3  2 10  5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y&lt;-<span class="dv">1</span>:<span class="dv">5</span>              <span class="co">#create a vector of consecutive integers y</span>
y<span class="dv">+2</span>                 <span class="co">#scalar addition</span></code></pre></div>
<pre><code>## [1] 3 4 5 6 7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">2</span>*y                 <span class="co">#scalar multiplication</span></code></pre></div>
<pre><code>## [1]  2  4  6  8 10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y^<span class="dv">2</span>                 <span class="co">#raise each component to the second power</span></code></pre></div>
<pre><code>## [1]  1  4  9 16 25</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">2</span>^y                 <span class="co">#raise 2 to the first through fifth power</span></code></pre></div>
<pre><code>## [1]  2  4  8 16 32</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y                   <span class="co">#y itself has not been unchanged</span></code></pre></div>
<pre><code>## [1] 1 2 3 4 5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y&lt;-y*<span class="dv">2</span>
y                   <span class="co">#it is now changed</span></code></pre></div>
<pre><code>## [1]  2  4  6  8 10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">r1&lt;-<span class="kw">rep</span>(<span class="dv">1</span>,<span class="dv">3</span>)        <span class="co"># create a vector of 1s, length 3</span>
<span class="kw">length</span>(r1)           <span class="co">#length of the vector</span></code></pre></div>
<pre><code>## [1] 3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(r1)            <span class="co"># class of the vector</span></code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a&lt;-<span class="dv">1</span>                <span class="co"># this is actually a vector length one</span></code></pre></div>
</div>
<div id="matrices" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Matrices</h3>
<p>A matrix refers to a numeric array of rows and columns. You can think of it as a stacked version of vectors where each row or column is a vector. One of the easiest ways to create a matrix is to combine vectors of equal length using <em>cbind()</em>, meaning ‘column bind’.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x&lt;-<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)
y&lt;-<span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>)
m1&lt;-<span class="kw">cbind</span>(x,y);m1</code></pre></div>
<pre><code>##      x y
## [1,] 1 4
## [2,] 2 5
## [3,] 3 6
## [4,] 4 7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">t</span>(m1)                <span class="co"># transpose of m1</span></code></pre></div>
<pre><code>##   [,1] [,2] [,3] [,4]
## x    1    2    3    4
## y    4    5    6    7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(m1)              <span class="co"># 2 by 5 matrix</span></code></pre></div>
<pre><code>## [1] 4 2</code></pre>
<p>You can also directly list the elements and specify the matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m2&lt;-<span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">5</span>,-<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">9</span>),<span class="dt">nrow=</span><span class="dv">3</span>)
m2</code></pre></div>
<pre><code>##      [,1] [,2] [,3]
## [1,]    1    5    2
## [2,]    3   -1    3
## [3,]    2    2    9</code></pre>
<p>Matrices and the next data structure <strong>data frames</strong> are tabular data structures. You can subset them using <strong><a href="genomic-intervals-and-r.html#section-107"></a></strong> and providing desired rows and columns to subset. Here is how that works conceptually: <img src="Intro2R_data/nonR_figures/slicingDataFrames.png" title="slicing/subsetting" alt="alt text" /> <em>slicing/subsetting of a matrix and data frame</em>.There are additional ways to subset data frames, see the next section.</p>
</div>
<div id="data-frames" class="section level3">
<h3><span class="header-section-number">2.3.3</span> Data Frames</h3>
<p>A data frame is more general than a matrix, in that different columns can have different modes (numeric, character, factor, etc.). A data frame can be constructed by data.frame() function. For example, we illustrate how to construct a data frame from genomic intervals or coordinates.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chr &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>, <span class="st">&quot;chr1&quot;</span>, <span class="st">&quot;chr2&quot;</span>, <span class="st">&quot;chr2&quot;</span>)
strand &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;-&quot;</span>,<span class="st">&quot;-&quot;</span>,<span class="st">&quot;+&quot;</span>,<span class="st">&quot;+&quot;</span>)
start&lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">200</span>,<span class="dv">4000</span>,<span class="dv">100</span>,<span class="dv">400</span>)
end&lt;-<span class="kw">c</span>(<span class="dv">250</span>,<span class="dv">410</span>,<span class="dv">200</span>,<span class="dv">450</span>)
mydata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(chr,start,end,strand)
<span class="co">#change column names</span>
<span class="kw">names</span>(mydata) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chr&quot;</span>,<span class="st">&quot;start&quot;</span>,<span class="st">&quot;end&quot;</span>,<span class="st">&quot;strand&quot;</span>)
mydata <span class="co"># OR this will work too</span></code></pre></div>
<pre><code>##    chr start end strand
## 1 chr1   200 250      -
## 2 chr1  4000 410      -
## 3 chr2   100 200      +
## 4 chr2   400 450      +</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mydata &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">chr=</span>chr,<span class="dt">start=</span>start,<span class="dt">end=</span>end,<span class="dt">strand=</span>strand)
mydata</code></pre></div>
<pre><code>##    chr start end strand
## 1 chr1   200 250      -
## 2 chr1  4000 410      -
## 3 chr2   100 200      +
## 4 chr2   400 450      +</code></pre>
<p>There are a variety of ways to extract the elements of a data frame. You can extract certain columns using column numbers or names, or you can extract certain rows by using row numbers. You can also extract data using logical arguments, such as extracting all rows that has a value in a column larger than your threshold.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mydata[,<span class="dv">2</span>:<span class="dv">4</span>] <span class="co"># columns 2,3,4 of data frame</span></code></pre></div>
<pre><code>##   start end strand
## 1   200 250      -
## 2  4000 410      -
## 3   100 200      +
## 4   400 450      +</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mydata[,<span class="kw">c</span>(<span class="st">&quot;chr&quot;</span>,<span class="st">&quot;start&quot;</span>)] <span class="co"># columns chr and start from data frame</span></code></pre></div>
<pre><code>##    chr start
## 1 chr1   200
## 2 chr1  4000
## 3 chr2   100
## 4 chr2   400</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mydata$start <span class="co"># variable start in the data frame</span></code></pre></div>
<pre><code>## [1]  200 4000  100  400</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mydata[<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>),] <span class="co"># get 1st and 3rd rows</span></code></pre></div>
<pre><code>##    chr start end strand
## 1 chr1   200 250      -
## 3 chr2   100 200      +</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mydata[mydata$start&gt;<span class="dv">400</span>,] <span class="co"># get all rows where start&gt;400</span></code></pre></div>
<pre><code>##    chr start end strand
## 2 chr1  4000 410      -</code></pre>
</div>
<div id="lists" class="section level3">
<h3><span class="header-section-number">2.3.4</span> Lists</h3>
<p>An ordered collection of objects (components). A list allows you to gather a variety of (possibly unrelated) objects under one name.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># example of a list with 4 components</span>
<span class="co"># a string, a numeric vector, a matrix, and a scalar</span>
w &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">name=</span><span class="st">&quot;Fred&quot;</span>,
       <span class="dt">mynumbers=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>),
       <span class="dt">mymatrix=</span><span class="kw">matrix</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dt">ncol=</span><span class="dv">2</span>),
       <span class="dt">age=</span><span class="fl">5.3</span>)
w</code></pre></div>
<pre><code>## $name
## [1] &quot;Fred&quot;
## 
## $mynumbers
## [1] 1 2 3
## 
## $mymatrix
##      [,1] [,2]
## [1,]    1    3
## [2,]    2    4
## 
## $age
## [1] 5.3</code></pre>
<p>You can extract elements of a list using the <strong>[<a href="genomic-intervals-and-r.html#section-107"></a>]</strong> convention using either its position in the list or its name.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">w[[<span class="dv">3</span>]] <span class="co"># 3rd component of the list</span></code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]    1    3
## [2,]    2    4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">w[[<span class="st">&quot;mynumbers&quot;</span>]] <span class="co"># component named mynumbers in list</span></code></pre></div>
<pre><code>## [1] 1 2 3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">w$age</code></pre></div>
<pre><code>## [1] 5.3</code></pre>
</div>
<div id="factors" class="section level3">
<h3><span class="header-section-number">2.3.5</span> Factors</h3>
<p>Factors are used to store categorical data. They are important for statistical modeling since categorical variables are treated differently in statistical models than continuous variables. This ensures categorical data treated accordingly in statistical models.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">features=<span class="kw">c</span>(<span class="st">&quot;promoter&quot;</span>,<span class="st">&quot;exon&quot;</span>,<span class="st">&quot;intron&quot;</span>)
f.feat=<span class="kw">factor</span>(features)</code></pre></div>
<p>Important thing to note is that when you are reading a data.frame with read.table() or creating a data frame with <strong>data.frame()</strong> character columns are stored as factors by default, to change this behavior you need to set <strong>stringsAsFactors=FALSE</strong> in <strong>read.table()</strong> and/or <strong>data.frame()</strong> function arguments.</p>
</div>
</div>
<div id="data-types" class="section level2">
<h2><span class="header-section-number">2.4</span> Data types</h2>
<p>There are four common data types in R, they are <strong>numeric</strong>, <strong>logical</strong>, <strong>character</strong> and <strong>integer</strong>. All these data types can be used to create vectors natively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#create a numeric vector x with 5 components</span>
x&lt;-<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">10</span>,<span class="dv">5</span>)
x</code></pre></div>
<pre><code>## [1]  1  3  2 10  5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#create a logical vector x</span>
x&lt;-<span class="kw">c</span>(<span class="ot">TRUE</span>,<span class="ot">FALSE</span>,<span class="ot">TRUE</span>)
x</code></pre></div>
<pre><code>## [1]  TRUE FALSE  TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create a character vector</span>
x&lt;-<span class="kw">c</span>(<span class="st">&quot;sds&quot;</span>,<span class="st">&quot;sd&quot;</span>,<span class="st">&quot;as&quot;</span>)
x</code></pre></div>
<pre><code>## [1] &quot;sds&quot; &quot;sd&quot;  &quot;as&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(x)</code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create an integer vector</span>
x&lt;-<span class="kw">c</span>(1L,2L,3L)
x</code></pre></div>
<pre><code>## [1] 1 2 3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(x)</code></pre></div>
<pre><code>## [1] &quot;integer&quot;</code></pre>
</div>
<div id="reading-and-writing-data" class="section level2">
<h2><span class="header-section-number">2.5</span> Reading and writing data</h2>
<p>Most of the genomics data sets are in the form of genomic intervals associated with a score. That means mostly the data will be in table format with columns denoting chromosome, start positions, end positions, strand and score. One of the popular formats is BED format used primarily by UCSC genome browser but most other genome browsers and tools will support BED format. We have all the annotation data in BED format. In R, you can easily read tabular format data with read.table() function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enh.df &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;intro2R_data/data/subset.enhancers.hg18.bed&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>)  <span class="co"># read enhancer marker BED file</span>
cpgi.df &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;intro2R_data/data/subset.cpgi.hg18.bed&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>) <span class="co"># read CpG island BED file</span>
<span class="co"># check first lines to see how the data looks like</span>
<span class="kw">head</span>(enh.df)</code></pre></div>
<pre><code>##      V1     V2     V3 V4   V5 V6    V7      V8 V9
## 1 chr20 266275 267925  . 1000  .  9.11 13.1693 -1
## 2 chr20 287400 294500  . 1000  . 10.53 13.0231 -1
## 3 chr20 300500 302500  . 1000  .  9.10 13.3935 -1
## 4 chr20 330400 331800  . 1000  .  6.39 13.5105 -1
## 5 chr20 341425 343400  . 1000  .  6.20 12.9852 -1
## 6 chr20 437975 439900  . 1000  .  6.31 13.5184 -1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(cpgi.df)</code></pre></div>
<pre><code>##      V1     V2     V3       V4
## 1 chr20 195575 195851  CpG:_28
## 2 chr20 207789 208148  CpG:_32
## 3 chr20 219055 219437  CpG:_33
## 4 chr20 225831 227155 CpG:_135
## 5 chr20 252826 256323 CpG:_286
## 6 chr20 275376 276977 CpG:_116</code></pre>
<p>You can save your data by writing it to disk as a text file. A data frame or matrix can be written out by using write.table() function. Now let us write out cpgi.df, we will write it out as a tab-separated file, pay attention to the arguments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(cpgi.df,<span class="dt">file=</span><span class="st">&quot;cpgi.txt&quot;</span>,<span class="dt">quote=</span><span class="ot">FALSE</span>,
            <span class="dt">row.names=</span><span class="ot">FALSE</span>,<span class="dt">col.names=</span><span class="ot">FALSE</span>,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</code></pre></div>
<p>You can save your R objects directly into a file using save() and saveRDS() and load them back in with load() and readRDS(). By using these functions you can save any R object whether or not they are in data frame or matrix classes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(cpgi.df,enh.df,<span class="dt">file=</span><span class="st">&quot;mydata.RData&quot;</span>)
<span class="kw">load</span>(<span class="st">&quot;mydata.RData&quot;</span>)
<span class="co"># saveRDS() can save one object at a type</span>
<span class="kw">saveRDS</span>(cpgi.df,<span class="dt">file=</span><span class="st">&quot;cpgi.rds&quot;</span>)
x=<span class="kw">readRDS</span>(<span class="st">&quot;cpgi.rds&quot;</span>)
<span class="kw">head</span>(x)</code></pre></div>
<pre><code>##      V1     V2     V3       V4
## 1 chr20 195575 195851  CpG:_28
## 2 chr20 207789 208148  CpG:_32
## 3 chr20 219055 219437  CpG:_33
## 4 chr20 225831 227155 CpG:_135
## 5 chr20 252826 256323 CpG:_286
## 6 chr20 275376 276977 CpG:_116</code></pre>
<p>One important thing is that with save() you can save many objects at a time and when they are loaded into memory with load() they retain their variable names. For example, in the above code when you use load(“mydata.RData”) in a fresh R session, an object names “cpg.df” will be created. That means you have to figure out what name you gave it to the objects before saving them. On the contrary to that, when you save an object by saveRDS() and read by readRDS() the name of the object is not retained, you need to assign the output of readRDS() to a new variable (“x” in the above code chunk).</p>
</div>
<div id="plotting-in-r" class="section level2">
<h2><span class="header-section-number">2.6</span> Plotting in R</h2>
<p>R has great support for plotting and customizing plots. We will show only a few below. Let us sample 50 values from normal distribution and plot them as a histogram.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># sample 50 values from normal distribution</span>
<span class="co"># and store them in vector x</span>
x&lt;-<span class="kw">rnorm</span>(<span class="dv">50</span>)
<span class="kw">hist</span>(x) <span class="co"># plot the histogram of those values</span></code></pre></div>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-21-1.png" width="672" /></p>
<p>We can modify all the plots by providing certain arguments to the plotting function. Now let’s give a title to the plot using <strong>‘main’</strong> argument. We can also change the color of the bars using <strong>‘col’</strong> argument. You can simply provide the name of the color. Below, we are using <strong>‘red’</strong> for the color. See Figure below for the result this chunk.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(x,<span class="dt">main=</span><span class="st">&quot;Hello histogram!!!&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</code></pre></div>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<p>Next, we will make a scatter plot. Scatter plots are one the most common plots you will encounter in data analysis. We will sample another set of 50 values and plotted those against the ones we sampled earlier. Scatterplot shows values of two variables for a set of data points. It is useful to visualize relationships between two variables. It is frequently used in connection with correlation and linear regression. There are other variants of scatter plots which show density of the points with different colors. We will show examples of those that in following chapters. The scatter plot from our sampling experiment is shown in the figure. Notice that, in addition to main we used <strong>“xlab”</strong> and <strong>“ylab”</strong> arguments to give labels to the plot. You can customize the plots even more than this. See <strong>?plot</strong> and <strong>?par</strong> for more arguments that can help you customize the plots.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># randomly sample 50 points from normal distribution</span>
y&lt;-<span class="kw">rnorm</span>(<span class="dv">50</span>)
<span class="co">#plot a scatter plot</span>
<span class="co"># control x-axis and y-axis labels</span>
<span class="kw">plot</span>(x,y,<span class="dt">main=</span><span class="st">&quot;scatterplot of random samples&quot;</span>,
        <span class="dt">ylab=</span><span class="st">&quot;y values&quot;</span>,<span class="dt">xlab=</span><span class="st">&quot;x values&quot;</span>)</code></pre></div>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-23-1.png" width="672" /></p>
<p>we can also plot boxplots for vectors x and y. Boxplots depict groups of numerical data through their quartiles. The edges of the box denote 1st and 3rd quartile, and the line that crosses the box is the median. Whiskers usually are defined using interquantile range:</p>
<p><em>lowerWhisker=Q1-1.5[IQR] and upperWhisker=Q1+1.5[IQR]</em></p>
<p>In addition, outliers can be depicted as dots. In this case, outliers are the values that remain outside the whiskers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">boxplot</span>(x,y,<span class="dt">main=</span><span class="st">&quot;boxplots of random samples&quot;</span>)</code></pre></div>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-24-1.png" width="672" /> Next up is bar plot which you can plot by <strong>barplot()</strong> function. We are going to plot four imaginary percentage values and color them with two colors, and this time we will also show how to draw a legend on the plot using <strong>legend()</strong> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">perc=<span class="kw">c</span>(<span class="dv">50</span>,<span class="dv">70</span>,<span class="dv">35</span>,<span class="dv">25</span>)
<span class="kw">barplot</span>(<span class="dt">height=</span>perc,<span class="dt">names.arg=</span><span class="kw">c</span>(<span class="st">&quot;CpGi&quot;</span>,<span class="st">&quot;exon&quot;</span>,<span class="st">&quot;CpGi&quot;</span>,<span class="st">&quot;exon&quot;</span>),
        <span class="dt">ylab=</span><span class="st">&quot;percentages&quot;</span>,<span class="dt">main=</span><span class="st">&quot;imagine %s&quot;</span>,
        <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;blue&quot;</span>))
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>,<span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;test&quot;</span>,<span class="st">&quot;control&quot;</span>),<span class="dt">fill=</span><span class="kw">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;blue&quot;</span>))</code></pre></div>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-25-1.png" width="672" /></p>
<p>## Saving plots If you want to save your plots to an image file there are couple of ways of doing that. Normally, you will have to do the following: 1. Open a graphics device 2. Create the plot 3. Close the graphics device</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pdf</span>(<span class="st">&quot;mygraphs/myplot.pdf&quot;</span>,<span class="dt">width=</span><span class="dv">5</span>,<span class="dt">height=</span><span class="dv">5</span>)
<span class="kw">plot</span>(x,y)
<span class="kw">dev.off</span>()</code></pre></div>
<p>Alternatively, you can first create the plot then copy the plot to a graphic device.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(x,y)
<span class="kw">dev.copy</span>(pdf,<span class="st">&quot;mygraphs/myplot.pdf&quot;</span>,<span class="dt">width=</span><span class="dv">7</span>,<span class="dt">height=</span><span class="dv">5</span>)
<span class="kw">dev.off</span>()</code></pre></div>
</div>
<div id="functions-and-control-structures-for-ifelse-etc." class="section level2">
<h2><span class="header-section-number">2.7</span> Functions and control structures (for, if/else etc.)</h2>
<div id="user-defined-functions" class="section level3">
<h3><span class="header-section-number">2.7.1</span> User defined functions</h3>
<p>Functions are useful for transforming larger chunks of code to re-usable pieces of code. Generally, if you need to execute certain tasks with variable parameters then it is time you write a function. A function in R takes different arguments and returns a definite output, much like mathematical functions. Here is a simple function takes two arguments, x and y, and returns the sum of their squares.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sqSum&lt;-function(x,y){
result=x^<span class="dv">2</span>+y^<span class="dv">2</span>
<span class="kw">return</span>(result)
}
<span class="co"># now try the function out</span>
<span class="kw">sqSum</span>(<span class="dv">2</span>,<span class="dv">3</span>)</code></pre></div>
<pre><code>## [1] 13</code></pre>
<p>Functions can also output plots and/or messages to the terminal. Here is a function that prints a message to the terminal:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sqSumPrint&lt;-function(x,y){
result=x^<span class="dv">2</span>+y^<span class="dv">2</span>
<span class="kw">cat</span>(<span class="st">&quot;here is the result:&quot;</span>,result,<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)
}
<span class="co"># now try the function out</span>
<span class="kw">sqSumPrint</span>(<span class="dv">2</span>,<span class="dv">3</span>)</code></pre></div>
<pre><code>## here is the result: 13</code></pre>
<p>Sometimes we would want to execute a certain part of the code only if certain condition is satisfied. This condition can be anything from the type of an object (Ex: if object is a matrix execute certain code), or it can be more complicated such as if object value is between certain thresholds. Let us see how they can be used3. They can be used anywhere in your code, now we will use it in a function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpgi.df &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;intro2R_data/data/subset.cpgi.hg18.bed&quot;</span>, <span class="dt">header =</span> <span class="ot">FALSE</span>)
<span class="co"># function takes input one row</span>
<span class="co"># of CpGi data frame</span>
largeCpGi&lt;-function(bedRow){
 cpglen=bedRow[<span class="dv">3</span>]-bedRow[<span class="dv">2</span>]+<span class="dv">1</span>
 if(cpglen&gt;<span class="dv">1500</span>){
    <span class="kw">cat</span>(<span class="st">&quot;this is large</span><span class="ch">\n</span><span class="st">&quot;</span>)
 }
 else if(cpglen&lt;=<span class="dv">1500</span> &amp;<span class="st"> </span>cpglen&gt;<span class="dv">700</span>){
    <span class="kw">cat</span>(<span class="st">&quot;this is normal</span><span class="ch">\n</span><span class="st">&quot;</span>)
 }
 else{
    <span class="kw">cat</span>(<span class="st">&quot;this is short</span><span class="ch">\n</span><span class="st">&quot;</span>)
 }
}
<span class="kw">largeCpGi</span>(cpgi.df[<span class="dv">10</span>,])
<span class="kw">largeCpGi</span>(cpgi.df[<span class="dv">100</span>,])
<span class="kw">largeCpGi</span>(cpgi.df[<span class="dv">1000</span>,])</code></pre></div>
</div>
<div id="loops-and-looping-structures-in-r" class="section level3">
<h3><span class="header-section-number">2.7.2</span> Loops and looping structures in R</h3>
<p>When you need to repeat a certain task or a execute a function multiple times, you can do that with the help of loops. A loop will execute the task until a certain condition is reached. The loop below is called a “for-loop” and it executes the task sequentially 10 times.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">for(i in <span class="dv">1</span>:<span class="dv">10</span>){ <span class="co"># number of repetitions</span>
<span class="kw">cat</span>(<span class="st">&quot;This is iteration&quot;</span>) <span class="co"># the task to be repeated</span>
<span class="kw">print</span>(i)
}</code></pre></div>
<pre><code>## This is iteration[1] 1
## This is iteration[1] 2
## This is iteration[1] 3
## This is iteration[1] 4
## This is iteration[1] 5
## This is iteration[1] 6
## This is iteration[1] 7
## This is iteration[1] 8
## This is iteration[1] 9
## This is iteration[1] 10</code></pre>
<p>The task above is a bit pointless, normally in a loop, you would want to do something meaningful. Let us calculate the length of the CpG islands we read in earlier. Although this is not the most efficient way of doing that particular task, it serves as a good example for looping. The code below will be execute hundred times, and it will calculate the length of the CpG islands for the first 100 islands in the data frame (by subtracting the end coordinate from the start coordinate).</p>
<p><strong>Note:</strong>If you are going to run a loop that has a lot of repetitions, it is smart to try the loop with few repetitions first and check the results. This will help you make sure the code in the loop works before executing it for thousands of times.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this is where we will keep the lenghts</span>
<span class="co"># for now it is an empty vector</span>
result=<span class="kw">c</span>()
<span class="co"># start the loop</span>
for(i in <span class="dv">1</span>:<span class="dv">100</span>){
    <span class="co">#calculate the length</span>
    len=cpgi.df[i,<span class="dv">3</span>]-cpgi.df[i,<span class="dv">2</span>]+<span class="dv">1</span>
    <span class="co">#append the length to the result</span>
    result=<span class="kw">c</span>(result,len)
}
<span class="co"># check the results</span>
<span class="kw">head</span>(result)</code></pre></div>
<pre><code>## [1]  277  360  383 1325 3498 1602</code></pre>
<div id="apply-family-functions-instead-of-loops" class="section level4">
<h4><span class="header-section-number">2.7.2.1</span> apply family functions instead of loops</h4>
<p>R has other ways of repeating tasks that tend to be more efficient than using loops. They are known as the <strong>“apply”</strong> family of functions, which include <em>apply, lapply, mapply and tapply</em> (and some other variants). All of these functions apply a given function to a set of instances and returns the result of those functions for each instance. The differences between them is that they take different type of inputs. For example apply works on data frames or matrices and applies the function on each row or column of the data structure. <em>lapply</em> works on lists or vectors and applies a function which takes the list element as an argument. Next we will demonstrate how to use <strong>apply()</strong> on a matrix. The example applies the sum function on the rows of a matrix, it basically sums up the values on each row of the matrix, which is conceptualized in Figure below.</p>
<div class="figure">
<img src="Intro2R_data/nonR_figures/apply.png" title="apply concept" alt="alt text" />
<p class="caption">alt text</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat=<span class="kw">cbind</span>(<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">0</span>,<span class="dv">3</span>,<span class="dv">3</span>),<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">3</span>),<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>),<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>),<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>))
result&lt;-<span class="kw">apply</span>(mat,<span class="dv">1</span>,sum)
result</code></pre></div>
<pre><code>## [1] 12  3  5  6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># OR you can define the function as an argument to apply()</span>
result&lt;-<span class="kw">apply</span>(mat,<span class="dv">1</span>,function(x) <span class="kw">sum</span>(x))
result</code></pre></div>
<pre><code>## [1] 12  3  5  6</code></pre>
<p>Notice that we used a second argument which equals to 1, that indicates that rows of the matrix/ data frame will be the input for the function. If we change the second argument to 2, this will indicate that columns should be the input for the function that will be applied. See Figure for the visualization of apply() on columns.</p>
<div class="figure">
<img src="Intro2R_data/nonR_figures/apply2.png" title="apply on columns concept" alt="alt text" />
<p class="caption">alt text</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">result&lt;-<span class="kw">apply</span>(mat,<span class="dv">2</span>,sum)
result</code></pre></div>
<pre><code>## [1] 9 3 6 2 3 3</code></pre>
<p>Next, we will use <strong>lapply()</strong>, which applies a function on a list or a vector. The function that will be applied is a simple function that takes the square of a given number.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input=<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>)
<span class="kw">lapply</span>(input,function(x) x^<span class="dv">2</span>)</code></pre></div>
<pre><code>## [[1]]
## [1] 1
## 
## [[2]]
## [1] 4
## 
## [[3]]
## [1] 9</code></pre>
<p><strong>mapply()</strong> is another member of apply family, it can apply a function on an unlimited set of vectors/lists, it is like a version of lapply that can handle multiple vectors as arguments. In this case, the argument to the mapply() is the function to be applied and the sets of parameters to be supplied as arguments of the function. This conceptualized Figure below, the function to be applied is a function that takes to arguments and sums them up. The arguments to be summed up are in the format of vectors, Xs and Ys. mapply() applies the summation function to each pair in Xs and Ys vector. Notice that the order of the input function and extra arguments are different for mapply. <img src="Intro2R_data/nonR_figures/mapply.png" title="mapply concept" alt="alt text" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Xs=<span class="dv">0</span>:<span class="dv">5</span>
Ys=<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">3</span>,<span class="dv">3</span>)
result&lt;-<span class="kw">mapply</span>(function(x,y) <span class="kw">sum</span>(x,y),Xs,Ys)
result</code></pre></div>
<pre><code>## [1] 2 3 4 6 7 8</code></pre>
</div>
<div id="apply-family-functions-on-multiple-cores" class="section level4">
<h4><span class="header-section-number">2.7.2.2</span> apply family functions on multiple cores</h4>
<p>If you have large data sets apply family functions can be slow (although probably still better than for loops). If that is the case, you can easily use the parallel versions of those functions from parallel package. These functions essentially divide your tasks to smaller chunks run them on separate CPUs and merge the results from those parallel operations. This concept is visualized at Figure below , mcapply runs the summation function on three different processors. Each processor executes the summation function on a part of the data set, and the results are merged and returned as a single vector that has the same order as the input parameters Xs and Ys. <img src="Intro2R_data/nonR_figures/mcmapply.png" title="mcmapply concept" alt="alt text" /></p>
</div>
<div id="vectorized-functions-in-r" class="section level4">
<h4><span class="header-section-number">2.7.2.3</span> Vectorized Functions in R</h4>
<p>The above examples have been put forward to illustrate functions and loops in R because functions using sum() are not complicated and easy to understand. You will probably need to use loops and looping structures with more complicated functions. In reality, most of the operations we used do not need the use of loops or looping structures because there are already vectorized functions that can achieve the same outcomes, meaning if the input arguments are R vectors the output will be a vector as well, so no need for loops or vectorization.</p>
<p>For example, instead of using mapply() and sum() functions we can just use + operator and sum up Xs and Ys.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">result=Xs+Ys
result</code></pre></div>
<pre><code>## [1] 2 3 4 6 7 8</code></pre>
<p>In order to get the column or row sums, we can use the vectorized functions colSums() and rowSums().</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colSums</span>(mat)</code></pre></div>
<pre><code>## [1] 9 3 6 2 3 3</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rowSums</span>(mat)</code></pre></div>
<pre><code>## [1] 12  3  5  6</code></pre>
<p>However, remember that not every function is vectorized in R, use the ones that are. But sooner or later, apply family functions will come in handy.</p>
</div>
</div>
</div>
<div id="session-info" class="section level2">
<h2><span class="header-section-number">2.8</span> Session info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.0 (2016-05-03)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.11.4 (El Capitan)
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.4.5   bookdown_0.0.71 digest_0.6.9    mime_0.4       
##  [5] R6_2.1.2        xtable_1.8-2    formatR_1.4     magrittr_1.5   
##  [9] evaluate_0.9    stringi_1.0-1   miniUI_0.1.1    rmarkdown_0.9.6
## [13] tools_3.3.0     stringr_1.0.0   shiny_0.13.2    httpuv_1.3.3   
## [17] yaml_2.1.13     htmltools_0.3.5 knitr_1.12.28</code></pre>
</div>
<div id="exercises" class="section level2">
<h2><span class="header-section-number">2.9</span> Exercises</h2>
<p>Set your working directory to the source file location in RStudio top menu:</p>
<p>‘Session &gt; Set Working Directory &gt; To Source File Location’</p>
<p>if you have done it correctly you should see this script when you type dir()</p>
<pre><code>&gt; dir()
[1] &quot;intro2R.exercises.html&quot;</code></pre>
<div id="computations-in-r-1" class="section level3">
<h3><span class="header-section-number">2.9.1</span> Computations in R</h3>
<div id="section" class="section level4">
<h4><span class="header-section-number">2.9.1.1</span> </h4>
<p>Sum 2 and 3, use +</p>
</div>
<div id="section-1" class="section level4">
<h4><span class="header-section-number">2.9.1.2</span> </h4>
<p>Take the square root of 36, use sqrt()</p>
</div>
<div id="section-2" class="section level4">
<h4><span class="header-section-number">2.9.1.3</span> </h4>
<p>Take the log10 of 1000, use function <code>log10()</code></p>
</div>
<div id="section-3" class="section level4">
<h4><span class="header-section-number">2.9.1.4</span> </h4>
<p>Take the log2 of 32, use function <code>log2()</code></p>
</div>
<div id="section-4" class="section level4">
<h4><span class="header-section-number">2.9.1.5</span> </h4>
<p>Assign the sum of 2,3 and 4 to variable x</p>
</div>
<div id="section-5" class="section level4">
<h4><span class="header-section-number">2.9.1.6</span> </h4>
<p>Find the absolute value of <code>5 - 145</code> using <code>abs()</code> function</p>
</div>
<div id="section-6" class="section level4">
<h4><span class="header-section-number">2.9.1.7</span> </h4>
<p>Calculate the square root of 625, divide it by 5 and assign it to variable <code>x</code>. Ex: <code>y= log10(1000)/5</code>, the previous statement takes log10 of 1000, divides it by 5 and assigns the value to variable y</p>
</div>
<div id="section-7" class="section level4">
<h4><span class="header-section-number">2.9.1.8</span> </h4>
<p>Multiply the value you get from previous exercise with 10000, assign it variable x Ex: <code>y=y*5</code>, multiplies y with 5 and assigns the value to y.</p>
<p><strong>KEY CONCEPT:</strong> results of computations or arbitrary values can be stored in variables we can re-use those variables later on and over-write them with new values</p>
</div>
</div>
<div id="data-structures-in-r" class="section level3">
<h3><span class="header-section-number">2.9.2</span> Data structures in R</h3>
<div id="section-8" class="section level4">
<h4><span class="header-section-number">2.9.2.1</span> </h4>
<p>Make a vector of 1,2,3,5 and 10 using <code>c()</code>, assign it to <code>vec</code> variable. Ex: <code>vec1=c(1,3,4)</code> makes a vector out of 1,3,4.</p>
</div>
<div id="section-9" class="section level4">
<h4><span class="header-section-number">2.9.2.2</span> </h4>
<p>Check the length of your vector with length(). Ex: <code>length(vec1)</code> should return 3</p>
</div>
<div id="section-10" class="section level4">
<h4><span class="header-section-number">2.9.2.3</span> </h4>
<p>Make a vector of all numbers between 2 and 15. Ex: <code>vec=1:6</code> makes a vector of numbers between 1 and 6, assigns to vec variable</p>
</div>
<div id="section-11" class="section level4">
<h4><span class="header-section-number">2.9.2.4</span> </h4>
<p>Make a vector of 4s repeated 10 times using <code>rep()</code> function. Ex: <code>rep(x=2,times=5)</code> makes a vector of 2s repeated 5 times</p>
</div>
<div id="section-12" class="section level4">
<h4><span class="header-section-number">2.9.2.5</span> </h4>
<p>Make a logical vector with TRUE, FALSE values of length 4, use <code>c()</code>. Ex: <code>c(TRUE,FALSE)</code></p>
</div>
<div id="section-13" class="section level4">
<h4><span class="header-section-number">2.9.2.6</span> </h4>
<p>Make a character vector of gene names PAX6,ZIC2,OCT4 and SOX2. Ex: <code>avec=c(&quot;a&quot;,&quot;b&quot;,&quot;c&quot;)</code> a makes a character vector of a,b and c</p>
</div>
<div id="section-14" class="section level4">
<h4><span class="header-section-number">2.9.2.7</span> </h4>
<p>Subset the vector using <code>[]</code> notation, get 5th and 6th elements. Ex: <code>vec1[1]</code> gets the first element. <code>vec1[c(1,3)]</code> gets 1st and 3rd elements</p>
</div>
<div id="section-15" class="section level4">
<h4><span class="header-section-number">2.9.2.8</span> </h4>
<p>You can also subset any vector using a logical vector in <code>[]</code>. Run the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myvec=<span class="dv">1</span>:<span class="dv">5</span>
myvec[<span class="kw">c</span>(<span class="ot">TRUE</span>,<span class="ot">TRUE</span>,<span class="ot">FALSE</span>,<span class="ot">FALSE</span>,<span class="ot">FALSE</span>)] <span class="co"># the length of the logical vector should be equal to length(myvec) </span>
myvec[<span class="kw">c</span>(<span class="ot">TRUE</span>,<span class="ot">FALSE</span>,<span class="ot">FALSE</span>,<span class="ot">FALSE</span>,<span class="ot">TRUE</span>)]</code></pre></div>
</div>
<div id="section-16" class="section level4">
<h4><span class="header-section-number">2.9.2.9</span> </h4>
<p><code>==,&gt;,&lt;, &gt;=, &lt;=</code> operators create logical vectors. See the results of the following operations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myvec &gt;<span class="st"> </span><span class="dv">3</span>
myvec ==<span class="st"> </span><span class="dv">4</span>
myvec &lt;=<span class="st"> </span><span class="dv">2</span>
myvec !=<span class="st"> </span><span class="dv">4</span></code></pre></div>
</div>
<div id="section-17" class="section level4">
<h4><span class="header-section-number">2.9.2.10</span> </h4>
<p>Use <code>&gt;</code> operator in <code>myvec[ ]</code> to get elements larger than 2 in <code>myvec</code> whic is described above</p>
</div>
<div id="section-18" class="section level4">
<h4><span class="header-section-number">2.9.2.11</span> </h4>
<p>make a 5x3 matrix (5 rows, 3 columns) using <code>matrix()</code>. Ex: matrix(1:6,nrow=3,ncol=2) makes a 3x2 matrix using numbers between 1 and 6</p>
</div>
<div id="section-19" class="section level4">
<h4><span class="header-section-number">2.9.2.12</span> </h4>
<p>What happens when you use <code>byrow = TRUE</code> in your matrix() as an additional argument? Ex: <code>mat=matrix(1:6,nrow=3,ncol=2,byrow = TRUE)</code></p>
</div>
<div id="section-20" class="section level4">
<h4><span class="header-section-number">2.9.2.13</span> </h4>
<p>Extract first 3 columns and first 3 rows of your matrix using <code>[]</code> notation.</p>
</div>
<div id="section-21" class="section level4">
<h4><span class="header-section-number">2.9.2.14</span> </h4>
<p>Extract last two rows of the matrix you created earlier.</p>
<p>Ex: <code>mat[2:3,]</code> or <code>mat[c(2,3),]</code> extracts 2nd and 3rd rows.</p>
</div>
<div id="section-22" class="section level4">
<h4><span class="header-section-number">2.9.2.15</span> </h4>
<p>Extract the first two columns and run <code>class()</code> on the result.</p>
</div>
<div id="section-23" class="section level4">
<h4><span class="header-section-number">2.9.2.16</span> </h4>
<p>Extract first column and run <code>class()</code> on the result, compare with the above exercise.</p>
</div>
<div id="section-24" class="section level4">
<h4><span class="header-section-number">2.9.2.17</span> </h4>
<p>Make a data frame with 3 columns and 5 rows, make sure first column is sequence of numbers 1:5, and second column is a character vector.</p>
<p>Ex: <code>df=data.frame(col1=1:3,col2=c(&quot;a&quot;,&quot;b&quot;,&quot;c&quot;),col3=3:1) # 3x3 data frame</code>.</p>
<p>Remember you need to make 3x5 data frame</p>
</div>
<div id="section-25" class="section level4">
<h4><span class="header-section-number">2.9.2.18</span> </h4>
<p>Extract first two columns and first two rows.</p>
<p><strong>HINT:</strong> Same notation as matrices</p>
</div>
<div id="section-26" class="section level4">
<h4><span class="header-section-number">2.9.2.19</span> </h4>
<p>Extract last two rows of the data frame you made.</p>
<p><strong>HINT:</strong> Same notation as matrices</p>
</div>
<div id="section-27" class="section level4">
<h4><span class="header-section-number">2.9.2.20</span> </h4>
<p>Extract last two columns using column names of the data frame you made.</p>
</div>
<div id="section-28" class="section level4">
<h4><span class="header-section-number">2.9.2.21</span> </h4>
<p>Extract second column using column names. You can use <code>[]</code> or <code>$</code> as in lists, use both in two different answers.</p>
</div>
<div id="section-29" class="section level4">
<h4><span class="header-section-number">2.9.2.22</span> </h4>
<p>Extract rows where 1st column is larger than 3. <strong>HINT:</strong> you can get a logical vector using <code>&gt;</code> operator ,logical vectors can be used in <code>[]</code> when subsetting.</p>
</div>
<div id="section-30" class="section level4">
<h4><span class="header-section-number">2.9.2.23</span> </h4>
<p>Extract rows where 1st column is larger than or equal to 3.</p>
</div>
<div id="section-31" class="section level4">
<h4><span class="header-section-number">2.9.2.24</span> </h4>
<p>Convert data frame to the matrix. <strong>HINT:</strong> use <code>as.matrix()</code>.</p>
<p>Observe what happens to numeric values in the data frame.</p>
</div>
<div id="section-32" class="section level4">
<h4><span class="header-section-number">2.9.2.25</span> </h4>
<p>Make a list using <code>list()</code> function, your list should have 4 elements the one below has 2.</p>
<p>Ex: <code>mylist= list(a=c(1,2,3),b=c(&quot;apple,&quot;orange&quot;))</code></p>
</div>
<div id="section-33" class="section level4">
<h4><span class="header-section-number">2.9.2.26</span> </h4>
<p>Select the 1st element of the list you made using <code>$</code> notation.</p>
<p>Ex: <code>mylist$a</code> selects first element named “a”</p>
</div>
<div id="section-34" class="section level4">
<h4><span class="header-section-number">2.9.2.27</span> </h4>
<p>Select the 4th element of the list you made earlier using <code>$</code> notation.</p>
</div>
<div id="section-35" class="section level4">
<h4><span class="header-section-number">2.9.2.28</span> </h4>
<p>Select the 1st element of your list using <code>[ ]</code> notation.</p>
<p>Ex: <code>mylist[1]</code> selects first element named “a”, you get a list with one element.</p>
<p>Ex: <code>mylist[&quot;a&quot;]</code> selects first element named “a”, you get a list with one element.</p>
</div>
<div id="section-36" class="section level4">
<h4><span class="header-section-number">2.9.2.29</span> </h4>
<p>select the 4th element of your list using ‘<a href="genomic-intervals-and-r.html#section-107"></a>’ notation.</p>
</div>
<div id="section-37" class="section level4">
<h4><span class="header-section-number">2.9.2.30</span> </h4>
<p>Make a factor using factor(), with 5 elements. Ex: <code>fa=factor(c(&quot;a&quot;,&quot;a&quot;,&quot;b&quot;))</code></p>
</div>
<div id="section-38" class="section level4">
<h4><span class="header-section-number">2.9.2.31</span> </h4>
<p>Convert a character vector to factor using <code>as.factor()</code>. First, make a character vector using <code>c()</code> then use <code>as.factor()</code>.</p>
</div>
<div id="section-39" class="section level4">
<h4><span class="header-section-number">2.9.2.32</span> </h4>
<p>Convert the factor you made above to character using <code>as.character()</code>.</p>
</div>
</div>
<div id="reading-in-and-writing-data-out-in-r" class="section level3">
<h3><span class="header-section-number">2.9.3</span> Reading in and writing data out in R</h3>
<div id="section-40" class="section level4">
<h4><span class="header-section-number">2.9.3.1</span> </h4>
<p>Read CpG island (CpGi) data from <code>intro2R_data/data/CpGi.table.hg18.txt</code>, this is a tab-separated file, store it in a variable called “cpgi”.</p>
</div>
<div id="section-41" class="section level4">
<h4><span class="header-section-number">2.9.3.2</span> </h4>
<p>Use <code>head()</code> on CpGi to see first few rows.</p>
</div>
<div id="section-42" class="section level4">
<h4><span class="header-section-number">2.9.3.3</span> </h4>
<p>Why doesn’t the following work? see ‘sep’ argument at <code>help(read.table)</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpgiSepComma=<span class="kw">read.table</span>(<span class="st">&quot;intro2R_data/data/CpGi.table.hg18.txt&quot;</span>,<span class="dt">header=</span><span class="ot">TRUE</span>,<span class="dt">sep=</span><span class="st">&quot;,&quot;</span>)
<span class="kw">head</span>(cpgiSepComma)</code></pre></div>
<pre><code>##   chrom.chromStart.chromEnd.name.length.cpgNum.gcNum.perCpg.perGc.obsExp
## 1         chr1\t18598\t19673\tCpG: 116\t1075\t116\t787\t21.6\t73.2\t0.83
## 2          chr1\t124987\t125426\tCpG: 30\t439\t30\t295\t13.7\t67.2\t0.64
## 3          chr1\t317653\t318092\tCpG: 29\t439\t29\t295\t13.2\t67.2\t0.62
## 4         chr1\t427014\t428027\tCpG: 84\t1013\t84\t734\t16.6\t72.5\t0.64
## 5         chr1\t439136\t440407\tCpG: 99\t1271\t99\t777\t15.6\t61.1\t0.84
## 6            chr1\t523082\t523977\tCpG: 94\t895\t94\t570\t21\t63.7\t1.04</code></pre>
</div>
<div id="section-43" class="section level4">
<h4><span class="header-section-number">2.9.3.4</span> </h4>
<p>What happens when <code>stringsAsFactors=FALSE</code> ? <code>cpgiHF=read.table(&quot;intro2R_data/data/CpGi.table.hg18.txt&quot;,header=FALSE,sep=&quot;\t&quot;,                   stringsAsFactors=FALSE)</code></p>
</div>
<div id="section-44" class="section level4">
<h4><span class="header-section-number">2.9.3.5</span> </h4>
<p>Read only first 10 rows of the CpGi table.</p>
</div>
<div id="section-45" class="section level4">
<h4><span class="header-section-number">2.9.3.6</span> </h4>
<p>Use <code>read.table(&quot;intro2R_data/data/CpGi.table.hg18.txt&quot;,...</code> with <code>header=FALSE</code>.</p>
<p>Do <code>head()</code> to see the results.</p>
</div>
<div id="section-46" class="section level4">
<h4><span class="header-section-number">2.9.3.7</span> </h4>
<p>Write CpG islands to a text file called “my.cpgi.file.txt”. Write the file to your home folder, you can use <code>file=&quot;~/my.cpgi.file.txt&quot;</code> in linux. <code>~/</code> denotes home folder.</p>
</div>
<div id="section-47" class="section level4">
<h4><span class="header-section-number">2.9.3.8</span> </h4>
<p>Same as above but this time make sure use <code>quote=FALSE</code>,<code>sep=&quot;\t&quot;</code> and <code>row.names=FALSE</code> arguments. Save the file to “my.cpgi.file2.txt” and compare it with “my.cpgi.file.txt”</p>
</div>
<div id="section-48" class="section level4">
<h4><span class="header-section-number">2.9.3.9</span> </h4>
<p>Write out the first 10 rows of ‘cpgi’ data frame. <strong>HINT:</strong> use subsetting for data frames we learned before.</p>
</div>
<div id="section-49" class="section level4">
<h4><span class="header-section-number">2.9.3.10</span> </h4>
<p>Write the first 3 columns of ‘cpgi’ data frame.</p>
</div>
<div id="section-50" class="section level4">
<h4><span class="header-section-number">2.9.3.11</span> </h4>
<p>Write CpG islands only on chr1. <strong>HINT:</strong> use subsetting with <code>[]</code>, feed a logical vector using <code>==</code> operator.</p>
</div>
<div id="section-51" class="section level4">
<h4><span class="header-section-number">2.9.3.12</span> </h4>
<p>Read two other data sets “intro2R_data/data/rn4.refseq.bed” and “intro2R_data/data/rn4.refseq2name.txt” with <code>header=FALSE</code>, assign them to df1 and df2 respectively.</p>
</div>
<div id="section-52" class="section level4">
<h4><span class="header-section-number">2.9.3.13</span> </h4>
<p>Use <code>head()</code> to see what is inside of the the data frames above.</p>
</div>
<div id="section-53" class="section level4">
<h4><span class="header-section-number">2.9.3.14</span> </h4>
<p>Merge data sets using <code>merge()</code> and assign the results to variable named ‘new.df’, and use <code>head()</code> to see the results.</p>
</div>
</div>
<div id="plotting-in-r-1" class="section level3">
<h3><span class="header-section-number">2.9.4</span> Plotting in R</h3>
<p>Please <strong>run the following</strong> for the rest of the exercises.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1001</span>)
x1=<span class="dv">1</span>:<span class="dv">100</span>+<span class="kw">rnorm</span>(<span class="dv">100</span>,<span class="dt">mean=</span><span class="dv">0</span>,<span class="dt">sd=</span><span class="dv">15</span>)
y1=<span class="dv">1</span>:<span class="dv">100</span></code></pre></div>
<div id="section-54" class="section level4">
<h4><span class="header-section-number">2.9.4.1</span> </h4>
<p>Make a scatter plot using <code>x1</code> and <code>y1</code> vectors generated above.</p>
</div>
<div id="section-55" class="section level4">
<h4><span class="header-section-number">2.9.4.2</span> </h4>
<p>Use <code>main</code> argument to give a title to <code>plot()</code> as in <code>plot(x,y,main=&quot;title&quot;)</code></p>
</div>
<div id="section-56" class="section level4">
<h4><span class="header-section-number">2.9.4.3</span> </h4>
<p>Use <code>xlab</code> argument to set a label to x-axis.Use <code>ylab</code> argument to set a label to y-axis.</p>
</div>
<div id="section-57" class="section level4">
<h4><span class="header-section-number">2.9.4.4</span> </h4>
<p>See what <code>mtext(side=3,text=&quot;hi there&quot;)</code> does. <strong>HINT:</strong> mtext stands for margin text.</p>
</div>
<div id="section-58" class="section level4">
<h4><span class="header-section-number">2.9.4.5</span> </h4>
<p>See what <code>mtext(side=2,text=&quot;hi there&quot;)</code> does.check your plot after execution.</p>
</div>
<div id="section-59" class="section level4">
<h4><span class="header-section-number">2.9.4.6</span> </h4>
<p>You can use <code>paste()</code> as ‘text’ argument in mtext() try that, you need to re-plot. your plot first. <strong>HINT:</strong> <code>mtext(side=3,text=paste(...))</code> See what <code>paste()</code> is used for.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">paste</span>(<span class="st">&quot;Text&quot;</span>,<span class="st">&quot;here&quot;</span>)</code></pre></div>
<pre><code>## [1] &quot;Text here&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myText=<span class="kw">paste</span>(<span class="st">&quot;Text&quot;</span>,<span class="st">&quot;here&quot;</span>)
myText</code></pre></div>
<pre><code>## [1] &quot;Text here&quot;</code></pre>
<p>Use <em>mtext()</em> and <em>paste()</em> to put a margin text on the plot.</p>
</div>
<div id="section-60" class="section level4">
<h4><span class="header-section-number">2.9.4.7</span> </h4>
<p><code>cor()</code> calculates correlation between two vectors. Pearson correlation is a measure of the linear correlation (dependence) between two variables X and Y.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">corxy=<span class="kw">cor</span>(x1,y1) <span class="co"># calculates pearson correlation</span></code></pre></div>
</div>
<div id="section-61" class="section level4">
<h4><span class="header-section-number">2.9.4.8</span> </h4>
<p>Try use <code>mtext()</code>,<code>cor()</code> and <code>paste()</code> to display correlation coefficient on your scatterplot ?</p>
</div>
<div id="section-62" class="section level4">
<h4><span class="header-section-number">2.9.4.9</span> </h4>
<p>Change the colors of your plot using <code>col</code> argument. Ex: <code>plot(x,y,col=&quot;red&quot;)</code></p>
</div>
<div id="section-63" class="section level4">
<h4><span class="header-section-number">2.9.4.10</span> </h4>
<p>Use <code>pch=19</code> as an argument in your <code>plot()</code> command.</p>
</div>
<div id="section-64" class="section level4">
<h4><span class="header-section-number">2.9.4.11</span> </h4>
<p>Use <code>pch=18</code> as an argument to your <code>plot()</code> command.</p>
</div>
<div id="section-65" class="section level4">
<h4><span class="header-section-number">2.9.4.12</span> </h4>
<p>Make histogram of x1 with <code>hist()</code> function.Histogram is a graphical representation of the data distribution.</p>
</div>
<div id="section-66" class="section level4">
<h4><span class="header-section-number">2.9.4.13</span> </h4>
<p>You can change colors with ‘col’, add labels with ‘xlab’, ‘ylab’, and add a ‘title’ with ‘main’ arguments. Try all these in a histogram.</p>
</div>
<div id="section-67" class="section level4">
<h4><span class="header-section-number">2.9.4.14</span> </h4>
<p>Make boxplot of y1 with <code>boxplot()</code>.</p>
</div>
<div id="section-68" class="section level4">
<h4><span class="header-section-number">2.9.4.15</span> </h4>
<p>Make boxplots of <code>x1</code> and <code>y1</code> vectors in the same plot.</p>
</div>
<div id="section-69" class="section level4">
<h4><span class="header-section-number">2.9.4.16</span> </h4>
<p>In boxplot use horizontal = TRUE argument</p>
</div>
<div id="section-70" class="section level4">
<h4><span class="header-section-number">2.9.4.17</span> </h4>
<p>make multiple plots with par(mfrow=c(2,1)) 1. run par(mfrow=c(2,1)) 2. make a boxplot 3. make a histogram</p>
</div>
<div id="section-71" class="section level4">
<h4><span class="header-section-number">2.9.4.18</span> </h4>
<p>Do the same as above but this time with par(mfrow=c(1,2))</p>
</div>
<div id="section-72" class="section level4">
<h4><span class="header-section-number">2.9.4.19</span> </h4>
<p>Save your plot using “Export” button</p>
</div>
<div id="section-73" class="section level4">
<h4><span class="header-section-number">2.9.4.20</span> </h4>
<p>Save your plot by running :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dev.copy</span>(pdf,<span class="dt">file=</span><span class="st">&quot;plot.file.pdf&quot;</span>);<span class="kw">dev.off</span>()</code></pre></div>
</div>
<div id="section-74" class="section level4">
<h4><span class="header-section-number">2.9.4.21</span> </h4>
<p>Save your plot running :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dev.copy</span>(png,<span class="dt">filename=</span><span class="st">&quot;plot.file.png&quot;</span>);<span class="kw">dev.off</span>()</code></pre></div>
</div>
<div id="section-75" class="section level4">
<h4><span class="header-section-number">2.9.4.22</span> </h4>
<p>Another way to save the plot is the following</p>
<ol style="list-style-type: decimal">
<li>Open a graphics device</li>
<li>Create the plot</li>
<li>Close the graphics device</li>
</ol>
</div>
<div id="section-76" class="section level4">
<h4><span class="header-section-number">2.9.4.23</span> </h4>
<p><strong>EXTRA: Making color density scatterplot</strong>. You can make a scatter plot showing density of points rather than points themselves. If you use points it looks like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x2=<span class="dv">1</span>:<span class="dv">1000</span>+<span class="kw">rnorm</span>(<span class="dv">1000</span>,<span class="dt">mean=</span><span class="dv">0</span>,<span class="dt">sd=</span><span class="dv">200</span>)
y2=<span class="dv">1</span>:<span class="dv">1000</span>
<span class="kw">plot</span>(x2,y2,<span class="dt">pch=</span><span class="dv">19</span>,<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-117-1.png" width="672" /></p>
<p>If you use smoothScatter() function, you get the densities.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">smoothScatter</span>(x2,y2,<span class="dt">colramp=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;white&quot;</span>,<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>,<span class="st">&quot;yellow&quot;</span>,<span class="st">&quot;red&quot;</span>))) </code></pre></div>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-118-1.png" width="672" /></p>
<p>Now, plot with <code>colramp=heat.colors</code> argument and then use a custom color scale using the following argument.</p>
<pre><code>colramp = colorRampPalette(c(&quot;white&quot;,&quot;blue&quot;, &quot;green&quot;,&quot;yellow&quot;,&quot;red&quot;)))</code></pre>
</div>
</div>
<div id="functions-and-control-structures-for-ifelse-etc.-1" class="section level3">
<h3><span class="header-section-number">2.9.5</span> Functions and control structures (for, if/else etc.)</h3>
<div id="section-77" class="section level4">
<h4><span class="header-section-number">2.9.5.1</span> </h4>
<p>Read CpG island data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpgi=<span class="kw">read.table</span>(<span class="st">&quot;intro2R_data/data/CpGi.table.hg18.txt&quot;</span>,<span class="dt">header=</span><span class="ot">TRUE</span>,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
<span class="kw">head</span>(cpgi)</code></pre></div>
<pre><code>##   chrom chromStart chromEnd     name length cpgNum gcNum perCpg perGc
## 1  chr1      18598    19673 CpG: 116   1075    116   787   21.6  73.2
## 2  chr1     124987   125426  CpG: 30    439     30   295   13.7  67.2
## 3  chr1     317653   318092  CpG: 29    439     29   295   13.2  67.2
## 4  chr1     427014   428027  CpG: 84   1013     84   734   16.6  72.5
## 5  chr1     439136   440407  CpG: 99   1271     99   777   15.6  61.1
## 6  chr1     523082   523977  CpG: 94    895     94   570   21.0  63.7
##   obsExp
## 1   0.83
## 2   0.64
## 3   0.62
## 4   0.64
## 5   0.84
## 6   1.04</code></pre>
</div>
<div id="section-78" class="section level4">
<h4><span class="header-section-number">2.9.5.2</span> </h4>
<p>Check values at perGc column using a histogram. ‘perGc’ stands for GC percent =&gt; percentage of C+G nucleotides</p>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-121-1.png" width="672" /></p>
</div>
<div id="section-79" class="section level4">
<h4><span class="header-section-number">2.9.5.3</span> </h4>
<p>Make a boxplot for ‘perGc’ column</p>
<p><img src="compgenomr_files/figure-html/unnamed-chunk-122-1.png" width="672" /></p>
</div>
<div id="section-80" class="section level4">
<h4><span class="header-section-number">2.9.5.4</span> </h4>
<p>Use if/else structure to decide if given GC percent high, low or medium. If it is low, high, or medium. low &lt; 60, high&gt;75, medium is between 60 and 75 use greater or less than operators &lt; or &gt; . Fill in the values in the in code below, where it is written ‘YOU_FILL_IN’</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GCper=<span class="dv">65</span>
 <span class="co">#result=&quot;low&quot;# set initial value</span>
  
  if(GCper &lt;<span class="st"> </span><span class="dv">60</span>){ <span class="co"># check if GC value is lower than 60, assign &quot;low&quot; to result</span>
    result=<span class="st">&quot;low&quot;</span>
    <span class="kw">cat</span>(<span class="st">&quot;low&quot;</span>)
  }
  else if(GCper &gt;<span class="st"> </span><span class="dv">75</span>){  <span class="co"># check if GC value is higher than 75, assign &quot;high&quot; to result</span>
    result=<span class="st">&quot;high&quot;</span>
    <span class="kw">cat</span>(<span class="st">&quot;high&quot;</span>)
  }else{ <span class="co"># if those two conditions fail then it must be &quot;medium&quot;</span>
    result=<span class="st">&quot;medium&quot;</span>
  }

result</code></pre></div>
</div>
<div id="section-81" class="section level4">
<h4><span class="header-section-number">2.9.5.5</span> </h4>
<p>Write a function that takes a value of GC percent and decides if it is low, high, or medium. low &lt; 60, high&gt;75, medium is between 60 and 75. Fill in the values in the in code below, where it is written ‘YOU_FILL_IN’</p>
<pre><code>GCclass&lt;-function(my.gc){
  
  YOU_FILL_IN
  
  return(result)
}
GCclass(10) # should return &quot;low&quot;
GCclass(90) # should return &quot;high&quot;
GCclass(65) # should return &quot;medium&quot;</code></pre>
</div>
<div id="section-82" class="section level4">
<h4><span class="header-section-number">2.9.5.6</span> </h4>
<p>Use a for loop to get GC percentage classes for gcValues below. Use the function you wrote above.</p>
<pre><code>gcValues=c(10,50,70,65,90)
for( i in YOU_FILL_IN){
  YOU_FILL_IN
}</code></pre>
</div>
<div id="section-83" class="section level4">
<h4><span class="header-section-number">2.9.5.7</span> </h4>
<p>Use <code>lapply</code> to get to get GC percentage classes for gcValues. Example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vec=<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">5</span>)
power2=function(x){ <span class="kw">return</span>(x^<span class="dv">2</span>)  }
    <span class="kw">lapply</span>(vec,power2)</code></pre></div>
</div>
<div id="section-84" class="section level4">
<h4><span class="header-section-number">2.9.5.8</span> </h4>
<p>Use sapply to get values to get GC percentage classes for gcValues</p>
</div>
<div id="section-85" class="section level4">
<h4><span class="header-section-number">2.9.5.9</span> </h4>
<p>Is there a way to decide on the GC percentage class of given vector of GCpercentages without using if/else structure and loops ? if so, how can you do it? <strong>HINT:</strong> subsetting using &lt; and &gt; operators</p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-genomics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="statistics-for-genomics.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="book_assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="book_assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="book_assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="book_assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
